{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2; Supplementary Figure 3; Supplementary Table 3\n",
    "\n",
    "Analysis for comparisons of PWM data generated from universal UV-Bind data. \n",
    "\n",
    "\n",
    "### Overview - Model alignment and trimming\n",
    "\n",
    "PWM files generated from Seed-and-Wobble are processed and compared. The models are by default an arbitrary length and an arbitrary orientation. To directly compare models across Non-UV and UV conditions of the same trancription factor (TF) these need to be addressed. For each transcription factor:\n",
    "\n",
    "1. A consensus sequence of the binding core is given to define a binding site. \n",
    "2. This sequence is scored across the PWM in both orientations. The top scoring position is considered the binding site. \n",
    "3. The submatrix for non-UV of the top scoring position is then compared to the UV matrix using KL divergence.\n",
    "4. The position of minimum divergence is considered the aligned position between the two models. \n",
    "5. The models are then extended by 3 to account for flanking sequence.\n",
    "\n",
    "\n",
    "### 3rd Party Packages\n",
    "\n",
    "1. Logomaker - Plotting PWM logos\n",
    "2. Matplotlib - Adjusting parameters for logomaker plots\n",
    "3. Numpy - Array usage\n",
    "4. Pandas - Dataframe usage\n",
    "5. Scipy - KL Divergence\n",
    "\n",
    "### UV Bind Analysis Core Imports\n",
    "\n",
    "- uac.PYDI: Tuple of pyrimidine dinucleotides and their reverse complements\n",
    "\n",
    "** Additional details can be found in the uvbind_analysis_core.py script.\n",
    "\n",
    "### Abbreviations:\n",
    "\n",
    "- pydi: Pyrimidine Dinucleotide\n",
    "- df: DataFrame\n",
    "- pdf: Probability DataFrame\n",
    "- kl: Kullbackâ€“Leibler\n",
    "- ic: Information Content\n",
    "- dict: Dictionary\n",
    "- pwm: Position Weight Matrix\n",
    "- nuv: Non-UV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports and Meta Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "from __future__ import annotations\n",
    "import os\n",
    "\n",
    "from bokeh.io import export_svg\n",
    "from bokeh.models import VBar\n",
    "from bokeh.plotting import figure, ColumnDataSource\n",
    "import logomaker\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "import pandas as pd\n",
    "import scipy.special\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "\n",
    "import uvbind_analysis_core as uac\n",
    "\n",
    "\n",
    "# Global Variables\n",
    "# Input folders\n",
    "DATA_FOLDER = \"../../Data/PWM_Files\"\n",
    "META_DATA_FILE = \"Meta_Data/Meta_Data_PWM_Comparisons.csv\"\n",
    "# Output folders\n",
    "OUTPUT_FOLDER_F2 = \"../Figure_2\"\n",
    "OUTPUT_FOLDER_SUPP = \"../Figure_S3\"\n",
    "OUTPUT_FOLDER_T = \"../Table_S3\"\n",
    "# Colors for PWM logos\n",
    "LOGO_COLOR_SCHEME = {\"A\":(0, 0.8, 0), # R G B\n",
    "                     \"C\":(0, 0, 0.8),\n",
    "                     \"G\":(0.8, 0.8, 0.1),\n",
    "                     \"T\":(0.8, 0, 0.1)}\n",
    "# Color for PWM probability differences boxplot and bargraphs\n",
    "# Color is UV conditions relative to non UV conditions\n",
    "INCREASE_COLOR = \"#8db0fe\"\n",
    "DECREASE_COLOR = \"#f4987a\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "for folder in (OUTPUT_FOLDER_F2, OUTPUT_FOLDER_SUPP, OUTPUT_FOLDER_T):\n",
    "    os.makedirs(folder, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2B-C; Figure 2S: PWM trimming and alignment\n",
    "\n",
    "PWM files generated from Seed-and-Wobble are processed and compared. The models are by default an arbitrary length and an arbitrary orientation. To directly compare models across Non-UV and UV conditions of the same trancription factor (TF) these need to be addressed. For each transcription factor:\n",
    "\n",
    "1. A consensus sequence of the binding core is given to define a binding site. \n",
    "2. This sequence is scored across the PWM in both orientations. The top scoring position is considered the binding site. \n",
    "3. The submatrix for non-UV of the top scoring position is then compared to the UV matrix using KL divergence.\n",
    "4. The position of minimum divergence is considered the aligned position between the two models. \n",
    "5. The models are then extended by 3 to account for flanking sequence.\n",
    "\n",
    "\n",
    "#### (1) Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matrix_position(file_path: str, matrix_name: str) -> int:\n",
    "    \"\"\"Returns number of lines to skip for given matrix name\n",
    "\n",
    "    :param file_path: Relative path to the PWM file\n",
    "    :type file_path: str\n",
    "    :param matrix_name: Name of the matrix (Probability, Enrichment, etc.)\n",
    "    :type matrix_name: str\n",
    "    :returns: Number of lines to skip when reading the file\n",
    "    \"\"\"\n",
    "    # Parse file for matrix to read\n",
    "    skip = -1\n",
    "    file_object = open(file_path, 'r')\n",
    "    for idx, line in enumerate(file_object):\n",
    "        if line.startswith(matrix_name):\n",
    "            skip = idx + 2\n",
    "            break\n",
    "    file_object.close()\n",
    "    if skip == -1:\n",
    "        raise ValueError(f\"Cannot find {matrix_name} in file.\")\n",
    "    return skip\n",
    "\n",
    "\n",
    "def read_pwm_matrix(file_path: str,\n",
    "                    matrix_name: str = \"Prob\") -> pd.DataFrame:\n",
    "    \"\"\"Returns top PWM from a PWM Seed-and-Wobble output\n",
    "\n",
    "    :param file_path: Relative path to the PWM file\n",
    "    :type file_path: str\n",
    "    :param skip: Number of rows to skip in reading the probability matrix\n",
    "    :type skip: int\n",
    "    :returns: PWM matrix in long form (columns = A, C, G, T)\n",
    "    \"\"\"\n",
    "    skip = matrix_position(file_path, matrix_name)\n",
    "    # Read matrix\n",
    "    pwm = pd.read_csv(file_path,\n",
    "                      sep='\\t',\n",
    "                      skiprows=skip,\n",
    "                      header=None,\n",
    "                      nrows=4)\n",
    "    pwm = pwm[pwm.columns[1:]]\n",
    "    pwm_long = pwm.T\n",
    "    pwm_long = pwm_long.rename(columns={0: 'A', 1: 'C', 2: 'G', 3: 'T'})\n",
    "    return pwm_long.reset_index(drop=True)\n",
    "\n",
    "\n",
    "def reverse_complement_matrix(pwm_long: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Reverse complement of long form PWM matrix\n",
    "\n",
    "    :param pwm_long: Long form PWM matrix\n",
    "    :type pwm_long: pd.DataFrame\n",
    "    :returns: Reverse complement of input matrix as a pd.Dataframe\n",
    "    \"\"\"\n",
    "    pwm = pwm_long.iloc[::-1, :]\n",
    "    pwm = pwm.reset_index(drop=True)\n",
    "    pwm = pwm.rename(columns={'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'})\n",
    "    pwm = pwm[['A', 'C', 'G', 'T']]\n",
    "    return pwm\n",
    "\n",
    "\n",
    "def plot_ic_matrix_logo(matrix: pd.DataFrame, ax, color_scheme) -> logomaker.Logo:\n",
    "    \"\"\"Save an information content logo from a long format matrix to output\"\"\"\n",
    "    logo = logomaker.Logo(matrix,ax=ax, color_scheme=color_scheme)\n",
    "    logo.style_spines(spines=[\"left\", \"bottom\"], visible=True)\n",
    "    # No tick marks or labels\n",
    "    logo.ax.set_xticks([])\n",
    "    logo.ax.set_yticks([])\n",
    "    logo.ax.set_yticklabels([])\n",
    "    # ylim is 0, 2 since the PWM information content ranges from 0 to 2\n",
    "    logo.ax.set_ylim([0, 2])\n",
    "    return logo\n",
    "\n",
    "\n",
    "def max_scoring_position(kmer: str,\n",
    "                         pwm: pd.DataFrame) -> tuple[int, float]:\n",
    "    \"\"\"Return the position of the max score and that score.\n",
    "\n",
    "    Given a kmer and pwm in long format, finds the maximum\n",
    "    scoring position of the kmer and returns a tuple of the\n",
    "    position and score.\n",
    "    \"\"\"\n",
    "    max_score = 0\n",
    "    max_position = -1\n",
    "    for i in range(len(pwm) - len(kmer) + 1):\n",
    "        # Get score of local range\n",
    "        subscore = 1\n",
    "        for idx, j in enumerate(kmer):\n",
    "            probability = pwm[j][i + idx]\n",
    "            subscore *= probability\n",
    "        if subscore > max_score:\n",
    "            max_score = subscore\n",
    "            max_position = i\n",
    "    return (max_position, max_score)\n",
    "\n",
    "\n",
    "def top_submatrix_from_kmer(kmer: str,\n",
    "                            pwm: pd.DataFrame,\n",
    "                            extend=3) -> pd.DataFrame:\n",
    "    \"\"\"Return the top scoring submatrix in either orientation from a kmer.\n",
    "    \n",
    "    Given a kmer, pwm in long format, and a parameter to extend the match,\n",
    "    return a submatrix of the pwm that overlaps with the top scoring position\n",
    "    of the kmer extended by the number of positions in the extend parameter if\n",
    "    possible. Return a submatrix in long format.\n",
    "    \"\"\"\n",
    "    orient_a = max_scoring_position(kmer, pwm)\n",
    "    pwm_reverse_complement = reverse_complement_matrix(pwm)\n",
    "    orient_b = max_scoring_position(kmer, pwm_reverse_complement)\n",
    "    if orient_a[1] > orient_b[1]:\n",
    "        orient = orient_a\n",
    "        pwm_oriented = pwm\n",
    "    else:\n",
    "        orient = orient_b\n",
    "        pwm_oriented = pwm_reverse_complement\n",
    "    start = max(0, orient[0] - extend)\n",
    "    end = min(len(pwm_oriented) - 1, orient[0] + len(kmer) + extend)\n",
    "    return pwm_oriented.iloc[start:end, :].reset_index(drop=True)\n",
    "\n",
    "\n",
    "def kl_divergence(p, q):\n",
    "    \"\"\"Kullback-Leibler divergence in bits\"\"\"\n",
    "    return np.sum(np.where(p != 0, p * np.log2(p / q), 0))\n",
    "\n",
    "\n",
    "def kl_divergence_sum(pwm_a: pd.DataFrame,\n",
    "                      pwm_b: pd.DataFrame) -> float:\n",
    "    \"\"\"Returns the sum of divergences\"\"\"\n",
    "    kl_sum = 0\n",
    "    for i in range(len(pwm_a)):\n",
    "        np_matrix_a = np.array(list(pwm_a.T[i]))\n",
    "        np_matrix_b = np.array(list(pwm_b.T[i]))\n",
    "        kl_sum += kl_divergence(np_matrix_a, np_matrix_b)\n",
    "    return kl_sum\n",
    "\n",
    "\n",
    "def min_kl_divergence_overlap(pwm_small: pd.DataFrame,\n",
    "                              pwm_large: pd.DataFrame):\n",
    "    \"\"\"Position and value of the minimum KL divergence.\n",
    "    \n",
    "    Given a pwm of less than or equal length (pwm_small) and a larger or \n",
    "    equal size pwm, returns a tuple conainting the following:\n",
    "    1. Position of minimal divergence\n",
    "    2. Value of minimal divergence\n",
    "    \"\"\"\n",
    "    best_match_position = -1\n",
    "    best_match_distance = float('inf')\n",
    "    for i in range(len(pwm_large) - len(pwm_small) + 1):\n",
    "        start = i\n",
    "        end = i + len(pwm_small)\n",
    "        pwm_b = pwm_large.iloc[start:end, :].reset_index(drop=True)\n",
    "        distance = kl_divergence_sum(pwm_small,\n",
    "                                     pwm_b)\n",
    "        if distance < best_match_distance:\n",
    "            best_match_position = i\n",
    "            best_match_distance = distance\n",
    "    return (best_match_position, best_match_distance)\n",
    "\n",
    "\n",
    "def best_match(pwm_a: pd.DataFrame,\n",
    "               pwm_b: pd.DataFrame,\n",
    "               consensus: str,\n",
    "               extend: int = 3) -> pd.DataFrame:\n",
    "    \"\"\"Returns the pwm_b subset with the minimum KL divergence from pwm_a.\"\"\"\n",
    "    # Return the pwm_a subset of the top scoring overlap of the consensus sequence.\n",
    "    pwm_query = top_submatrix_from_kmer(consensus, pwm_a, extend=0)\n",
    "    pwm_b_reverse_complement = reverse_complement_matrix(pwm_b)\n",
    "    # Compare both orientations of the subset and pick the one with the best match\n",
    "    orient_a = min_kl_divergence_overlap(pwm_query, pwm_b)\n",
    "    orient_b = min_kl_divergence_overlap(pwm_query,\n",
    "                                         pwm_b_reverse_complement)\n",
    "    if orient_a[1] < orient_b[1]:\n",
    "        orient = orient_a\n",
    "        pwm_oriented = pwm_b\n",
    "    else:\n",
    "        orient = orient_b\n",
    "        pwm_oriented = pwm_b_reverse_complement\n",
    "    # Based on the best match, return the subset of b in the best matching orientation\n",
    "    start = max(0, orient[0] - extend)\n",
    "    end = min(len(pwm_oriented) - 1, orient[0] + len(consensus) + extend)\n",
    "    return pwm_oriented.iloc[start:end, :].reset_index(drop=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Analysis:\n",
    "\n",
    "The PWM files are drawn in a group of 2 figures, each containing a row of PWMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read meta data\n",
    "full_meta_data = pd.read_csv(META_DATA_FILE)\n",
    "ic_matrix_list = []\n",
    "pwm_dict = {}\n",
    "# For each row of PWM logos\n",
    "for subplot, subplot_name in ((0, \"Top_Row\"), (1, \"Bottom_Row\")):\n",
    "    meta_data = full_meta_data[full_meta_data[\"Subplot\"] == subplot].reset_index(drop=True)\n",
    "    # For each row, generate the plots specified\n",
    "    fig, ax_array = plt.subplots(2, 5)\n",
    "    fig.set_size_inches(40, 5)\n",
    "    for row in meta_data.itertuples():\n",
    "        # Read PWM files for non-UV and UV\n",
    "        pwm_nuv = read_pwm_matrix(f\"{DATA_FOLDER}/{row.Non_UV}_{row.File_Non_UV}\")\n",
    "        pwm_uv = read_pwm_matrix(f\"{DATA_FOLDER}/{row.UV}_{row.File_UV}\")\n",
    "        # Trim the non-UV matrix based on the consensus site\n",
    "        pwm_nuv = top_submatrix_from_kmer(row.Consensus, pwm_nuv)\n",
    "        # Find best match to the UV matrix\n",
    "        pwm_uv = best_match(pwm_nuv, pwm_uv, row.Consensus)\n",
    "        # List of pairs to save\n",
    "        pwm_list = []\n",
    "        # Plot the logos\n",
    "        for pwm, name, rn in zip((pwm_nuv, pwm_uv), (row.Non_UV, row.UV), (0, 1)):\n",
    "            information_matrix = logomaker.transform_matrix(pwm,\n",
    "                                                            from_type=\"probability\",\n",
    "                                                            to_type=\"information\")\n",
    "            logo = plot_ic_matrix_logo(information_matrix,\n",
    "                                       ax=ax_array[rn][row.Figure_Column],\n",
    "                                       color_scheme=LOGO_COLOR_SCHEME)\n",
    "            information_matrix[\"Experiment\"] = name\n",
    "            ic_matrix_list.append(information_matrix)\n",
    "            pwm_list.append(pwm)\n",
    "        # Add PWM list pair to the dictionary as a tuple\n",
    "        pwm_dict[row.Protein] = tuple(pwm_list)\n",
    "    plt.subplots_adjust(wspace=0.05)\n",
    "    plt.savefig(fname=f\"{OUTPUT_FOLDER_F2}/{subplot_name}.svg\",format='svg')\n",
    "    plt.savefig(fname=f\"{OUTPUT_FOLDER_SUPP}/{subplot_name}.svg\",format='svg')\n",
    "    plt.close()\n",
    "# Concatinate the raw plotted data for the logos and save\n",
    "ic_plotted_data = pd.concat(ic_matrix_list)\n",
    "ic_plotted_data = ic_plotted_data.reset_index()\n",
    "ic_plotted_data = ic_plotted_data.rename(columns={\"pos\":\"Position\"})\n",
    "ic_plotted_data[\"Experiment\"] = ic_plotted_data[\"Experiment\"].apply(lambda x: \"_\".join(x.split('_')[:2]))\n",
    "ic_plotted_data[\"Experiment\"] = ic_plotted_data[\"Experiment\"].apply(lambda x: x.replace(\"_WC\", \"_NonUV\"))\n",
    "ic_plotted_data.to_csv(f\"{OUTPUT_FOLDER_T}/Table_S3A.csv\", index=False, float_format=\"%.6f\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2C, Figure 2SA; Pyrimidine dinucleotide formation probabilities\n",
    "\n",
    "Given a sliding 2-bp window, calculate the probability of forming pyrimidine dinucleotides for PWMs of each transcription factor in non-UV and UV conditions. Then calculate the difference in the probability they form.\n",
    "\n",
    "When assembling the figures, the position with the largest difference will be colored. The color is based on a diverging color palette from -1 to 1. \n",
    "\n",
    "#### (1) Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pydi_formation_probability(pwm: pd.DataFrame,\n",
    "                               prefix: str,\n",
    "                               pydi_tuples=uac.PYDI) -> pd.DataFrame:\n",
    "    \"\"\"Return probabilities for pyrimidine dinucelotide formation.\n",
    "\n",
    "    Calculates the probability of any pyrimidine dinucleotide forming across a\n",
    "    2-position sliding window for a PWM model. Returns a dataframe with the\n",
    "    first position in the window and the probability.\n",
    "    \"\"\"\n",
    "    # Create a list of probabilities at each 2-position window\n",
    "    probabilities = []\n",
    "    for i in range(len(pwm) - 1):\n",
    "        # Sum probabilities for each pyrimidine dinucleotide\n",
    "        any_pydi_probability = 0\n",
    "        for left, right in map(tuple, pydi_tuples):\n",
    "            any_pydi_probability += pwm[left][i] * pwm[right][i + 1]\n",
    "        probabilities.append((i, any_pydi_probability))\n",
    "    # Type cast the probabilities list into a dataframe\n",
    "    df = pd.DataFrame(probabilities)\n",
    "    df = df.rename(columns={0: \"First_Position\",\n",
    "                            1: f\"{prefix}_PyDi_Probability\"})\n",
    "    return df\n",
    "\n",
    "\n",
    "def compare_pydi_formation(protein_name: str,\n",
    "                           df_nuv: pd.DataFrame,\n",
    "                           df_uv: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Calculate the difference in pydi formation for a pair of pwm models.\n",
    "\n",
    "    Given a pwm model for non-UV and UV, determines the difference in the\n",
    "    probability of pyrimidine dinucleotide formation. Return a dataframe with\n",
    "    the differences between the two models. \n",
    "    \"\"\"\n",
    "    pdfa = pydi_formation_probability(df_nuv, \"NUV\")\n",
    "    pdfb = pydi_formation_probability(df_uv, \"UV\")\n",
    "    pdf = pd.merge(pdfa, pdfb, on=\"First_Position\")\n",
    "    pdf[\"NUV_minus_UV\"] = pdf[pdf.columns[1]] - pdf[pdf.columns[2]]\n",
    "    pdf[\"Absolute_Difference\"] = abs(pdf[\"NUV_minus_UV\"])\n",
    "    pdf[\"Comparison\"] = protein_name\n",
    "    return pdf\n",
    "\n",
    "def color_from_mpl_palette(palette: str,\n",
    "                           query_value: float,\n",
    "                           value_min: float,\n",
    "                           value_max: float) -> str:\n",
    "    \"\"\"Return the hex color code from a matplotlib color palette.\n",
    "    \n",
    "    Given a query value, the value range the pallette is being used for, and \n",
    "    the name of the matplotlib palette, returns a hex color code as a string for\n",
    "    the value.\n",
    "    \"\"\"\n",
    "    # Color map object\n",
    "    color_map = matplotlib.cm.get_cmap(palette)\n",
    "    # Get normalization of values based on value range\n",
    "    normalization = matplotlib.colors.Normalize(vmin=value_min, vmax=value_max)\n",
    "    # Output includes rgb and luminence but only rgb is needed for hex code\n",
    "    rgb_colors = color_map(normalization(query_value))[:3]\n",
    "    color = matplotlib.colors.rgb2hex(rgb_colors)\n",
    "    return color\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis\n",
    "results = [compare_pydi_formation(\n",
    "    x, pwm_dict[x][0], pwm_dict[x][1]) for x in pwm_dict]\n",
    "pydi_probabilities = pd.concat(results).reset_index(drop=True)\n",
    "# Add color information\n",
    "pydi_probabilities[\"Plot_Position\"] = pydi_probabilities[\"First_Position\"] + 1\n",
    "pydi_probabilities[\"Color_F2\"] = pydi_probabilities[\"NUV_minus_UV\"].apply(lambda x: color_from_mpl_palette(\"coolwarm\", x, -1, 1))\n",
    "pydi_probabilities[\"Color_F2S\"] = pydi_probabilities[\"NUV_minus_UV\"].apply(lambda x: DECREASE_COLOR if x >= 0 else INCREASE_COLOR)\n",
    "di = {DECREASE_COLOR:\"Decrease\", INCREASE_COLOR:\"Increase\"}\n",
    "pydi_probabilities[\"Group\"] = pydi_probabilities[\"Color_F2S\"].apply(lambda x: di[x])\n",
    "# Save results\n",
    "pydi_probabilities.to_csv(f\"{OUTPUT_FOLDER_SUPP}/Table_Probability_Diff.csv\", index=False)\n",
    "# Find max absolute difference and save to table\n",
    "max_differences = pydi_probabilities[[\"Comparison\", \"Absolute_Difference\"]].groupby(by=\"Comparison\").aggregate(np.max)\n",
    "max_differences = max_differences.reset_index()\n",
    "max_differences = max_differences.rename(columns={\"Absolute_Difference\":\"Max_Absolute_Difference\"})\n",
    "max_differences.to_csv(f\"{OUTPUT_FOLDER_F2}/Table_MaxAbsDiff.csv\")\n",
    "max_differences.to_csv(f\"{OUTPUT_FOLDER_SUPP}/Table_MaxAbsDiff.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2SA: Comparison of pyrimidine dinucleotide formation probabilities\n",
    "\n",
    "In figure 2, the largest change in pyrimidine dinucelotide formation was indicated with a highlighted position. For this figure, changes across the models are shown using a barplot above the PWM logos. The following code block generates a barplot for each transcription factor and saves it in svg format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for protein in list(set(pydi_probabilities[\"Comparison\"])):\n",
    "    # Subset the pydi_probabilities dataframe by a given protein\n",
    "    plot_source_df = pydi_probabilities.query(\"Comparison == @protein\").reset_index(drop=True)\n",
    "    # Create a Bokeh figure object and source from the plot_source_df\n",
    "    source=ColumnDataSource(plot_source_df)\n",
    "    p = figure(y_range=(0, 1),\n",
    "               x_range=(0, max(plot_source_df[\"Plot_Position\"]) + 1),\n",
    "               height=200,\n",
    "               width=400,\n",
    "               toolbar_location=None, tools=\"\")\n",
    "    # Draw the vertical bars in the bar plot\n",
    "    glyph = VBar(x=\"Plot_Position\",\n",
    "                 top=\"Absolute_Difference\",\n",
    "                 bottom=0, width=0.75,\n",
    "                 fill_color=\"Color_F2S\")\n",
    "    p.add_glyph(source, glyph)\n",
    "    # Give setting for grids, plot ticks, font, outlines\n",
    "    p.xgrid.grid_line_color = None\n",
    "    p.ygrid.grid_line_color = None\n",
    "    p.xaxis.minor_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    p.xaxis.major_tick_line_color = None\n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.outline_line_width = 1\n",
    "    p.outline_line_color = 'black'\n",
    "    p.yaxis.ticker = []\n",
    "    p.xaxis.ticker = []\n",
    "    # Set output backend to svg and save\n",
    "    p.output_backend='svg'\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_SUPP}/{protein}_PyDi_Barplot.svg\", obj=p)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2SB: Comparison of increasing and decreasing positions\n",
    "\n",
    "Across all positions for the 10 PWMs in Figure 2 and 2S, are the magnitudes in change different for increases and decreases? \n",
    "\n",
    "This analysis makes a boxplot of both groups and then runs a wilxocon rank-sum test to compare the groups. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUsAAAJNCAYAAAC4BlceAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAcaElEQVR4nO3df7zlBV3n8dfb4aeEmHD9sRcm0NAe5EPJRgpi/VUZuLXUSCtomVo7YTK1tbW5u61uj9oe+ajMbcRYqlFrc2mr0aWiqEeG5I+SwQDDhJ1GiztojOgKyAQMfvaP8x27XS5zPzPM995z576ej8d93HO+53vP/TBzeM33e875fk+qCknS/j1mpQeQpNXAWEpSg7GUpAZjKUkNxlKSGoylJDUcsdIDHIyTTjqpTj311JUeQ9Jh5oYbbvhMVc0sdtuqjOWpp57K9u3bV3oMSYeZJH/3SLe5Gy5JDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpYVV+FK50ONm2bRu7du1att+3e/duAGZmFv147FHMzs6ycePGZft9YzCW0hpz//33r/QIq5KxlFbYcm9xbdmyBYDNmzcv6+9d7XzOUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWoYPZZJzktya5IdSV6/yO0vSPL5JDcOX28YeyZJOlCjnkgjyTrgMuCbgTng+iRXVdXHFqz651X1rWPOIkmPxthblmcBO6pqZ1U9AFwJXDDy75SkQ27sWM4Ct8+7PjcsW+jsJDcl+cMkXz3yTJJ0wMY+n2UWWVYLrn8E+IqqujfJS4D3AKc/7I6STcAmgPXr1x/iMSVp/8bespwDTpl3/WTgjvkrVNXdVXXvcPlq4MgkJy28o6q6oqo2VNWG5TwdviTB+LG8Hjg9yWlJjgIuAq6av0KSJyfJcPmsYaa7Rp5Lkg7IqLvhVbU3yaXANcA6YGtV3ZLkkuH2y4ELgdcm2QvsAS6qqoW76pK0okb/DJ5h1/rqBcsun3f5rcBbx55Dkh4Nj+CRpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJalh9FgmOS/JrUl2JHn9ftZ7bpKHklw49kySdKBGjWWSdcBlwPnAGcDFSc54hPXeBFwz5jySdLDG3rI8C9hRVTur6gHgSuCCRdbbDPwucOfI80jSQRk7lrPA7fOuzw3LviTJLPAdwOUjzyJJB23sWGaRZbXg+luAH6+qh/Z7R8mmJNuTbN+9e/ehmk+SWo4Y+f7ngFPmXT8ZuGPBOhuAK5MAnAS8JMneqnrP/JWq6grgCoANGzYsDK4kjWrsWF4PnJ7kNGAXcBHw8vkrVNVp+y4neQfw+wtDKUkrbdRYVtXeJJcyeZV7HbC1qm5Jcslwu89TSloVxt6ypKquBq5esGzRSFbVq8aeR5IOhkfwSFKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqeGAYpnkuLEGkaRp1oplknOSfAz4m+H6s5O8bdTJJGmKdLcsfxH4FuAugKq6CXjeWENJ0rRp74ZX1e0LFj10iGeRpKl1RHO925OcA1SSo4AfZNgll6S1oLtleQnwOmAWmAPOHK5L0prQimVVfaaqXlFVT6qqJ1bVd1XVXZ2fTXJekluT7Ejy+kVuvyDJzUluTLI9ybkH+h8hSWPrvhr+ziSPn3f9y5NsbfzcOuAy4HzgDODiJGcsWO1PgWdX1ZnAa4Bf7Y0uScunuxv+rKr6f/uuVNXngK9p/NxZwI6q2llVDwBXAhfMX6Gq7q2qGq4eBxSSNGW6sXxMki/fdyXJE+i9ODQLzH8VfW5Y9s8k+Y4kHwf+gMnWpSRNle6r4b8AfDDJ7wzXvxP4b42fyyLLHrblWFXvBt6d5HnATwHf9LA7SjYBmwDWr1/fHFuSDo3uCzy/DlwI/ANwJ7Cxqn6j8aNzwCnzrp8M3LGf33Md8LQkJy1y2xVVtaGqNszMzHTGlqRDprtlCfBx4HP7fibJ+qr6+yV+5nrg9CSnAbuAi4CXz18hyVcCf1tVleQ5wFEMRwpJ0rRoxTLJZuCNTLYsH2Kye13As/b3c1W1N8mlwDXAOmBrVd2S5JLh9suBlwKvTPIgsAd42bwXfCRpKnS3LH8IeEb3vZXzVdXVwNULll0+7/KbgDcd6P1K0nLqvhp+O/D5MQeRpGnW3bLcCVyb5A+A+/ctrKo3jzKVJE2Zbiz/fvg6aviSpDWlFcuq+kmYnCm9qr4w7kiSNH26x4af7ZnSJa1l3Rd43oJnSpe0hnmmdElq8EzpktTgmdIlqWHJLcvhBL5vqapXLMM8kjSVltyyrKqHgJlh91uS1qTuc5afBD6Q5CrgS++z9AgeSWtFN5Z3DF+PAY4fbxxJmk4ewSNJDR7BI0kNHsEjSQ0ewSNJDR7BI0kNHsEjSQ373bJM8qaq+nHghR7BI2ktW2rL8iVJjgT+43IMI0nTaqnnLP8I+AxwXJK7+aePwA1QVfW4keeTpKmw1JblT1TVCcAfVNXjqur4+d+XY0BJmgZLxfJDw/e7xx5EkqbZUrvhRyX5HuCcJBsX3lhV28YZS5Kmy1KxvAR4BfB44NsW3FaAsZS0Juw3llX1fuD9SbZX1a8t00ySNHWWep/li6rqvcDn3A2XtJYttRv+fOC9PHwXHNwNl7SGLLUb/sbh+6uXZxxJmk6dDyx7BrAJ+Kph0d8AV1TVbWMOJknTZL/vs0xyNnAtcC9wBfArTD6D59okXz/6dJI0JZbasnwDcHFVXTtv2XuSvBd4I3D+WINJ0jRZ6giepy0IJQBV9T7gqaNMJElTaKlY3rOf2/zgMklrxlK74ack+aVFlofJiYAlaU1YKpY/tp/bth/KQSRpmi31Pst3du4kyZaq2nxoRpJW3rZt29i1a9dKjzGKubk5ALZs2bLCk4xndnaWjRsfdtDho9L9wLKlfMMhuh9pKuzatYvbP/G3PPHYI1d6lEPuiIf2AnD/p/9+hScZx517Hhzlfg9VLKXDzhOPPZKXP31mpcfQAXrXbbtHud/254ZL0lp2qGKZQ3Q/kjSVDiiWSY57hJv++yGYRZKmViuWSc5J8jEmJ9EgybOTvG3f7VX1jnHGk6Tp0N2y/EXgW4C7AKrqJuB5Yw0lSdOmvRteVbcvWPTQIZ5FkqZW961Dtyc5B6gkRwE/yLBLLklrQXfL8hLgdUyOB58DzgR+YKSZJGnqdLcsn1FVr5i/IMk3AB849CNJ0vTpblkudhDp4XtgqSQtsNRH4Z4NnAPMJPmReTc9Dlg35mCSNE2W2g0/CviyYb3j5y2/G7hwrKEkadosdYq29wHvS/KOqvq7ZZpJkqZO9wWedySphQur6kWHeB5JmkrdWP7ovMvHAC8F9h76cSRpOrViWVU3LFj0gSTvG2EeSZpKrVgmecK8q48BvhZ48igTSdIU6u6G3wAUk/NW7gU+AXzvWENJ0rTp7oafNvYgkjTNlnpT+n4/Hq2qth3acSRpOi21Zflt+7mtAGMpaU1Y6k3pr16uQSRpmnU/VuKEJG9Osn34+oUkJ4w9nCRNi+5Zh7YC9wD/Zvi6G3j7WENJ0rTpvnXoaVX10nnXfzLJjSPMI0lTqbtluSfJufuuDCf+3TPOSJI0fbpblq8F3jk8Txngs8CrxhpKkqZN903pNwLPTvK44frdYw4lSdOm+2r4Dw2hvAd4c5KPJHnxuKNJ0vToPmf5mmFr8sXAE4FXAz872lSSNGW6sczw/SXA26vqpnnLJOmw143lDUn+mEksr0lyPPDF8caSpOnSfTX8e4EzgZ1VdV+SE5nsikvSmtB9NfyLSU4Fvmv4LJ73V9W7R51MkqZI99XwtwGXAB8F/hr4/iSXjTmYJE2T7m7484FnVlUBJHknk3BK0prQfYHnVmD9vOunADcf+nEkaTotdab032Nykt8TgL9J8uHhpucCHxp5NkmaGkvthv/8IssCnAtcfOjHkaTptNSZ0r/02eBJzgRezuR8lp8ALh91MkmaIkvthj8duIjJVuRdwG8BqaoXLsNskjQ1ltoN/zjw58C3VdUOgCQ/PPpUkjRllno1/KXAp4E/S/IrSb4RjwmXtAbtN5ZV9e6qehnwVcC1wA8DT0ryy56iTdJa0nqfZVV9oap+s6q+FTgZuBF4/ZiDSdI06b4p/Uuq6rNV9T+q6kVjDCRJ0+iAYylJa9HosUxyXpJbk+xI8rBd9ySvSHLz8PXBJM8eeyZJOlCjxjLJOuAy4HzgDODiJGcsWO0TwPOr6lnATwFXjDmTJB2MsbcszwJ2VNXOqnoAuBK4YP4KVfXBqvrccPUvmLyAJElTZexYzgK3z7s+Nyx7JN8L/OGoE0nSQeiez/JgLfYG9lp0xeSFTGJ57iPcvgnYBLB+/frFVpGk0Yy9ZTnH5NyX+5wM3LFwpSTPAn4VuKCq7lrsjqrqiqraUFUbZmZmRhlWkh7J2LG8Hjg9yWlJjmJyUo6r5q+QZD2wDfjuqrpt5Hkk6aCMuhteVXuTXApcA6wDtlbVLUkuGW6/HHgDcCLwtiQAe6tqw5hzSdKBGvs5S6rqauDqBcsun3f5+4DvG3sOSXo0PIJHkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKQGYylJDcZSkhqMpSQ1GEtJajCWktRgLCWpwVhKUoOxlKSGI1Z6gMPVtm3b2LVr17L9vt27dwMwMzOzbL9zdnaWjRs3Ltvvk1aSsTxM3H///Ss9gnRYM5YjWe4tri1btgCwefPmZf290lrhc5aS1GAsJanBWEpSg7GUpIbRY5nkvCS3JtmR5PWL3P5VST6U5P4kPzr2PJJ0MEZ9NTzJOuAy4JuBOeD6JFdV1cfmrfZZ4AeBbx9zFkl6NMbesjwL2FFVO6vqAeBK4IL5K1TVnVV1PfDgyLNI0kEbO5azwO3zrs8NyyRpVRk7lllkWR3UHSWbkmxPsn3foX2StFzGjuUccMq86ycDdxzMHVXVFVW1oao2LOfxz5IE48fyeuD0JKclOQq4CLhq5N8pSYfcqK+GV9XeJJcC1wDrgK1VdUuSS4bbL0/yZGA78Djgi0n+HXBGVd095mySdCBGP5FGVV0NXL1g2eXzLn+aye65JE0tzzokLWL37t38430P8q7bfDFxtbnzvgc5ZoQXgT3cUZIa3LKUFjEzM8P9D+3h5U/3nRerzbtu283RI7xjxi1LSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajKUkNRhLSWowlpLUYCwlqcFYSlKDsZSkBmMpSQ3GUpIajljpAZbLtm3b2LVr10qPMZq5uTkAtmzZssKTjGd2dpaNGzeu9Bhao9ZMLHft2sXOT85x7PFPXulRRrG3jgTgU3ftXeFJxrHnnk+v9Aha49ZMLAGOPf7JPPWs16z0GDoIOz+8daVH0Brnc5aS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGo5Y6QGWy+7du9nzhX9k54e3rvQoOgh77vkUu794zLL+zjv3PMi7btu9rL9zOXzu/r0AfPnRh+f//nfueZBTRrjfw/NPS3qUZmdnV3qE0eydmwPg6CefvMKTjOMUxvn7WzOxnJmZYe9j9vLUs16z0qPoIOz88FZmTly+h+vGjRuX7Xctty1btgCwefPmFZ5kdfE5S0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDWsmTelA+y559OH7eGO9993FwBHP/bEFZ5kHHvu+TSceHgecaLVYc3E8nA+fA1gbs+DADxlGY9yWVYnnnzY/x1quh2m/2c93OF8+Bp4CJs0Np+zlKQGYylJDcZSkhqMpSQ1GEtJahg9lknOS3Jrkh1JXr/I7UnyS8PtNyd5ztgzSdKBGjWWSdYBlwHnA2cAFyc5Y8Fq5wOnD1+bgF8ecyZJOhhjb1meBeyoqp1V9QBwJXDBgnUuAH69Jv4CeHySp4w8lyQdkLHflD4L3D7v+hzwdY11ZoFPjTvauLZt28auXbuW7ffNDR9Cte/N6cthdnb2sH+z/3LwsbI6jB3LLLKsDmIdkmxispvO+vXrH/1kh5mjjz56pUfQKuFj5eCMHcs5+Gcf4XsycMdBrENVXQFcAbBhw4aHxXTarPZ/RbV8fKysDmM/Z3k9cHqS05IcBVwEXLVgnauAVw6vin898PmqWtW74JIOP6NuWVbV3iSXAtcA64CtVXVLkkuG2y8HrgZeAuwA7gNePeZMknQwRj/rUFVdzSSI85ddPu9yAa8bew5JejQ8gkeSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1GAsJanBWEpSg7GUpAZjKUkNxlKSGoylJDUYS0lqMJaS1JDJ54WtLkl2A3+30nNMoZOAz6z0EFoVfKws7iuqamaxG1ZlLLW4JNurasNKz6Hp52PlwLkbLkkNxlKSGozl4eWKlR5Aq4aPlQPkc5aS1OCWpSQ1GMsVlOTelZ5B0y3JQ0luTHJLkpuS/EgS/79dAUes9AB6dJIcUVV7V3oOjWZPVZ0JkOSJwLuAE4A3Ppo79XFz4PwXagokeUGSa5P8TpKPJ/nNJBlue26SDw5bFR9OcnySVyX57SS/B/xxkuOSbE1yfZK/SnLB8LOnJvnzJB8Zvs4Zlj8lyXXDFstfJ/mXw/IXJ/nQsO5vJ/myFftD0cNU1Z3AJuDSTKxL8nPD3/vNSb5/37pJ/kOSjw6Pm58dll2b5GeSvA/4oSRfm+R9SW5Ick2Spwzr/dvhPm9K8rtJHjss/87h8XJTkuuGZY84w2GnqvxaoS/g3uH7C4DPAycz+QfsQ8C5wFHATuC5w3qPY7I38CpgDnjCsPxngO8aLj8euA04DngscMyw/HRg+3D53wP/ebi8DjieyREd1wHHDct/HHjDSv8ZrfWvfY+RBcs+BzyJSTh/Ylh2NLAdOA04H/gg8Njhtn2Pk2uBtw2XjxzWmRmuvwzYOlw+cd7v+mlg83D5o8DsvsfZ8H3RGVb6z22ML3fDp8eHq2oOIMmNwKlMAvqpqroeoKruHm4H+JOq+uzwsy8G/nWSHx2uHwOsB+4A3prkTOAh4OnD7dcDW5McCbynqm5M8nzgDOADw/0fxSTamj4Zvr8YeFaSC4frJzD5R/GbgLdX1X0A8x4nAL81fH8G8EzgT4a/73XAp4bbnpnkp5n8w/tlwDXD8g8A70jyv4FtS8zwiUf/nzldjOX0uH/e5YeY/N0EeKT3dn1h3uUAL62qW+evkOS/Av8APJvJFus/AlTVdUmeB/wr4DeS/ByTrZU/qaqLH/1/isaS5KlMHh93Mvl731xV1yxY5zyWftwEuKWqzl5knXcA315VNyV5FZM9H6rqkiRfx+Rxc+Pwj/CiMxyOfM5yun0c+BdJngswPF+52D9w1wCb5z3P+TXD8hOYbJl+EfhuJlsPJPkK4M6q+hXg14DnAH8BfEOSrxzWeWySp6OpkWQGuBx4a032e68BXjvsIZDk6UmOA/4YeM285xqfsMjd3QrMJDl7WOfIJF893HY88Knhfl8x7/c/rar+sqrewOQkHKfsZ4bDjluWU6yqHkjyMmBLkmOBPUx2sRb6KeAtwM1DMD8JfCvwNuB3k3wn8Gf801bFC4AfS/IgcC/wyqraPWxF/K8kRw/r/QST5z+1co4dnpY5EtgL/Abw5uG2X2XydM1Hhr/33Uy2CP9o2OrbnuQB4GrgP82/0+GxdSHwS0lOYNKCtwC3AP8F+EsmZ/b6KJN4AvxcktOZbE3+KXATcPNiMxzKP4Bp4RE8ktTgbrgkNRhLSWowlpLUYCwlqcFYSlKDsdSqkeRJSd6VZOdwPPOHknzHSs+ltcFYalUY3sP3HuC6qnpqVX0tcBGT4+nnr+d7hzUK32epVSHJNzI5scfzF7ntVUwOwTuGyQlELgS2Ak8F7gM2VdXNw+Gf91bVzw8/99dM3rwP8EdM3oj9NUzeiP/KfcdWS+CWpVaPrwY+sp/bzwa+p6peBPwk8FdV9SwmR678euP+nwFcMfzM3cAPPMp5dZgxllqVklw2nFfx+mHR/LMwncvksECq6r3AicMhfftze1V9YLj8P4f7kL7EWGq1uIXJCT8AqKrXAd8IzAyLFp6FaaFicmz1/Mf8MQtuX7i+9CXGUqvFe4Fjkrx23rLHPsK61zGcLSfJC4DPDOcC/SRDcJM8h8mJcvdZv+8MPMDFwPsP1eA6PPgCj1aN4WMPfhH4OiZnt/kCk1OWHQtsqKpLh/WeALydSQznv8BzLPB/gCcyOQHyuUzOKg6TM/NcB5wD/F/gu32BR/MZS615SU4Ffr+qnrnSs2h6uRsuSQ1uWUpSg1uWktRgLCWpwVhKUoOxlKQGYylJDcZSkhr+P1vegM0J36WnAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 360x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create and save boxplot of the 2 groups\n",
    "rcParams['figure.figsize'] = 5,10\n",
    "sns.boxplot(x=\"Group\",\n",
    "            y=\"Absolute_Difference\",\n",
    "            data=pydi_probabilities,\n",
    "            order=[\"Increase\", \"Decrease\"],\n",
    "            palette=['#8db0fe', '#f4987a'],\n",
    "            showfliers=False)\n",
    "plt.savefig(fname=f\"{OUTPUT_FOLDER_SUPP}/Absolute_Difference_Boxplot.svg\",format='svg')\n",
    "# Run rank sum test\n",
    "decrease = list(pydi_probabilities[pydi_probabilities[\"Color_F2S\"] == \"#f4987a\"][\"Absolute_Difference\"])\n",
    "increase = list(pydi_probabilities[pydi_probabilities[\"Color_F2S\"] == \"#8db0fe\"][\"Absolute_Difference\"])\n",
    "comparison_ranksum = stats.ranksums(decrease, increase)\n",
    "# Save result to a text file\n",
    "comparison_out = open(f\"{OUTPUT_FOLDER_SUPP}/Absolute_Difference_Boxplot_Statistics.txt\", 'w')\n",
    "comparison_out.write(f\"Rank Sum Statistics: {comparison_ranksum.statistic}\\nP-Value: {comparison_ranksum.pvalue}\")\n",
    "comparison_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Color bar for use in Figure 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgAAAABACAYAAABsv8+/AAAAF3RFWHRUaXRsZQBjb29sd2FybSBjb2xvcm1hcPHW9i4AAAAddEVYdERlc2NyaXB0aW9uAGNvb2x3YXJtIGNvbG9ybWFw2KgRBwAAADB0RVh0QXV0aG9yAE1hdHBsb3RsaWIgdjMuNS4xLCBodHRwczovL21hdHBsb3RsaWIub3Jn97fOzwAAADJ0RVh0U29mdHdhcmUATWF0cGxvdGxpYiB2My41LjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmfZEVHoAAACIUlEQVR4nO3WO3LbQBBF0QaW5TV40d6XTIwTCCo2MQQYOHrnJKr5NRip7vLr959RVbUsS1VVLetaVVXr+rzu50s7X2fnt+cu+/n5/nF+8Z315buz9/3310fzXvfrfP73nOn8mnzv+f7xfva9/bzPu9w/vnM+b36v3t+72J/fGxfvx739mt1v82tyXv13jPO5x7175y/fm53f3F/7fm333vf7o+2P7cP18371e+P9nOrvR/995/e+59ZkPZ17rN/fO+Zufe75vdr67znfv3o3tnv3Lt/t69HW3+fH/Zc579+/rLc2r82fzevf7/ePv+Pq3uTdf9rfHu/v/5y3e/v+1uc/tud3j8m84/0+72tf/x2frb+e/18BAEEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEOgf6GZUalYlbg8AAAAASUVORK5CYII=\n",
      "text/html": [
       "<div style=\"vertical-align: middle;\"><strong>coolwarm</strong> </div><div class=\"cmap\"><img alt=\"coolwarm colormap\" title=\"coolwarm\" style=\"border: 1px solid #555;\" src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAgAAAABACAYAAABsv8+/AAAAF3RFWHRUaXRsZQBjb29sd2FybSBjb2xvcm1hcPHW9i4AAAAddEVYdERlc2NyaXB0aW9uAGNvb2x3YXJtIGNvbG9ybWFw2KgRBwAAADB0RVh0QXV0aG9yAE1hdHBsb3RsaWIgdjMuNS4xLCBodHRwczovL21hdHBsb3RsaWIub3Jn97fOzwAAADJ0RVh0U29mdHdhcmUATWF0cGxvdGxpYiB2My41LjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmfZEVHoAAACIUlEQVR4nO3WO3LbQBBF0QaW5TV40d6XTIwTCCo2MQQYOHrnJKr5NRip7vLr959RVbUsS1VVLetaVVXr+rzu50s7X2fnt+cu+/n5/nF+8Z315buz9/3310fzXvfrfP73nOn8mnzv+f7xfva9/bzPu9w/vnM+b36v3t+72J/fGxfvx739mt1v82tyXv13jPO5x7175y/fm53f3F/7fm333vf7o+2P7cP18371e+P9nOrvR/995/e+59ZkPZ17rN/fO+Zufe75vdr67znfv3o3tnv3Lt/t69HW3+fH/Zc579+/rLc2r82fzevf7/ePv+Pq3uTdf9rfHu/v/5y3e/v+1uc/tud3j8m84/0+72tf/x2frb+e/18BAEEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEEgAAEAgAQAAgQQAAAQSAAAQSAAAQCABAACBBAAABBIAABBIAABAIAEAAIEEAAAEEgAAEOgf6GZUalYlbg8AAAAASUVORK5CYII=\"></div><div style=\"vertical-align: middle; max-width: 514px; display: flex; justify-content: space-between;\"><div style=\"float: left;\"><div title=\"#3b4cc0ff\" style=\"display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #3b4cc0ff;\"></div> under</div><div style=\"margin: 0 auto; display: inline-block;\">bad <div title=\"#00000000\" style=\"display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #00000000;\"></div></div><div style=\"float: right;\">over <div title=\"#b40426ff\" style=\"display: inline-block; width: 1em; height: 1em; margin: 0; vertical-align: middle; border: 1px solid #555; background-color: #b40426ff;\"></div></div>"
      ],
      "text/plain": [
       "<matplotlib.colors.LinearSegmentedColormap at 0x7fd8765aa070>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matplotlib.cm.get_cmap(\"coolwarm\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
