{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4; Supplementary Figure 5; Supplementary Table 5\n",
    "\n",
    "Analysis comparing non-UV and UV conditions to find that the 5U01 sequence seems to increase in binding for CREB1 and EGR1 after UV irradiation. \n",
    "\n",
    "### Files\n",
    "The following files are used to generate the data:\n",
    "\n",
    "1. Initial observation:\n",
    "    1. CREB1_WC_ID0_alldata.txt\n",
    "    2. CREB1_UV_ID1_alldata.txt\n",
    "    3. EGR1_WC_ID2_alldata.txt\n",
    "    4. EGR1_UV_ID3_alldata.txt\n",
    "    1. CREB1_WC_ID0_7mers_1111111.txt\n",
    "    2. CREB1_UV_ID1_7mers_1111111.txt\n",
    "    3. EGR1_WC_ID2_7mers_1111111.txt\n",
    "    4. EGR1_UV_ID3_7mers_1111111.txt\n",
    "2. Follow-up experiments:\n",
    "    1. CREB1_WC_ID38_alldata.txt\n",
    "    2. CREB1_UV_ID39_alldata.txt\n",
    "    3. EGR1_WC_ID40_alldata.txt\n",
    "    4. EGR1_UV_ID41_alldata.txt\n",
    "3. Meta data file:\n",
    "    1. Signal_Increase_UV_5U01_Scatterplots.csv\n",
    "\n",
    "\n",
    "### 3rd Party Packages\n",
    "\n",
    "1. Bokeh - Plotting (Scatterplots)\n",
    "2. Matplotlib - Plotting (Boxplot)\n",
    "3. Numpy - ln function from np.log\n",
    "4. Pandas - Dataframe usage\n",
    "5. Seaborn - Plotting (Boxplots)\n",
    "6. Scipy - Linear regression\n",
    "7. Statsmodels: Multiple test correction\n",
    "\n",
    "### UV Bind Analysis Core Imports\n",
    "\n",
    "- uac.ols: Train OLS regression on sequences that cannot form pyrimidine dimers.\n",
    "- uac.scale_uv_on_non_uv: Scale UV probes on Non-UV by making the slope of a linear regression for sequences that cannot form pyrimidine dimers equal to 1. \n",
    "- uac.read_kmer_file_pair: Given a pair of k-mer files, reads them as dataframes and merges them into one.\n",
    "\n",
    "** Additional details can be found in the uvbind_analysis_core.py script.\n",
    "\n",
    "### Abbreviations:\n",
    "\n",
    "- pydi: Pyrimidine Dinucleotide\n",
    "- df: DataFrame\n",
    "- dn: Dinucleotide\n",
    "- idx: index\n",
    "\n",
    "#### Imports and Global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "from collections import namedtuple\n",
    "import os\n",
    "import re\n",
    "\n",
    "\n",
    "from bokeh.io import export_svg, export_png\n",
    "from bokeh.plotting import figure, show, ColumnDataSource\n",
    "from bokeh.models import CDSView, GroupFilter, Span\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import uvbind_analysis_core as uac\n",
    "\n",
    "ALLDATA_FOLDER = \"../../Data/AllData_Files\"\n",
    "KMER_FOLDER = \"../../Data/Kmer_Files\"\n",
    "SCATTERPLOT_META_DATA = \"Meta_Data/Signal_Increase_UV_5U01_Scatterplots.csv\"\n",
    "FIGS5_META_DATA = \"Meta_Data/Meta_Data_S5B.csv\"\n",
    "DN_DESIGN_META_DATA = \"Meta_Data/Array_86684_DN_Design.csv\"\n",
    "\n",
    "OUTPUT_FOLDER_MAIN = \"../Figure_4\"\n",
    "OUTPUT_FOLDER_SUPP = \"../Figure_S5\"\n",
    "OUTPUT_FOLDER_TABLE = \"../Table_S5\"\n",
    "\n",
    "\n",
    "# Figure 6S Parameters\n",
    "PALETTE = (\"#2b8cbe\", \"#a6bddb\", \"black\", \"black\") \n",
    "BOXPLOT_PALETTE = (\"lightgrey\", \"grey\") # (Non_UV, UV)\n",
    "# Do we still need these global vars? - Used in FS6g, replace with updated meta?\n",
    "CREB1_KMERS = (\"CREB1_WC_ID0_7of9mers_1111111.txt\",\n",
    "               \"CREB1_UV_ID1_7of9mers_1111111.txt\")\n",
    "EGR1_KMERS = (\"EGR1_WC_ID2_7of9mers_1111111.txt\",\n",
    "               \"EGR1_UV_ID3_7of9mers_1111111.txt\")\n",
    "# Panel A\n",
    "PANEL_A_FILES = ((\"CREB1\",\"Non_UV\" ,\"CREB1_WC_ID0_alldata.txt\"),\n",
    "              (\"CREB1\", \"UV\", \"CREB1_UV_ID1_alldata.txt\"),\n",
    "              (\"EGR1\", \"Non_UV\", \"EGR1_WC_ID2_alldata.txt\"),\n",
    "              (\"EGR1\", \"UV\", \"EGR1_UV_ID3_alldata.txt\"))\n",
    "\n",
    "# Updated meta data\n",
    "KMER_FILES = ((\"CREB1\",\"Non_UV\" ,\"CREB1_WC_ID0_7of9mers_1111111.txt\"),\n",
    "              (\"CREB1\", \"UV\", \"CREB1_UV_ID1_7of9mers_1111111.txt\"),\n",
    "              (\"EGR1\", \"Non_UV\", \"EGR1_WC_ID2_7of9mers_1111111.txt\"),\n",
    "              (\"EGR1\", \"UV\", \"EGR1_UV_ID3_7of9mers_1111111.txt\"))\n",
    "KMER_PROFILE_META_DATA = \"Meta_Data/Meta_Data_F6S_Profile_Plots.csv\"\n",
    "\n",
    "# Figure 4 parameters\n",
    "CREB1_HIGHLIGHT = (\"Clst_CREB1_C1314_TD25_O2\",\n",
    "                   \"Clst_CREB1_C604_TD25\",\n",
    "                   \"Clst_CREB1_C1039_TD25\")\n",
    "EGR1_HIGHLIGHT = (\"Clst_EGR1_C394_TD25\",\n",
    "                  \"Clst_EGR1_C537_TD25\",\n",
    "                  \"Clst_EGR1_C390_TD25\")\n",
    "KDE_BOUNDS = {\"CREB1\": ((5.8, 13),(0, 1.4)), # x_bounds, y_bounds\n",
    "              \"EGR1\": ((5.8, 13),(0, 4))}\n",
    "\n",
    "PDB_VAR_LIB = {\"ELK1\",\n",
    "              \"ETS1_monomer\",\n",
    "              \"ETS1_dimer\", \n",
    "              \"Creb1\", \n",
    "              \"CTCF\", \n",
    "              \"Myc+Mad\",\n",
    "              \"p53_1\", \n",
    "              \"RelA_dimer\", \n",
    "              \"Runx1\", \n",
    "              \"Stat3\",\n",
    "              \"TBP\",\n",
    "              \"TBP_Ariel_A\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure output folders exist\n",
    "for i in (OUTPUT_FOLDER_MAIN, OUTPUT_FOLDER_SUPP, OUTPUT_FOLDER_TABLE):\n",
    "    os.makedirs(i, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table S5A: Scatterplot of non-Universal probes\n",
    "\n",
    "#### (1) Read and display meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Panel</th>\n",
       "      <th>Protein</th>\n",
       "      <th>File_Non_UV</th>\n",
       "      <th>File_UV</th>\n",
       "      <th>Range_Start_X</th>\n",
       "      <th>Range_End_X</th>\n",
       "      <th>Range_Start_Y</th>\n",
       "      <th>Range_End_Y</th>\n",
       "      <th>Ticker_A_X</th>\n",
       "      <th>Ticker_B_X</th>\n",
       "      <th>Ticker_C_X</th>\n",
       "      <th>Ticker_A_Y</th>\n",
       "      <th>Ticker_B_Y</th>\n",
       "      <th>Ticker_C_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1_WC_ID0_alldata.txt</td>\n",
       "      <td>CREB1_UV_ID1_alldata.txt</td>\n",
       "      <td>50</td>\n",
       "      <td>75000</td>\n",
       "      <td>200</td>\n",
       "      <td>75000</td>\n",
       "      <td>50</td>\n",
       "      <td>2500</td>\n",
       "      <td>75000</td>\n",
       "      <td>200</td>\n",
       "      <td>2500</td>\n",
       "      <td>75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1_WC_ID2_alldata.txt</td>\n",
       "      <td>EGR1_UV_ID3_alldata.txt</td>\n",
       "      <td>200</td>\n",
       "      <td>25000</td>\n",
       "      <td>1000</td>\n",
       "      <td>15000</td>\n",
       "      <td>250</td>\n",
       "      <td>2500</td>\n",
       "      <td>25000</td>\n",
       "      <td>1500</td>\n",
       "      <td>5000</td>\n",
       "      <td>15000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Panel Protein               File_Non_UV                   File_UV  \\\n",
       "0     A   CREB1  CREB1_WC_ID0_alldata.txt  CREB1_UV_ID1_alldata.txt   \n",
       "1     A    EGR1   EGR1_WC_ID2_alldata.txt   EGR1_UV_ID3_alldata.txt   \n",
       "\n",
       "   Range_Start_X  Range_End_X  Range_Start_Y  Range_End_Y  Ticker_A_X  \\\n",
       "0             50        75000            200        75000          50   \n",
       "1            200        25000           1000        15000         250   \n",
       "\n",
       "   Ticker_B_X  Ticker_C_X  Ticker_A_Y  Ticker_B_Y  Ticker_C_Y  \n",
       "0        2500       75000         200        2500       75000  \n",
       "1        2500       25000        1500        5000       15000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = pd.read_csv(SCATTERPLOT_META_DATA)\n",
    "meta_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "PRIMER = uac.PRIMER\n",
    "\n",
    "def return_group(x: str):\n",
    "    \"\"\"Return a substring representing the group a label belongs to.\"\"\"\n",
    "    # Determine how many subgroups deliminated by '_' to return\n",
    "    n = 3\n",
    "    if x.startswith(\"MITF_Ex\"):\n",
    "        n = 4\n",
    "    elif x.startswith(\"random_damageable\"):\n",
    "        n = 2\n",
    "    # Return the subgroup\n",
    "    return '_'.join(x.split('_')[:n])\n",
    "\n",
    "\n",
    "def filter_alldata_dataframe(dataframe: pd.DataFrame,\n",
    "                             primer: str = PRIMER) -> pd.DataFrame:\n",
    "    \"\"\"Remove flagged sequence and non-double stranded sequences.\n",
    "\n",
    "    Given a dataframe with column names in the alldata format, returns\n",
    "    a dataframe with all flagged rows, rows with no sequence, and\n",
    "    rows with a sequence without a primer removed.\n",
    "    \"\"\"\n",
    "    alexa = dataframe.columns[7]\n",
    "    # Filter out flags\n",
    "    dataframe = dataframe[(dataframe[\"Cy3Flags\"].isnull()) &\n",
    "                          (dataframe[f\"{alexa}Flags\"].isnull())]\n",
    "    # Remove rows with no sequence\n",
    "    dataframe = dataframe[dataframe[\"Sequence\"].notnull()]\n",
    "    # Remove rows without primer\n",
    "    dataframe = dataframe[dataframe[\"Sequence\"].str.endswith(primer)]\n",
    "    dataframe = dataframe.reset_index(drop=True)\n",
    "    return dataframe\n",
    "\n",
    "\n",
    "def replicate_values(dataframe: pd.DataFrame,\n",
    "                     column: str,\n",
    "                     min_count: int = 2) -> pd.DataFrame:\n",
    "    \"\"\"Return a dataframe of values with a minimum number of replicates.\n",
    "\n",
    "    Given a dataframe and column, finds the number of replicates of that\n",
    "    column and returns a dataframe with a minimum amount of replicates.\n",
    "    \"\"\"\n",
    "    # Calculate value counts and return as a dataframe\n",
    "    count = pd.DataFrame(dataframe[column].value_counts()).reset_index()\n",
    "    # Rename the columns\n",
    "    count = count.rename(columns={column: \"Count\"})\n",
    "    count = count.rename(columns={\"index\": column})\n",
    "    # Query for items that have a minimum value count\n",
    "    count = count.query(\"Count >= @min_count\")\n",
    "    return count\n",
    "\n",
    "\n",
    "def process_alldata_file(file_path: str) -> pd.DataFrame:\n",
    "    \"\"\"Given an alldata file, return a dataframe of custom sequences.\n",
    "\n",
    "    Takes an alldata file and does the following pipeline:\n",
    "    1. Read as a dataframe\n",
    "    2. Remove universal sequences\n",
    "    3. Remove trailing replicate labels so replicate data can be queried\n",
    "    4. Subset the column into the sequence name, sequence, and signal values\n",
    "\n",
    "    Returns a dataframe with a Name, Sequence, and Signal columns.\n",
    "    \"\"\"\n",
    "    # Read dataframe\n",
    "    dataframe = pd.read_csv(file_path, sep='\\t')\n",
    "    alexa = dataframe.columns[7]  # Name is the specific alexa wavelength for the experiment\n",
    "    dataframe = filter_alldata_dataframe(dataframe)\n",
    "    # Remove universal UV-Bind Probes\n",
    "    dataframe = dataframe[~dataframe[\"Name\"].str.startswith(\"All\")]\n",
    "    dataframe = dataframe.reset_index(drop=True)\n",
    "    # Remove trailing replicate label for replicate counting\n",
    "    dataframe[\"Name_Group\"] = dataframe[\"Name\"]\\\n",
    "        .apply(lambda x: '_'.join(x.split('_')[:-1]))\n",
    "    # Filter for sequences with 2+ replicates\n",
    "    replicate_counts = replicate_values(dataframe, \"Name_Group\")\n",
    "    name_set = set(replicate_counts[\"Name_Group\"])\n",
    "    dataframe = dataframe[dataframe[\"Name_Group\"]\n",
    "                          .isin(name_set)].reset_index(drop=True)\n",
    "    # Subset by name, sequence and signal columns\n",
    "    dataframe = dataframe[[\"Name\", \"Sequence\", alexa]]\n",
    "    dataframe = dataframe.rename(columns={alexa: \"Signal\"})\n",
    "    return dataframe\n",
    "\n",
    "\n",
    "def filter_rep(x):\n",
    "    \"\"\"Remove the _r suffix from a label.\"\"\"\n",
    "    if re.search(pattern=\"_r\\d*$\", string=x):\n",
    "        return '_'.join(x.split('_')[:-1])\n",
    "    else:\n",
    "        return x\n",
    "    \n",
    "def decompress_tbp_labels(df):\n",
    "    dup = df[df[\"Name\"].apply(lambda x: \"Ariel_A\" in x and \"Ariel_G\" in x)].copy()\n",
    "    dup = dup.reset_index(drop=True)\n",
    "    dup[\"Group\"] = \"TBP_Ariel_G\"\n",
    "    df = pd.concat([df, dup]).reset_index(drop=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "def show_scatterplot(plot_df: pd.DataFrame,\n",
    "                     x: str,\n",
    "                     y: str,\n",
    "                     color_on: str,\n",
    "                     title: str,\n",
    "                     color_map: str = \"viridis\"):\n",
    "    \"\"\"Show a scatterplot of 2 columns in a dataframe.\"\"\"\n",
    "    figure(figsize=(8, 8), dpi=90)\n",
    "    plt.scatter(plot_df[x],\n",
    "               plot_df[y],\n",
    "               c=plot_df[color_on],\n",
    "               cmap=color_map)\n",
    "    plt.title(title)\n",
    "    plt.colorbar()\n",
    "    plt.axline((1, 1), slope=1, color='black')\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def show_hexbin(plot_df: pd.DataFrame,\n",
    "                x: str,\n",
    "                y: str,\n",
    "                color_on: str,\n",
    "                title: str,\n",
    "                color_map: str = \"viridis\"):\n",
    "    \"\"\"Show a hexbin of 2 columns in a dataframe aggregating by median.\"\"\"\n",
    "    figure(figsize=(8, 8), dpi=80)\n",
    "    plt.hexbin(plot_df[x],\n",
    "               plot_df[y],\n",
    "               C=plot_df[color_on],\n",
    "               reduce_C_function=np.median,\n",
    "               gridsize=50,\n",
    "               cmap=color_map)\n",
    "    plt.title(title)\n",
    "    plt.axline((1, 1), slope=1, color='black')\n",
    "    plt.colorbar()\n",
    "    plt.show()\n",
    "    \n",
    "def scatterplot_of_groups(plot_df,\n",
    "                          plot_range_x,\n",
    "                          plot_range_y,\n",
    "                          circle_size,\n",
    "                          palette,\n",
    "                          ticker_x,\n",
    "                          ticker_y,\n",
    "                          groups,\n",
    "                          x,\n",
    "                          y):\n",
    "    \"\"\"\"\"\"\n",
    "    # Create bokeh plot object\n",
    "    p = figure(width=800, height=800, x_range=plot_range_x, y_range=plot_range_y)\n",
    "    source = ColumnDataSource(plot_df)\n",
    "    # For each group and their corresponding palette color, draw datapoints\n",
    "    for i, color in zip(groups, palette):\n",
    "        view = CDSView(source=source,\n",
    "                      filters=[GroupFilter(column_name=\"Group\", group=i)])\n",
    "        p.circle(x,\n",
    "                 y,\n",
    "                 source=source,\n",
    "                 view=view,\n",
    "                 size=circle_size,\n",
    "                 color=color)\n",
    "    p.xaxis.ticker = ticker_x\n",
    "    p.yaxis.ticker = ticker_y\n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.xgrid.grid_line_color = None\n",
    "    p.ygrid.grid_line_color = None\n",
    "    p.xaxis.minor_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    p.toolbar_location = None\n",
    "    return p\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CREB1_Non_UV</th>\n",
       "      <th>CREB1_UV</th>\n",
       "      <th>EGR1_UV</th>\n",
       "      <th>EGR1_Non_UV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5U01</td>\n",
       "      <td>ACGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.757891</td>\n",
       "      <td>7.783016</td>\n",
       "      <td>9.059808</td>\n",
       "      <td>7.613325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5U01</td>\n",
       "      <td>CCGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.795791</td>\n",
       "      <td>7.829630</td>\n",
       "      <td>9.110741</td>\n",
       "      <td>7.486613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GAGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.844187</td>\n",
       "      <td>7.780512</td>\n",
       "      <td>9.090486</td>\n",
       "      <td>7.577634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GCAGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.757891</td>\n",
       "      <td>7.674385</td>\n",
       "      <td>9.049232</td>\n",
       "      <td>7.464223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GCCGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.653960</td>\n",
       "      <td>7.740664</td>\n",
       "      <td>9.214880</td>\n",
       "      <td>7.744570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACCTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>7.274826</td>\n",
       "      <td>8.461786</td>\n",
       "      <td>7.610853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACGTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.553933</td>\n",
       "      <td>7.393571</td>\n",
       "      <td>8.194920</td>\n",
       "      <td>7.307538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACTTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.540308</td>\n",
       "      <td>7.485211</td>\n",
       "      <td>8.407155</td>\n",
       "      <td>7.605890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCAGATTTCCCGTAAATCATACA</td>\n",
       "      <td>5.930918</td>\n",
       "      <td>7.298783</td>\n",
       "      <td>8.305608</td>\n",
       "      <td>7.038784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCATATTTCCCGTAAATCATACA</td>\n",
       "      <td>6.336826</td>\n",
       "      <td>7.630947</td>\n",
       "      <td>8.547043</td>\n",
       "      <td>7.409742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Group                   Sequence  CREB1_Non_UV  CREB1_UV   EGR1_UV  \\\n",
       "0     5U01  ACGGAAATTCCCGGGAATTTCCGCG      4.757891  7.783016  9.059808   \n",
       "1     5U01  CCGGAAATTCCCGGGAATTTCCGCG      4.795791  7.829630  9.110741   \n",
       "2     5U01  GAGGAAATTCCCGGGAATTTCCGCG      4.844187  7.780512  9.090486   \n",
       "3     5U01  GCAGAAATTCCCGGGAATTTCCGCG      4.757891  7.674385  9.049232   \n",
       "4     5U01  GCCGAAATTCCCGGGAATTTCCGCG      4.653960  7.740664  9.214880   \n",
       "..     ...                        ...           ...       ...       ...   \n",
       "556  Other  GCGCACCTTTCCCGTAAATCATACA      6.135565  7.274826  8.461786   \n",
       "557  Other  GCGCACGTTTCCCGTAAATCATACA      6.553933  7.393571  8.194920   \n",
       "558  Other  GCGCACTTTTCCCGTAAATCATACA      6.540308  7.485211  8.407155   \n",
       "559  Other  GCGCAGATTTCCCGTAAATCATACA      5.930918  7.298783  8.305608   \n",
       "560  Other  GCGCATATTTCCCGTAAATCATACA      6.336826  7.630947  8.547043   \n",
       "\n",
       "     EGR1_Non_UV  \n",
       "0       7.613325  \n",
       "1       7.486613  \n",
       "2       7.577634  \n",
       "3       7.464223  \n",
       "4       7.744570  \n",
       "..           ...  \n",
       "556     7.610853  \n",
       "557     7.307538  \n",
       "558     7.605890  \n",
       "559     7.038784  \n",
       "560     7.409742  \n",
       "\n",
       "[561 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in each file and concatinate into a single df\n",
    "panel_a_data = []\n",
    "for protein, condition, file in PANEL_A_FILES:\n",
    "    df = process_alldata_file(f\"{ALLDATA_FOLDER}/{file}\")\n",
    "    df[\"Group\"] = df[\"Name\"].apply(lambda x: x.split('-')[0])\n",
    "    df[\"Group\"] = df[\"Group\"].apply(lambda x: filter_rep(x))\n",
    "    df = decompress_tbp_labels(df)\n",
    "    df[\"Name\"] = df[\"Name\"].apply(lambda x: \"_\".join(x.split(\"_\")[:-1]))\n",
    "    df[\"Sequence\"] = df[\"Sequence\"].apply(lambda x: x[:25])\n",
    "    df = df.groupby(by=[\"Group\",\"Name\", \"Sequence\"]).aggregate(np.median).reset_index()\n",
    "    df[\"Protein\"] = protein\n",
    "    df[\"Condition\"] = condition\n",
    "    panel_a_data.append(df)\n",
    "panel_a_data = pd.concat(panel_a_data).reset_index(drop=True)\n",
    "panel_a_data = panel_a_data[panel_a_data[\"Group\"].isin(PDB_VAR_LIB)]\n",
    "del panel_a_data[\"Name\"]\n",
    "# Output Table S5A\n",
    "panel_a_data[\"Signal\"] = panel_a_data[\"Signal\"].apply(lambda x: np.log(x))\n",
    "panel_a_data[\"Group\"] = panel_a_data[\"Group\"].apply(lambda x: \"5U01\" if x == \"RelA_dimer\" else \"Other\")\n",
    "panel_a_data = panel_a_data.sort_values(by=\"Group\").reset_index(drop=True)\n",
    "#panel_a_data.to_csv(f\"{OUTPUT_FOLDER_TABLE}/Table_S5A.csv\", index=False)\n",
    "panel_a_data = panel_a_data.pivot(index=[\"Group\",\"Sequence\"],\n",
    "                   values=\"Signal\",\n",
    "                   columns=[\"Protein\", \"Condition\"]).reset_index()\n",
    "panel_a_data.columns = [\"_\".join(i) for i in panel_a_data.columns]\n",
    "panel_a_data.columns = list(map(lambda x: x.rstrip('_'), panel_a_data.columns))\n",
    "panel_a_data.to_csv(f\"{OUTPUT_FOLDER_TABLE}/Table_S5A.csv\", index=False)\n",
    "panel_a_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S5A\n",
    "\n",
    "Scatterplot of the non-Universal probes in supplementary figure 5 for CREB1 and EGR1. The sequences corresponding to a single-nucleotide variation design of the DNA sequence in PDB ID 5U01 are highlighted in a different color. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (1) Read and display meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Panel</th>\n",
       "      <th>Protein</th>\n",
       "      <th>File_Non_UV</th>\n",
       "      <th>File_UV</th>\n",
       "      <th>Range_Start_X</th>\n",
       "      <th>Range_End_X</th>\n",
       "      <th>Range_Start_Y</th>\n",
       "      <th>Range_End_Y</th>\n",
       "      <th>Ticker_A_X</th>\n",
       "      <th>Ticker_B_X</th>\n",
       "      <th>Ticker_C_X</th>\n",
       "      <th>Ticker_A_Y</th>\n",
       "      <th>Ticker_B_Y</th>\n",
       "      <th>Ticker_C_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1_WC_ID0_alldata.txt</td>\n",
       "      <td>CREB1_UV_ID1_alldata.txt</td>\n",
       "      <td>50</td>\n",
       "      <td>75000</td>\n",
       "      <td>200</td>\n",
       "      <td>75000</td>\n",
       "      <td>50</td>\n",
       "      <td>2500</td>\n",
       "      <td>75000</td>\n",
       "      <td>200</td>\n",
       "      <td>2500</td>\n",
       "      <td>75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1_WC_ID2_alldata.txt</td>\n",
       "      <td>EGR1_UV_ID3_alldata.txt</td>\n",
       "      <td>200</td>\n",
       "      <td>25000</td>\n",
       "      <td>1000</td>\n",
       "      <td>15000</td>\n",
       "      <td>250</td>\n",
       "      <td>2500</td>\n",
       "      <td>25000</td>\n",
       "      <td>1500</td>\n",
       "      <td>5000</td>\n",
       "      <td>15000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Panel Protein               File_Non_UV                   File_UV  \\\n",
       "0     A   CREB1  CREB1_WC_ID0_alldata.txt  CREB1_UV_ID1_alldata.txt   \n",
       "1     A    EGR1   EGR1_WC_ID2_alldata.txt   EGR1_UV_ID3_alldata.txt   \n",
       "\n",
       "   Range_Start_X  Range_End_X  Range_Start_Y  Range_End_Y  Ticker_A_X  \\\n",
       "0             50        75000            200        75000          50   \n",
       "1            200        25000           1000        15000         250   \n",
       "\n",
       "   Ticker_B_X  Ticker_C_X  Ticker_A_Y  Ticker_B_Y  Ticker_C_Y  \n",
       "0        2500       75000         200        2500       75000  \n",
       "1        2500       25000        1500        5000       15000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_data = pd.read_csv(SCATTERPLOT_META_DATA)\n",
    "meta_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Create Scatterplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_a_data[\"Sequence_Group\"] = panel_a_data[\"Group\"]\n",
    "\n",
    "panel_a_meta = meta_data[meta_data[\"Panel\"] == \"A\"].reset_index(drop=True)\n",
    "for row in panel_a_meta.itertuples():\n",
    "    # Transform plot range and ticker settings to ln scale\n",
    "    ln_plot_range_x = list(map(np.log, (row.Range_Start_X, row.Range_End_X)))\n",
    "    ln_plot_range_y = list(map(np.log, (row.Range_Start_Y, row.Range_End_Y)))\n",
    "    ln_plot_ticker_x = list(map(np.log,\n",
    "                              (row.Ticker_A_X, row.Ticker_B_X, row.Ticker_C_X)))\n",
    "    ln_plot_ticker_y = list(map(np.log,\n",
    "                              (row.Ticker_A_Y, row.Ticker_B_Y, row.Ticker_C_Y)))\n",
    "    # Plot\n",
    "    plot = scatterplot_of_groups(panel_a_data,\n",
    "                                  ln_plot_range_x,\n",
    "                                 ln_plot_range_y,\n",
    "                                  10,\n",
    "                                  (PALETTE[2], PALETTE[0]),\n",
    "                                  ln_plot_ticker_x,\n",
    "                                 ln_plot_ticker_y,\n",
    "                                 (\"Other\", \"5U01\"),\n",
    "                                 x = f\"{row.Protein}_Non_UV\",\n",
    "                                 y = f\"{row.Protein}_UV\")\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_SUPP}/Scatterplot_{row.Protein}_A.svg\",\n",
    "               obj=plot)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S5B\n",
    "\n",
    "Given data from a follow-up experiment, Prediction intervals based on an OLS of non-damageable sequences in the library\n",
    "2. The 5U01 full and half sequences in 2 different colors\n",
    "3. Sequences that either contain a consensus sequence or a 1-nt variation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Panel</th>\n",
       "      <th>Protein</th>\n",
       "      <th>File_Non_UV</th>\n",
       "      <th>File_UV</th>\n",
       "      <th>Range_Start</th>\n",
       "      <th>Range_End</th>\n",
       "      <th>Ticker_A</th>\n",
       "      <th>Ticker_B</th>\n",
       "      <th>Ticker_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1_WC_ID38_alldata.txt</td>\n",
       "      <td>CREB1_UV_ID39_alldata.txt</td>\n",
       "      <td>500</td>\n",
       "      <td>110000</td>\n",
       "      <td>1000</td>\n",
       "      <td>10000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1_WC_ID40_alldata.txt</td>\n",
       "      <td>EGR1_UV_ID41_alldata.txt</td>\n",
       "      <td>500</td>\n",
       "      <td>110000</td>\n",
       "      <td>1000</td>\n",
       "      <td>10000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Panel Protein                File_Non_UV                    File_UV  \\\n",
       "0     C   CREB1  CREB1_WC_ID38_alldata.txt  CREB1_UV_ID39_alldata.txt   \n",
       "1     C    EGR1   EGR1_WC_ID40_alldata.txt   EGR1_UV_ID41_alldata.txt   \n",
       "\n",
       "   Range_Start  Range_End  Ticker_A  Ticker_B  Ticker_C  \n",
       "0          500     110000      1000     10000    100000  \n",
       "1          500     110000      1000     10000    100000  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_b_meta = pd.read_csv(FIGS5_META_DATA)\n",
    "panel_b_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Group</th>\n",
       "      <th>Name</th>\n",
       "      <th>WT_Seq</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>Binding_Site</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>DN_CREB1_CrystalTGACGTCA</td>\n",
       "      <td>DN_CREB1_CrystalTGACGTCA_posW-&gt;WT</td>\n",
       "      <td>GCATACGGCTGACGTCAGCCACATACGCACACATACACATACACAC...</td>\n",
       "      <td>0</td>\n",
       "      <td>TGACGTCA</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>DN_CREB1_TGACGTCA</td>\n",
       "      <td>DN_CREB1_TGACGTCA_posW-&gt;WT</td>\n",
       "      <td>GCATACATATGACGTCATATGTATGCGCACACATACACATACACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>TGACGTCA</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>DN_EGR1_CGCCCACGC</td>\n",
       "      <td>DN_EGR1_CGCCCACGC_posW-&gt;WT</td>\n",
       "      <td>ACGTATGCACGCCCACGCGTATGTACGCACACATACACATACACAC...</td>\n",
       "      <td>3</td>\n",
       "      <td>CGCCCACGC</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>DN_EGR1_GCGTGGGCG</td>\n",
       "      <td>DN_EGR1_GCGTGGGCG_posW-&gt;WT</td>\n",
       "      <td>ACATACGCGTGGGCGTGCATACGTACGCACACATACACATACACAC...</td>\n",
       "      <td>6</td>\n",
       "      <td>GCGTGGGCG</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein                     Group                               Name  \\\n",
       "0   CREB1  DN_CREB1_CrystalTGACGTCA  DN_CREB1_CrystalTGACGTCA_posW->WT   \n",
       "1   CREB1         DN_CREB1_TGACGTCA         DN_CREB1_TGACGTCA_posW->WT   \n",
       "2    EGR1         DN_EGR1_CGCCCACGC         DN_EGR1_CGCCCACGC_posW->WT   \n",
       "3    EGR1         DN_EGR1_GCGTGGGCG         DN_EGR1_GCGTGGGCG_posW->WT   \n",
       "\n",
       "                                              WT_Seq  Variant_ID Binding_Site  \\\n",
       "0  GCATACGGCTGACGTCAGCCACATACGCACACATACACATACACAC...           0     TGACGTCA   \n",
       "1  GCATACATATGACGTCATATGTATGCGCACACATACACATACACAC...           1     TGACGTCA   \n",
       "2  ACGTATGCACGCCCACGCGTATGTACGCACACATACACATACACAC...           3    CGCCCACGC   \n",
       "3  ACATACGCGTGGGCGTGCATACGTACGCACACATACACATACACAC...           6    GCGTGGGCG   \n",
       "\n",
       "   Start  End  \n",
       "0      9   17  \n",
       "1      9   17  \n",
       "2      9   18  \n",
       "3      6   15  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_design = pd.read_csv(DN_DESIGN_META_DATA)\n",
    "consensus_design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (1) Functions and Named Tuples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Named tuple for group names based of dinucelotide variant designs\n",
    "Dinucleotide_Group = namedtuple(\"DN_Group\", [\"consensus\", \"snv\", \"dnv\"])\n",
    "\n",
    "def dinucleotide_group_names(query_df: pd.DataFrame,\n",
    "                             design_df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Labels sequences into consensus, snv, or dnv across multiple groups.\n",
    "    \n",
    "    For dinucleotide variation assays, probes can be in one of the following\n",
    "    categories:\n",
    "    \n",
    "    1. Consensus: The sequence is the WT or has a variation outside the binding\n",
    "    site.\n",
    "    2. SNV: The sequence has a 1-nt variation in the binding site\n",
    "    3. DNV: The sequences has a 2-nt variation in the binding site\n",
    "    \n",
    "    A namedtuple is returned containing sets of sequence names belonging to each\n",
    "    group.\n",
    "    \"\"\"\n",
    "    consensus_set = set()\n",
    "    snv_set = set()\n",
    "    dnv_set = set()\n",
    "    for row in design_df.itertuples():\n",
    "        # Query for a group of sequences\n",
    "        df = query_df[query_df[\"Name\"].str.startswith(row.Group)]\n",
    "        df = df.reset_index(drop=True)\n",
    "        for name, sequence in zip(df[\"Name\"], df[\"Sequence\"]):\n",
    "            binding_site = sequence[row.Start:row.End]\n",
    "            if binding_site == row.Binding_Site:\n",
    "                consensus_set.add(name)\n",
    "            elif uac.hamming_distance(binding_site, row.Binding_Site) == 1:\n",
    "                snv_set.add(name)\n",
    "            elif uac.hamming_distance(binding_site, row.Binding_Site) == 2:\n",
    "                dnv_set.add(name)\n",
    "            else:\n",
    "                raise ValueError(\"Unexpected difference found\")\n",
    "    return Dinucleotide_Group(consensus_set, snv_set, dnv_set)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "wt_dict = {\"CREB1\":(\"DN_CREB1_CrystalTGACGTCA\",\"DN_CREB1_TGACGTCA\", \"DN_RELA_Double\", \"DN_RELA_Single\" ),\n",
    "           \"EGR1\": (\"DN_EGR1_CGCCCACGC\", \"DN_EGR1_GCGTGGGCG\", \"DN_RELA_Double\", \"DN_RELA_Single\")}\n",
    "panel_b_data = []\n",
    "for row in panel_b_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    df = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    # Add OLS analysis\n",
    "    df = uac.ols(df, \"Signal_Non_UV\", \"Signal_UV\", 0.01, \"Has_PyDi\")\n",
    "    df = df[df[\"Name\"].apply(lambda x: any(x.startswith(y) for y in wt_dict[row.Protein]))]\n",
    "    groups = []\n",
    "    for name in df[\"Name\"]:\n",
    "        if name.startswith(\"DN_RELA_Double\"):\n",
    "            groups.append(\"Full_5U01\")\n",
    "        elif name.startswith(\"DN_RELA_Single\"):\n",
    "            groups.append(\"Half_5U01\")\n",
    "        else:\n",
    "            groups.append(\"Other\")\n",
    "    df[\"Group\"] = groups\n",
    "    # Transform plot range and ticker settings to ln scale\n",
    "    ln_plot_range = list(map(np.log, (row.Range_Start, row.Range_End)))\n",
    "    ln_plot_ticker = list(map(np.log,\n",
    "                              (row.Ticker_A, row.Ticker_B, row.Ticker_C)))\n",
    "    # Plot\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                 ln_plot_range,\n",
    "                                 10,\n",
    "                                 (\"#2b8cbe\", \"#a6bddb\", \"black\"),\n",
    "                                 ln_plot_ticker,\n",
    "                                 (\"Full_5U01\", \"Half_5U01\", \"Other\"),\n",
    "                                 True)\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_SUPP}/FigS5B_{row.Protein}.svg\",\n",
    "               obj=plot)\n",
    "    df[\"Protein\"] = row.Protein\n",
    "    panel_b_data.append(df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table S5B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_s5b = pd.concat(panel_b_data).reset_index(drop=True)\n",
    "table_s5b = table_s5b[[\"Group\",\n",
    "                       \"Sequence\",\n",
    "                       \"Signal_Non_UV\",\n",
    "                       \"Signal_UV\",\n",
    "                       \"Prediction_Upper\",\n",
    "                       \"Prediction_Lower\",\n",
    "                       \"Predicted\",\n",
    "                       \"Protein\"]]\n",
    "table_s5b[\"Sequence\"] = table_s5b[\"Sequence\"].apply(lambda x: x[:25])\n",
    "table_s5b = table_s5b.sort_values(by=[\"Protein\", \"Group\", \"Sequence\"]).reset_index(drop=True)\n",
    "table_s5b.to_csv(f\"{OUTPUT_FOLDER_TABLE}/Table_S5B.csv\", index=False, float_format='%g')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S6C-F: K-mer profile comparison\n",
    "\n",
    "For groups of sequences related to all possible 1-nt variations of consensus sites, their flanking sequence and the 5U01 sequence, generate k-mer profiles. Each profile shows the e-score for the overlapping k-mer from a specified dataset. These profiles can then be compared across conditions.\n",
    "\n",
    "#### (1) Kmer profile meta data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scored_Sequence_Group</th>\n",
       "      <th>Protein_Kmer_Source</th>\n",
       "      <th>Condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5U01</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5U01</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>Non_UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5U01</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5U01</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>Non_UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>Non_UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>Non_UV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Scored_Sequence_Group Protein_Kmer_Source Condition\n",
       "0                  5U01                EGR1        UV\n",
       "1                  5U01                EGR1    Non_UV\n",
       "2                  5U01               CREB1        UV\n",
       "3                  5U01               CREB1    Non_UV\n",
       "4                 CREB1               CREB1        UV\n",
       "5                 CREB1               CREB1    Non_UV\n",
       "6                  EGR1                EGR1        UV\n",
       "7                  EGR1                EGR1    Non_UV"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read meta data and display\n",
    "kmer_profile_meta = pd.read_csv(KMER_PROFILE_META_DATA)\n",
    "kmer_profile_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmer_dict_from_df(input_df, condition, protein):\n",
    "    \"\"\"DOC\"\"\"\n",
    "    df= input_df[(input_df[\"Condition\"] == condition) &\n",
    "                        (input_df[\"Protein\"] == protein)]\n",
    "    result = dict(zip(df[\"7-mer\"], df[\"E-score\"]))\n",
    "    result.update(dict(zip(df[\"7-mer.1\"], df[\"E-score\"])))\n",
    "    return result\n",
    "\n",
    "def kmer_scores(kmer_dict, sequence, k=7, thr=0.35):\n",
    "    \"\"\"DOC\"\"\"\n",
    "    pos_scores = []\n",
    "    for i in range(len(sequence) - k + 1):\n",
    "        subsequence = sequence[i:i+k]\n",
    "        score = kmer_dict[subsequence]\n",
    "        pos_scores.append((i, score))\n",
    "    result = pd.DataFrame(pos_scores)\n",
    "    result = result.rename(columns={0:\"Position\", 1:\"Score\"})\n",
    "    result[\"Thr_Group\"] = \"Above\" if max(result[\"Score\"]) >= thr else \"Below\"\n",
    "    return result\n",
    "\n",
    "def kmer_scores_across_sequences(kmer_dict, sequence_list, k=7):\n",
    "    \"\"\"DOC\"\"\"\n",
    "    dfs = []\n",
    "    for sequence in sequence_list:\n",
    "        kmer_score_df = kmer_scores(kmer_dict, sequence, k=k)\n",
    "        kmer_score_df[\"Sequence\"] = sequence\n",
    "        dfs.append(kmer_score_df)\n",
    "    result = pd.concat(dfs).reset_index(drop=True)\n",
    "    return result\n",
    "\n",
    "def score_sequence_group(sequences, KMER_FILES=KMER_FILES):\n",
    "    dfs = []\n",
    "    for protein, condition, file in KMER_FILES:\n",
    "        kmer_dict = kmer_dict_from_df(kmer_df, condition, protein)\n",
    "        score_df = kmer_scores_across_sequences(kmer_dict, sequences)\n",
    "        score_df[\"Condition\"] = condition\n",
    "        score_df[\"Protein\"] = protein\n",
    "        dfs.append(score_df)\n",
    "    result = pd.concat(dfs).reset_index(drop=True)\n",
    "    return result\n",
    "\n",
    "\n",
    "def kmer_profile_plot(scored_sequences_df,\n",
    "                      protein,\n",
    "                      condition,\n",
    "                      threshold=0.35):\n",
    "    \"\"\"Makes a profile plot of k-mer scores for given parameters.\n",
    "    \n",
    "    Given a scored sequence dataframe, protein, condition, generate a bokeh figure with\n",
    "    line plots for each unique sequence. Add a horizontal, dashed line indicating a score\n",
    "    threshold.\n",
    "    \"\"\"\n",
    "    # Subset the scored sequence dataframe by condition and protein\n",
    "    plot_df = scored_sequences_df.query(\"Protein == @protein & Condition == @condition\")\n",
    "    # Create a bokeh figure object\n",
    "    p = figure(y_range=(-0.52, 0.55))\n",
    "    # For the subsetted dataframe, get all unique sequences and make a line plot for each\n",
    "    unique_sequences = list(set(plot_df[\"Sequence\"]))\n",
    "    for seq in unique_sequences:\n",
    "        df = plot_df[plot_df[\"Sequence\"] == seq].sort_values(by=\"Position\")\n",
    "        p.line(df[\"Position\"], df[\"Score\"], color=\"black\", line_width=2, alpha=0.15)\n",
    "    # Plot a horizontal, dashed line indicating the threshold\n",
    "    hline = Span(location=threshold,\n",
    "                 dimension='width',\n",
    "                 line_color='black',\n",
    "                 line_dash='dashed',\n",
    "                 line_width=3)\n",
    "    hline2 = Span(location=0.5,\n",
    "                 dimension='width',\n",
    "                 line_color='black',\n",
    "                 line_dash='dashed',\n",
    "                 line_width=3)\n",
    "    p.renderers.extend([hline])\n",
    "    p.renderers.extend([hline2])\n",
    "    # Set plot parameters for the grid line color, ticks, and outline\n",
    "    p.xgrid.grid_line_color = None\n",
    "    p.ygrid.grid_line_color = None\n",
    "    p.xaxis.minor_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.outline_line_width = 1\n",
    "    p.outline_line_color = 'black'\n",
    "    # Remove toolbar and grids\n",
    "    p.toolbar_location = None\n",
    "    p.xgrid.visible = False\n",
    "    p.ygrid.visible = False\n",
    "    # Return the plot object\n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (3) Read and concatinate k-mer dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7-mer</th>\n",
       "      <th>7-mer.1</th>\n",
       "      <th>E-score</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>TTTTTTT</td>\n",
       "      <td>0.07776</td>\n",
       "      <td>Non_UV</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAC</td>\n",
       "      <td>GTTTTTT</td>\n",
       "      <td>0.03014</td>\n",
       "      <td>Non_UV</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAG</td>\n",
       "      <td>CTTTTTT</td>\n",
       "      <td>0.07091</td>\n",
       "      <td>Non_UV</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAT</td>\n",
       "      <td>ATTTTTT</td>\n",
       "      <td>0.02219</td>\n",
       "      <td>Non_UV</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAACA</td>\n",
       "      <td>TGTTTTT</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>Non_UV</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32763</th>\n",
       "      <td>TTGCCAA</td>\n",
       "      <td>TTGGCAA</td>\n",
       "      <td>-0.11773</td>\n",
       "      <td>UV</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32764</th>\n",
       "      <td>TTGGAAA</td>\n",
       "      <td>TTTCCAA</td>\n",
       "      <td>0.16756</td>\n",
       "      <td>UV</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32765</th>\n",
       "      <td>TTGTAAA</td>\n",
       "      <td>TTTACAA</td>\n",
       "      <td>0.01032</td>\n",
       "      <td>UV</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32766</th>\n",
       "      <td>TTTAAAA</td>\n",
       "      <td>TTTTAAA</td>\n",
       "      <td>0.29989</td>\n",
       "      <td>UV</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32767</th>\n",
       "      <td>TTTCAAA</td>\n",
       "      <td>TTTGAAA</td>\n",
       "      <td>0.07186</td>\n",
       "      <td>UV</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32768 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         7-mer  7-mer.1  E-score Condition Protein\n",
       "0      AAAAAAA  TTTTTTT  0.07776    Non_UV   CREB1\n",
       "1      AAAAAAC  GTTTTTT  0.03014    Non_UV   CREB1\n",
       "2      AAAAAAG  CTTTTTT  0.07091    Non_UV   CREB1\n",
       "3      AAAAAAT  ATTTTTT  0.02219    Non_UV   CREB1\n",
       "4      AAAAACA  TGTTTTT  0.12400    Non_UV   CREB1\n",
       "...        ...      ...      ...       ...     ...\n",
       "32763  TTGCCAA  TTGGCAA -0.11773        UV    EGR1\n",
       "32764  TTGGAAA  TTTCCAA  0.16756        UV    EGR1\n",
       "32765  TTGTAAA  TTTACAA  0.01032        UV    EGR1\n",
       "32766  TTTAAAA  TTTTAAA  0.29989        UV    EGR1\n",
       "32767  TTTCAAA  TTTGAAA  0.07186        UV    EGR1\n",
       "\n",
       "[32768 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read k-mer files, label, and concatinate into a single dataframe\n",
    "dfs = []\n",
    "for protein, condition, file in KMER_FILES:\n",
    "    df = pd.read_csv(f\"{KMER_FOLDER}/{file}\", sep='\\t')\n",
    "    # Subset columns by k-mer and e-score\n",
    "    df = df[[\"7-mer\", \"7-mer.1\", \"E-score\"]]\n",
    "    # Add information on the condition and protein\n",
    "    df[\"Condition\"] = condition\n",
    "    df[\"Protein\"] = protein\n",
    "    dfs.append(df)\n",
    "kmer_df = pd.concat(dfs).reset_index(drop=True)\n",
    "# Show kmer_df\n",
    "kmer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (4) Analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CREB1_Non_UV</th>\n",
       "      <th>CREB1_UV</th>\n",
       "      <th>EGR1_UV</th>\n",
       "      <th>EGR1_Non_UV</th>\n",
       "      <th>Sequence_Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5U01</td>\n",
       "      <td>ACGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.757891</td>\n",
       "      <td>7.783016</td>\n",
       "      <td>9.059808</td>\n",
       "      <td>7.613325</td>\n",
       "      <td>5U01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5U01</td>\n",
       "      <td>CCGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.795791</td>\n",
       "      <td>7.829630</td>\n",
       "      <td>9.110741</td>\n",
       "      <td>7.486613</td>\n",
       "      <td>5U01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GAGGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.844187</td>\n",
       "      <td>7.780512</td>\n",
       "      <td>9.090486</td>\n",
       "      <td>7.577634</td>\n",
       "      <td>5U01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GCAGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.757891</td>\n",
       "      <td>7.674385</td>\n",
       "      <td>9.049232</td>\n",
       "      <td>7.464223</td>\n",
       "      <td>5U01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5U01</td>\n",
       "      <td>GCCGAAATTCCCGGGAATTTCCGCG</td>\n",
       "      <td>4.653960</td>\n",
       "      <td>7.740664</td>\n",
       "      <td>9.214880</td>\n",
       "      <td>7.744570</td>\n",
       "      <td>5U01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACCTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>7.274826</td>\n",
       "      <td>8.461786</td>\n",
       "      <td>7.610853</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACGTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.553933</td>\n",
       "      <td>7.393571</td>\n",
       "      <td>8.194920</td>\n",
       "      <td>7.307538</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCACTTTTCCCGTAAATCATACA</td>\n",
       "      <td>6.540308</td>\n",
       "      <td>7.485211</td>\n",
       "      <td>8.407155</td>\n",
       "      <td>7.605890</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCAGATTTCCCGTAAATCATACA</td>\n",
       "      <td>5.930918</td>\n",
       "      <td>7.298783</td>\n",
       "      <td>8.305608</td>\n",
       "      <td>7.038784</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>Other</td>\n",
       "      <td>GCGCATATTTCCCGTAAATCATACA</td>\n",
       "      <td>6.336826</td>\n",
       "      <td>7.630947</td>\n",
       "      <td>8.547043</td>\n",
       "      <td>7.409742</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Group                   Sequence  CREB1_Non_UV  CREB1_UV   EGR1_UV  \\\n",
       "0     5U01  ACGGAAATTCCCGGGAATTTCCGCG      4.757891  7.783016  9.059808   \n",
       "1     5U01  CCGGAAATTCCCGGGAATTTCCGCG      4.795791  7.829630  9.110741   \n",
       "2     5U01  GAGGAAATTCCCGGGAATTTCCGCG      4.844187  7.780512  9.090486   \n",
       "3     5U01  GCAGAAATTCCCGGGAATTTCCGCG      4.757891  7.674385  9.049232   \n",
       "4     5U01  GCCGAAATTCCCGGGAATTTCCGCG      4.653960  7.740664  9.214880   \n",
       "..     ...                        ...           ...       ...       ...   \n",
       "556  Other  GCGCACCTTTCCCGTAAATCATACA      6.135565  7.274826  8.461786   \n",
       "557  Other  GCGCACGTTTCCCGTAAATCATACA      6.553933  7.393571  8.194920   \n",
       "558  Other  GCGCACTTTTCCCGTAAATCATACA      6.540308  7.485211  8.407155   \n",
       "559  Other  GCGCAGATTTCCCGTAAATCATACA      5.930918  7.298783  8.305608   \n",
       "560  Other  GCGCATATTTCCCGTAAATCATACA      6.336826  7.630947  8.547043   \n",
       "\n",
       "     EGR1_Non_UV Sequence_Group  \n",
       "0       7.613325           5U01  \n",
       "1       7.486613           5U01  \n",
       "2       7.577634           5U01  \n",
       "3       7.464223           5U01  \n",
       "4       7.744570           5U01  \n",
       "..           ...            ...  \n",
       "556     7.610853          Other  \n",
       "557     7.307538          Other  \n",
       "558     7.605890          Other  \n",
       "559     7.038784          Other  \n",
       "560     7.409742          Other  \n",
       "\n",
       "[561 rows x 7 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_a_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get sequence data from design\n",
    "for row in panel_b_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    sequence_library_b = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    break\n",
    "\n",
    "sequence_dict = {}\n",
    "sequence_dict[\"CREB1\"] = sequence_library_b[(sequence_library_b[\"Name\"].str.startswith(\"DN_CREB1_TGACGTCA\")) &\n",
    "                               (sequence_library_b[\"Name\"].str.contains(\"-DN\"))][\"Sequence\"]\n",
    "sequence_dict[\"EGR1\"] = sequence_library_b[(sequence_library_b[\"Name\"].str.startswith(\"DN_EGR1_CGCCCACGC\")) &\n",
    "                               (sequence_library_b[\"Name\"].str.contains(\"-DN\"))][\"Sequence\"]\n",
    "\n",
    "for protein, condition, file in PANEL_A_FILES:\n",
    "    sequence_library_a = process_alldata_file(f\"{ALLDATA_FOLDER}/{file}\")\n",
    "    break\n",
    "sequence_dict[\"5U01\"] = sequence_library_a[(sequence_library_a[\"Name\"].str.startswith(\"RelA\"))][\"Sequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary with keys as group names and values as sequence lists\n",
    "# Add pd.Series to given groups\n",
    "\n",
    "\n",
    "for key in sequence_dict:\n",
    "    # convert the pd.Series into a list of sequence subset by the first 25 characters\n",
    "    sequence_dict[key] = list(map(lambda x: x[:25], sequence_dict[key]))\n",
    "    # Score the sequences and save as a pd.DataFrame\n",
    "    sequence_dict[key] = score_sequence_group(sequence_dict[key])\n",
    "    \n",
    "# Create profile plots based on parameters in k-mer profile plot meta data\n",
    "for row in kmer_profile_meta.itertuples():\n",
    "    plot = kmer_profile_plot(sequence_dict[row.Scored_Sequence_Group],\n",
    "                             row.Protein_Kmer_Source,\n",
    "                             row.Condition)\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(plot, filename=(f\"{OUTPUT_FOLDER_SUPP}/{row.Protein_Kmer_Source}_\"\n",
    "                               f\"{row.Condition}_{row.Scored_Sequence_Group}.svg\"))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S6G: Unique k-mer boxplot comparison\n",
    "\n",
    "Given the increase is repeatable, could the k-mer data between non-UV and UV offer an explanation as to why the increase occurs? This analysis:\n",
    "\n",
    "1. Retrieves all 5U01 sequences and finds all unique k-mers. \n",
    "2. Subsets the CREB1 and EGR1 k-mer data by thes k-mers\n",
    "3. Determines if there a difference between non-UV and UV conditions for these k-mers\n",
    "\n",
    "A wilcoxon rank-sum test and benjamini-hochberg correction results are saved as a table and a boxplot of the data is saved as a svg file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>P_Value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>1.130440e-11</td>\n",
       "      <td>1.130440e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>2.000925e-12</td>\n",
       "      <td>4.001851e-12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein       P_Value           FDR\n",
       "0   CREB1  1.130440e-11  1.130440e-11\n",
       "1    EGR1  2.000925e-12  4.001851e-12"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEGCAYAAAB7DNKzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAUkElEQVR4nO3df5BdZ33f8fdXPyyvZRJk2cDaxt5VZEhNSt10ATeUIiLcYifFyTStTVowmWRdPEAgIRR73WnTJpHdSRvCTEg9WkMwlImUBEocRoQxGyu4FFrLxmRia4ztXbBV30pCUoxZLYpW+vaPe4Sv13efXe3uvedq9/2a2dl7znnuOd/dubuf+5zn3OdEZiJJ0mxW1V2AJKm3GRSSpCKDQpJUZFBIkooMCklS0Zq6C+iE888/PwcGBuouQ5LOGA888MB3MvOCdtuWZVAMDAywZ8+eusuQpDNGRHx7tm2eepIkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklRkUEiSigwKSafl8OHD3HLLLRw5cqTuUtQly3L22LqNjo4yMTGxqH00Gg0A+vv7F7yPwcFBhoeHF1WHNNPOnTt55JFH2LFjBzfddFPd5agL7FH0qKmpKaampuouQ3qew4cPMzY2RmYyNjZmr2KFsEfRAUvxLn5kZASAbdu2LXpfdbJ3tbzs3LmTkydPAnDy5El7FSuEPQr1PHtXvWP37t1MT08DMD09ze7du+stSF1hj0IdZe9qedmyZQv33HMP09PTrFmzhi1bttRdkrrAoJBWmMWcDjx+/PgPehQnTpxgfHz8B0F+unrhdOBiT40uxWlR6I3fRYlBIWne1q5dy+rVqzlx4gQbNmxg7dq1dZdUq5VySrTWoIiItwAfAVYDd2bm7bO0ew3wNeC6zPyTLpYoLTuLfef6wQ9+kKeeeooPf/jDbNiwYYmqqsdifxcr5bRobYPZEbEa+ChwNXA58LaIuHyWdv8Z+GJ3K5TUztq1a9m0adMZHxKavzqvenot8Hhmjmfm3wI7gGvbtHsv8BngQDeLkyQ11RkUFwFPtSzvq9b9QERcBPwscMdcO4uIGyNiT0TsOXjw4JIWKkkrWZ1BEW3W5Yzl3wU+lJkn5tpZZm7PzKHMHLrggguWoj5JEvUOZu8DXt6yfDHw9Iw2Q8COiAA4H7gmIqYz83NdqVCSVGtQ3A9cFhGDwP8Frgd+vrVBZg6eehwRnwA+b0hIUnfVFhSZOR0R76F5NdNq4OOZ+XBEvKvaPue4hCSp82r9HEVm7gJ2zVjXNiAy853dqEmS9HxOCihJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSkUEhSSoyKCRJRQaFJKnIoJAkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklRkUEiSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkorW1F2AJC3E6OgoExMTtdYwPj4OwMjISK11DA4OMjw83LH9GxSSzkgTExPs3buXvr6+2mo4fvw4AN/61rdqq2Fqaqrjx6g1KCLiLcBHgNXAnZl5+4zt/wr4ULX4PeCmzPxGd6uU1Kv6+vp45StfWXcZtXr00Uc7fozaxigiYjXwUeBq4HLgbRFx+YxmE8AbM/PVwG8A27tbpSSpzsHs1wKPZ+Z4Zv4tsAO4trVBZv6vzDxSLX4NuLjLNUrSilfnqaeLgKdalvcBryu0/0XgC7NtjIgbgRsBLrnkkqWoTzhg2KrTA4ZSr6ozKKLNumzbMOJNNIPiH822s8zcTnVqamhoqO1+dPocMGzqxoCh1KvqDIp9wMtbli8Gnp7ZKCJeDdwJXJ2Zh7pUm1o4YNidAUOpV9U5RnE/cFlEDEbEWcD1wN2tDSLiEuCzwNsz85s11ChJK15tPYrMnI6I9wBfpHl57Mcz8+GIeFe1/Q7g3wMbgd+PCIDpzByqq2ZJWolq/RxFZu4Cds1Yd0fL418CfqnbdUmSnuNcT5KkIoNCklRkUEiSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpCJvhSrpjNRoNDh69OiKn7Dx6NGjNBqNjh7DHoUkqcgehaQzUn9/P8eOHXMK/Ecfpb+/v6PHsEchSSoyKCRJRQaFJKnIoJAkFTmYLZ1BRkdHmZiYqLWG8fFxAEZGRmqto9OXhOo5BoV0BpmYmODxxx/v+FUuJWvWNP9tTE5O1lZDo9Fg1apVrFrlSZFuMCikM0x/fz/Dw8N1l1Gr0dFR9u/fX3cZK4ZxLEkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklTk5bEqcs7/pm7M+S/1qjl7FBHx0oj4WER8oVq+PCJ+sfOlSZJ6wXx6FJ8A/gC4tVr+JrAT+FiHalIPcc7/pm7M+S/1qvmMUZyfmX8EnATIzGngREerkiT1jPkExWREbAQSICKuBJ7paFWSpJ4xn1NPvwrcDfxIRHwFuAD4uY5WJUnqGXMGRWY+GBFvBF4JBPBoZh7veGWSNIepqalar8g7duwYAOvWrauthqmpqY4fY86giIh3zFj14xFBZn6yQzVJ0pz6+vrYtGlTrTWcujfHwMBArXUMDg52dP/zOfX0mpbHZwNbgQeBRQdFRLwF+AiwGrgzM2+fsT2q7dcAR4F3ZuaDiz2upDNff38/27Ztq7WGUzdvqruOTpvPqaf3ti5HxA8Dn1rsgSNiNfBR4CpgH3B/RNydmY+0NLsauKz6eh3w36rvHeVdxJ7jh8wkLeST2Udp/uNerNcCj2fmOEBE7ACuBVqD4lrgk5mZwNci4sUR0Z+ZHf3v5V3EmryLmCSY3xjFn1FdGkvzctrLgT9agmNfBDzVsryPF/YW2rW5CHhBUETEjcCNAJdccsmii/MuYt5FTFLTfHoU/6Xl8TTw7czctwTHjjbrcgFtmisztwPbAYaGhtq2kSSdvvmMUfxlh469D3h5y/LFwNMLaCNJ6qBZTz5HxLMR8d02X89GxHeX4Nj3A5dFxGBEnAVcT/ODfa3uBt4RTVcCz3R6fEKS9Hyz9igy80WdPHBmTkfEe4Av0rw89uOZ+XBEvKvafgewi+alsY/THET/hU7WJPW6RqPB5OQko6OjdZdSq0ajwfr16+suY8WY91VPEfESmp+jACAzn1zswTNzF80waF13R8vjBN692ONIkhZuPlc9vRX4r8CFwAHgUmAv8KrOliZppv7+fiYnJ70ib3TUHkUXzecC+d8ArgS+mZmDND+Z/ZWOViVJ6hnzCYrjmXkIWBURqzLzXuCKzpYlSeoV8xmj+JuIOBe4D/h0RByg+XkKSdIKMJ8exZeBFwPvA/4ceAL4Zx2sSZLUQ+YTFEHzEtbdwLnAzupUlCRpBZgzKDLzP2bmq2hepnoh8JcR8aWOVyZJ6gmnMy3oAeD/AYeAl3SmHElSr5kzKCLipojYDYwB5wPDmfnqThcmSeoN87nq6VLg/Zn5UIdrkST1oPnMHntzNwqRJPUmb10mSSoyKCRJRQaFJKnIoJAkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpaD6TAkrqIY1Gg9HR0dqOf+hQ875lGzdurK2GRqPB5s2bazv+SmNQSGeQwcHBuktg//79AKxfv762GjZv3twTv4uVwqDQnKampnj00UdrO/6xY8cAWLduXW01TE1N1XbsVsPDw3WXwMjICADbtm2ruRJ1i0Ghor6+PjZt2lRrDePj4wAMDAzUWofvYLVSGRQq6u/vr/2do+9gpXoZFG00Gg0mJydrHTDsBY1Go9bz0JJ6g5fHSpKK7FG00d/fz+TkZE8MHNZpdHTUHoUkexSSpDKDQpJUZFBIkopqCYqIOC8i7omIx6rvG9q0eXlE3BsReyPi4Yh4Xx21StJKV1eP4mZgLDMvA8aq5ZmmgQ9k5t8BrgTeHRGXd7FGSRL1BcW1wF3V47uAn5nZIDMbmflg9fhZYC9wUbcKlCQ11RUUL83MBjQDAXhJqXFEDAB/H/jfhTY3RsSeiNhz8ODBpaxVkla0jn2OIiK+BLyszaZbT3M/5wKfAd6fmd+drV1mbge2AwwNDeXpHEOSNLuOBUVmvnm2bRGxPyL6M7MREf3AgVnaraUZEp/OzM92qFRJUkFdp57uBm6oHt8A/OnMBhERwMeAvZn5O12sTZLUoq6guB24KiIeA66qlomICyNiV9Xm9cDbgZ+MiIeqr2vqKVeSVq5a5nrKzEPA1jbrnwauqR7/TyC6XJokaQY/mS1JKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSUS0fuJOkXjA6OsrExMSCnz8+Pg7AyMjIouoYHBxkeHh4UfvoJINCkhaor6+v7hK6wqCQtGL18rv4XuIYhSSpyB7FLBqNBqOjo7Ud/9ChQwBs3LixthoajQabN2+u7fiSeoNB0cbg4GDdJbB//34A1q9fX1sNmzdv7onfhaR6GRRt9MJ5y1NXUWzbtq3mSiStdAaFtMJ4SahOl4PZkk5LX1/firksdC6HDx/mlltu4ciRI3WX0lH2KKQVxnfxS2fnzp088sgj7Nixg5tuuqnucjrGHoUkLcDhw4cZGxsjMxkbG1vWvQqDQpIWYOfOnZw8eRKAkydPsmPHjpor6hyDQpIWYPfu3UxPTwMwPT3N7t276y2ogwwKSVqALVu2sGZNc5h3zZo1bNmypd6COsigkKQFuO6661i1qvkvdNWqVVx//fU1V9Q5BoUkLcB5553H1q1biQi2bt3Khg0b6i6pY7w8VpIW6LrrruPJJ59c1r0JMCgkacHOO+88brvttrrL6DhPPUmSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpKJagiIizouIeyLiser7rB9pjIjVEfH1iPh8N2uUJDXV1aO4GRjLzMuAsWp5Nu8D9nalKknSC9QVFNcCd1WP7wJ+pl2jiLgY+Cngzu6UJUmaqa6geGlmNgCq7y+Zpd3vAv8WODnXDiPixojYExF7Dh48uGSFStJK17G5niLiS8DL2my6dZ7P/2ngQGY+EBFb5mqfmduB7QBDQ0M5/0olSSUdC4rMfPNs2yJif0T0Z2YjIvqBA22avR54a0RcA5wN/FBE/PfM/NcdKlmS1EZdp57uBm6oHt8A/OnMBpl5S2ZenJkDwPXAXxgSktR9dQXF7cBVEfEYcFW1TERcGBG7aqpJktRGLfejyMxDwNY2658Grmmzfjewu+OFSZJewE9mS5KKDApJUpFBIUkqMigkSUW1DGZr5RgdHWViYmJR+xgfHwdgZGRkwfsYHBxkeHh4UXVIK5VBoZ7X19dXdwnSihaZy2+2i6GhodyzZ09tx1/Kd9GbNm1a8D58Fy1pviLigcwcarfNHkWP8l20pF5hUHSA7+IlLSde9SRJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSkUEhSSoyKCRJRctyUsCIOAh8u+46lonzge/UXYQ0C1+fS+fSzLyg3YZlGRRaOhGxZ7YZJaW6+frsDk89SZKKDApJUpFBoblsr7sAqcDXZxc4RiFJKrJHIUkqMigkSUUGxQoQES+LiB0R8UREPBIRuyLiFRExFREPVes+GRFrq/ZbIuKZatuprzdX205Uy9+IiAcj4idajvPnEfE3EfH5un5WnblaXlunvm6u1q+JiG0R8VjLtlvbPO+vI+LPIuLFLdt8TS4B75m9zEVEAP8DuCszr6/WXQG8FHgiM6+IiNXAPcC/BD5dPfW+zPzpNrucyswrqv38U+A24I3Vtt8GzgH+TWd+Gi1zP3htzfCbwMuAv5uZ34+IFwEfaPe8iLgLeDfwW9U2X5NLwB7F8vcm4Hhm3nFqRWY+BDzVsnwC+D/ARae57x8CjrTsZwx4djHFSq0i4hxgGHhvZn4fIDOfzcxfn+UpX6XldexrcmnYo1j+fgx4oNQgIs4GXge8r2X1GyLioZblf56ZTwB91fqzgX7gJ5e0Wq1kfTNec7cBe4EnM3POf/ZVz3gr8LHOlLdyGRQr249Uf5iXAX+SmX/Vsm0+p57+IfDJiPix9DprLd4LTj1FxKtnLP8CzTc0G4GfyMyneC5gBmi+KbqnG8WuJJ56Wv4eBv7BLNueqP4wNwNXRsRbT2fHmflVmpOytZ1ITFoCjwOXVOMSZOYfVK/ZZ4DVVZtTAXMpcBbNMQotIYNi+fsLYF1EDJ9aERGvoflHBUBmNoCbgVtOZ8cR8aM0/1gPLU2p0vNl5lGap5J+rzpFeuoU01lt2j4D/DLwa6eu4NPSMCiWueqU0M8CV1WXxz4M/Drw9IymnwPOiYg3VMtvmHGp4s9V6/tOrQN2AjdUg+FExH3AHwNbI2JfdVWUNF99M15zt1frbwUawF9HxNeB+4C7eOFrmMz8OvAN4NQVfr4ml4BTeEiSiuxRSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQ5tAyO+nD1ay5vxoRxb+diBiIiJ/vQC3vr+Y/krrGoJDmNpWZV2Tmq4CrgGuA/zDHcwaAJQ8K4P00Z0OVusagkE5DZh4AbgTeE00DEXFfdW+O1vtz3M5zH1r8ldnaRUR/RHy55X4Kb6jW/5OI+GrV9o8j4tyI+GXgQuDeiLi3jp9fK5MfuJPmEBHfy8xzZ6w7AvwozSmsT1b3SbgM+MPMHIqILcCvnZpYsTpd1K7dB4CzM/O3qqkpzgHWAZ8Frs7MyYj4ELAuM/9TRHwLGMrM73Tlh5dw9lhpoaL6vpbmPERXACeAV8zSfrZ29wMfr+Ym+lxmPhQRbwQuB77SvO8UZ9G8z4JUC4NCOk0RsYnmP/sDNMcq9gN/j+ap3O/P8rRfadcuM78cEf8Y+CngUxHx2zRvBnVPZr6tkz+HNF+OUUinISIuAO4Afq+acPGHgUZmngTeznNTXz8LvKjlqW3bRcSlwIHMHKU5S+qPA18DXh8Rm6s250TEK2bZr9Rx9iikuZ26Mc5aYBr4FPA71bbfBz4TEf8CuBeYrNb/FTAdEd8APlFotwX4YEQcB74HvCMzD0bEO4E/jIh1Vbt/B3wT2A58ISIamfmmzvy40vM5mC1JKvLUkySpyKCQJBUZFJKkIoNCklRkUEiSigwKSVKRQSFJKvr/cgoKAK12b7gAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get 5u01 sequences from panel A as a list\n",
    "#sequences_5u01 = panel_a_data[(panel_a_data[\"Protein\"] == \"CREB1\") &\n",
    "#                              (panel_a_data[\"Name\"].str.startswith(\"RelA\"))]\n",
    "sequences_5u01 = list(panel_a_data[panel_a_data[\"Group\"] == \"5U01\"][\"Sequence\"])\n",
    "# Determine all unique 7-mers from the list of 5u01 sequences\n",
    "unique_5u01_kmers = set()\n",
    "for sequence in sequences_5u01:\n",
    "    unique_5u01_kmers.update(uac.unique_kmers(sequence, 7))\n",
    "# For each transcription factor\n",
    "kmer_dfs = []\n",
    "for kmer_files, tf in ((CREB1_KMERS, \"CREB1\"),\n",
    "                         (EGR1_KMERS, \"EGR1\")):\n",
    "    # Read the k-mer data as a pd.DataFrame\n",
    "    non_uv_file, uv_file = kmer_files\n",
    "    kmers = uac.read_kmer_file_pair(f\"{KMER_FOLDER}/{non_uv_file}\",\n",
    "                                    f\"{KMER_FOLDER}/{uv_file}\",\n",
    "                                    (\"_Non_UV\", \"_UV\"))\n",
    "    # Subset it by those k-mers present in 5U01 sequences\n",
    "    kmers_5u01 = kmers[(kmers[\"7-mer\"].isin(unique_5u01_kmers)) |\n",
    "                      (kmers[\"7-mer.1\"].isin(unique_5u01_kmers))]\n",
    "    kmers_5u01 = kmers_5u01.reset_index(drop=True)\n",
    "    # Add information on the TF, append it to the dataframes processed\n",
    "    kmers_5u01[\"Dataset\"] = tf\n",
    "    kmer_dfs.append(kmers_5u01)\n",
    "# Concatintate dataframes into one\n",
    "kmer_measurements_5u01 = pd.concat(kmer_dfs).reset_index(drop=True)\n",
    "# Compare groups with wilcoxon rank sum test\n",
    "wilcoxon_results = []\n",
    "for dataset, df in kmer_measurements_5u01.groupby(by=\"Dataset\"):\n",
    "    ranksum_result = stats.ranksums(df[\"E-score_Non_UV\"], df[\"E-score_UV\"])\n",
    "    wilcoxon_results.append((dataset, ranksum_result.pvalue))\n",
    "# Store the results as a pd.DataFrame\n",
    "escore_comparison = pd.DataFrame(wilcoxon_results)\n",
    "escore_comparison = escore_comparison.rename(columns={0:\"Protein\", 1:\"P_Value\"})\n",
    "# Benjamini correction for multiple tests to produce false discovery rates\n",
    "fdr_values = multipletests(escore_comparison[\"P_Value\"], method=\"fdr_bh\")[1]\n",
    "escore_comparison[\"FDR\"] = fdr_values\n",
    "# Save results\n",
    "escore_comparison.to_csv(f\"{OUTPUT_FOLDER_SUPP}/Panel_E_Statistics.csv\")\n",
    "# Create a boxplot of the data and save\n",
    "# Transform the data to long form\n",
    "plot_df = pd.melt(kmer_measurements_5u01,\n",
    "                  id_vars=['7-mer', \"Dataset\"],\n",
    "                  value_vars=['E-score_Non_UV', \"E-score_UV\"])\n",
    "# Create a boxplot\n",
    "sns.boxplot(x=\"Dataset\", y=\"value\", hue=\"variable\",\n",
    "            data=plot_df,\n",
    "            palette = BOXPLOT_PALETTE,\n",
    "           hue_order=[\"E-score_Non_UV\", \"E-score_UV\"])\n",
    "plt.ylim(-0.52, 0.52)\n",
    "plt.legend([],[], frameon=False)\n",
    "plt.savefig(f\"{OUTPUT_FOLDER_SUPP}/Panel_E_Boxplot.svg\")\n",
    "escore_comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4A-B: Sequences of concatinated k-mers plotting\n",
    "\n",
    "Scatterplot comparing sequences made of concatinated k-mers that had higher E-scores in UV conditions than non-UV conditions (UV-preffered). The plots are colored in the following priority:\n",
    "\n",
    "1. Green: Sequences with a consensus sequence of the given transcription factor\n",
    "2. Light Green: Sequences with a 1-nt variation from the consensus sequences\n",
    "3. Black: Sequences made from concatinated UV-preffered k-mers\n",
    "4. Grey: Sequences used to train the OLS model\n",
    "\n",
    "The data uses the same information as panel C for Figure S6, with the only changes being the sequences being plotted. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Panel</th>\n",
       "      <th>Protein</th>\n",
       "      <th>File_Non_UV</th>\n",
       "      <th>File_UV</th>\n",
       "      <th>Range_Start</th>\n",
       "      <th>Range_End</th>\n",
       "      <th>Ticker_A</th>\n",
       "      <th>Ticker_B</th>\n",
       "      <th>Ticker_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1_WC_ID38_alldata.txt</td>\n",
       "      <td>CREB1_UV_ID39_alldata.txt</td>\n",
       "      <td>500</td>\n",
       "      <td>110000</td>\n",
       "      <td>1000</td>\n",
       "      <td>10000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>EGR1_WC_ID40_alldata.txt</td>\n",
       "      <td>EGR1_UV_ID41_alldata.txt</td>\n",
       "      <td>500</td>\n",
       "      <td>110000</td>\n",
       "      <td>1000</td>\n",
       "      <td>10000</td>\n",
       "      <td>100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Panel Protein                File_Non_UV                    File_UV  \\\n",
       "0     C   CREB1  CREB1_WC_ID38_alldata.txt  CREB1_UV_ID39_alldata.txt   \n",
       "1     C    EGR1   EGR1_WC_ID40_alldata.txt   EGR1_UV_ID41_alldata.txt   \n",
       "\n",
       "   Range_Start  Range_End  Ticker_A  Ticker_B  Ticker_C  \n",
       "0          500     110000      1000     10000    100000  \n",
       "1          500     110000      1000     10000    100000  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_b_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read meta data and query for data related to panel C\n",
    "consensus_design = pd.read_csv(DN_DESIGN_META_DATA)\n",
    "panel_c_data = []\n",
    "for row in panel_b_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    df = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    # Add OLS analysis\n",
    "    df = uac.ols(df, \"Signal_Non_UV\", \"Signal_UV\", 0.01, \"Has_PyDi\")\n",
    "    # Define sequences with a consensus site or a snv in the consensus\n",
    "    consensus_df = consensus_design[consensus_design[\"Protein\"] == row.Protein]\n",
    "    dn_groups = dinucleotide_group_names(df, consensus_df)\n",
    "    # Label Groups\n",
    "    groups = []\n",
    "    for name in df[\"Name\"]:\n",
    "        if name.startswith(\"DN_RELA_Double\"):\n",
    "            groups.append(\"Full_5U01\")\n",
    "        elif name.startswith(\"DN_RELA_Single\"):\n",
    "            groups.append(\"Half_5U01\")\n",
    "        elif name in dn_groups.consensus:\n",
    "            groups.append(\"Consensus\")\n",
    "        elif name in dn_groups.snv:\n",
    "            groups.append(\"SNV\")\n",
    "        else:\n",
    "            groups.append(\"Other_Not_Plotted\")\n",
    "    df[\"Group\"] = groups\n",
    "    df[\"Protein\"] = row.Protein\n",
    "    panel_c_data.append(df)\n",
    "panel_c_data = pd.concat(panel_c_data).reset_index(drop=True)\n",
    "\n",
    "\n",
    "relabeled_dfs = {}\n",
    "for protein in (\"CREB1\", \"EGR1\"):\n",
    "    df = panel_c_data[panel_c_data[\"Protein\"] == protein]\n",
    "    df = df.reset_index(drop=True)\n",
    "    # Rename groups to plot\n",
    "    groups = []\n",
    "    for row in df.itertuples():\n",
    "        if row.Name.startswith(f\"Clst_{protein}\"):\n",
    "            groups.append(\"Concat\")\n",
    "        elif (row.Has_PyDi is False) and (row.Group not in (\"Consensus\", \"SNV\")):\n",
    "            groups.append(\"Training\")\n",
    "        else:\n",
    "            groups.append(row.Group)\n",
    "    df[\"Group\"] = groups\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                     list(map(np.log, (400, 110000))),\n",
    "                                     10,\n",
    "                                     (\"lightgrey\", \"black\", \"green\", \"lightgreen\"),\n",
    "                                     list(map(np.log, (1000, 10000, 100000))),\n",
    "                                     (\"Training\", \"Concat\", \"Consensus\", \"SNV\"),\n",
    "                                     True)\n",
    "    plot.output_backend = 'svg'\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_MAIN}/Scatterplot_{protein}.svg\",\n",
    "               obj=plot)\n",
    "    relabeled_dfs[protein] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_4_table = pd.concat([relabeled_dfs[\"CREB1\"], relabeled_dfs[\"EGR1\"]]).reset_index(drop=True)\n",
    "figure_4_table = figure_4_table[figure_4_table[\"Group\"].isin((\"Training\", \"Concat\", \"Consensus\", \"SNV\"))].reset_index(drop=True)\n",
    "figure_4_table = figure_4_table[[\"Group\",\n",
    "                                 \"Sequence\",\n",
    "                                 \"Signal_Non_UV\",\n",
    "                                 \"Signal_UV\",\n",
    "                                 \"Prediction_Upper\",\n",
    "                                 \"Prediction_Lower\",\n",
    "                                 \"Predicted\",\n",
    "                                 \"Protein\"]]\n",
    "figure_4_table[\"Sequence\"] = figure_4_table[\"Sequence\"].apply(lambda x: x[:25])\n",
    "figure_4_table = figure_4_table.sort_values(by=[\"Protein\", \"Group\"]).reset_index(drop=True)\n",
    "figure_4_table.to_csv(f\"{OUTPUT_FOLDER_TABLE}/Table_S5C.csv\", index=False, float_format='%g')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4C-D - Highlights and KDE comparisons\n",
    "\n",
    "This panel is comprised of a figure with the following sub-plots:\n",
    "\n",
    "1. A plot at the top showing the change from non-UV to UV for 3 selected sequences as highlights\n",
    "2. A set of plots showing KDE comparisons for the following groups:\n",
    "    1. Consensus sequences\n",
    "    2. 1-nt variation sequences\n",
    "    3. Concatinated UV-preffered k-mer sequences\n",
    "\n",
    "These components are then combined into a single plot which shows the data at the same scale to compare across componenets. \n",
    "\n",
    "#### (1) Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_plot(df: pd.DataFrame,\n",
    "                   column_x: str,\n",
    "                   column_y: str,\n",
    "                   hue: str,\n",
    "                   palette: tuple[str, str],\n",
    "                   order: tuple[str, str],\n",
    "                   x_ticks: tuple[int, int],\n",
    "                   ax,\n",
    "                  x_limit):\n",
    "    sns.set_style('ticks')\n",
    "    #L=5.8\n",
    "    #U=13\n",
    "    p = sns.stripplot(data=df,\n",
    "                      x=column_x,\n",
    "                      y=column_y,\n",
    "                      hue=hue,\n",
    "                      order=order,\n",
    "                      orient='h',\n",
    "                      jitter=False,\n",
    "                      s=15,\n",
    "                      palette=['blue', 'red'],\n",
    "                      alpha=0.5,\n",
    "                      linewidth=0,\n",
    "                      ax=ax)\n",
    "    ax.get_legend().remove()\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label  \n",
    "    ax.set_xlim(x_limit[0], x_limit[1])\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "\n",
    "def kde_plot(df,\n",
    "             column_x,\n",
    "             hue,\n",
    "             x_limit,\n",
    "             y_limit,\n",
    "             ax,\n",
    "             show_x_ticks=False):\n",
    "    p = sns.kdeplot(data=df,\n",
    "                    x=column_x,\n",
    "                    hue=hue,\n",
    "                    fill=True,\n",
    "                    common_norm=False,\n",
    "                    palette=[\"blue\", 'red'],\n",
    "                    alpha=0.5,\n",
    "                    linewidth=8,\n",
    "                    ax=ax)\n",
    "    ax.set_xlim(x_limit[0], x_limit[1])\n",
    "    ax.set_ylim(y_limit[0], y_limit[1])\n",
    "    ax.get_legend().remove()\n",
    "    plt.legend('',frameon=False)\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label    \n",
    "    p.yaxis.major_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    if show_x_ticks:\n",
    "        p.set(xticks=[np.log(1000), np.log(10000), np.log(100000)])\n",
    "        p.set(xticklabels=['1000', '10000', '100000'])\n",
    "    else:\n",
    "        ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    return p\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (2) Analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "groups = (\"Concat\",\"SNV\", \"Consensus\")\n",
    "for protein, highlights in ((\"CREB1\", CREB1_HIGHLIGHT),\n",
    "                            (\"EGR1\", EGR1_HIGHLIGHT)):\n",
    "    fig, ax_array = plt.subplots(4)\n",
    "    fig.set_size_inches(10, 10)\n",
    "    kde_x, kde_y = KDE_BOUNDS[protein]\n",
    "    plot_df = relabeled_dfs[protein].copy()\n",
    "    plot_df = plot_df[plot_df[\"Name\"].isin(set(highlights))]\n",
    "    plot_df = pd.melt(plot_df,\n",
    "                      id_vars=[\"Name\"],\n",
    "                      value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    highlight_plot(plot_df,\n",
    "                   \"value\",\n",
    "                   \"Name\",\n",
    "                   \"variable\",\n",
    "                   ('blue', 'red'),\n",
    "                   highlights,\n",
    "                   [np.log(1000), np.log(10000), np.log(100000)],\n",
    "                  ax_array[0],\n",
    "                  kde_x)\n",
    "    protein_df = relabeled_dfs[protein].copy()\n",
    "    protein_df = pd.melt(protein_df, id_vars=[\"Name\", \"Group\"], value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    for idx, group in enumerate(groups):\n",
    "        plot_df = protein_df[protein_df[\"Group\"] == group]\n",
    "        plot_x_ticks = group == \"Consensus\"\n",
    "        kde_plot(plot_df,\n",
    "                 \"value\",\n",
    "                 \"variable\", \n",
    "                 kde_x,\n",
    "                 kde_y,\n",
    "                 ax_array[idx+1],\n",
    "                 plot_x_ticks)\n",
    "    plt.subplots_adjust(hspace=0.0000001)\n",
    "    plt.savefig(fname=f\"{OUTPUT_FOLDER_MAIN}/{protein}_KDE.svg\",format='svg')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
