{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6ec0023e",
   "metadata": {},
   "source": [
    "## Supplementary Figure 7 - Mutations\n",
    "\n",
    "\n",
    "\n",
    "### Dependencies \n",
    "\n",
    "These need to be installed beforehand and added to PATH. \n",
    "\n",
    "1. Bedtools\n",
    "2. MEME Suite\n",
    "    - FIMO\n",
    "    - uniprobe2meme\n",
    "    - fasta-get-markov\n",
    "        - Note: This one is optional if you pre-calculate the background file once.\n",
    "        - Note: Also not sure it makes any significant difference, uniform bkg might be better.\n",
    "\n",
    "### External file setup\n",
    "\n",
    "#### skcm_mutations_0base.bed\n",
    "\n",
    "```python\n",
    "# Adjust Coordinates (subtract start by 1) - Just done once\n",
    "with open(f\"{OUTPUT_SX}/skcm_mutations.bed\") as read_obj, open(skcm_mutations_file, 'w') as write_obj:\n",
    "    for line in read_obj:\n",
    "        chrom, start, end, wt, snp = line.strip().split()\n",
    "        start = int(start) - 1\n",
    "        write_obj.write(f\"{chrom}\\t{start}\\t{end}\\t{wt}\\t{snp}\\n\")\n",
    "```\n",
    "\n",
    "\n",
    "#### Footprint data setup\n",
    "\n",
    "\n",
    "```bash\n",
    "# Download file\n",
    "wget https://resources.altius.org/~jvierstra/projects/footprinting.2020/consensus.index/consensus_footprints_and_collapsed_motifs_hg38.bed.gz\n",
    "# Unzip\n",
    "gunzip consensus_footprints_and_collapsed_motifs_hg38.bed.gz\n",
    "# Retain only first 3 columns (coordinants)\n",
    "cut -f 1-3 consensus_footprints_and_collapsed_motifs_hg38.bed >\n",
    "consensus_footprints_and_collapsed_motifs_justcoord_hg38.bed\n",
    "# Run Liftover\n",
    "./liftover consensus_footprints_and_collapsed_motifs_justcoord_hg38.bed hg38ToHg19.over.chain consensus_footprints_and_collapsed_motifs_justcoord_hg19.bed consensus_footprints_and_collapsed_motifs_justcoord_hg19_unmapped\n",
    "```\n",
    "\n",
    "#### Imports and Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "749d450f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import gzip\n",
    "import multiprocessing as mp\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "from Bio import SeqIO\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import uvbind_analysis_core as uac\n",
    "\n",
    "#######################\n",
    "# IO Global Variables #\n",
    "#######################\n",
    "\n",
    "# Location of UV preferred 6-mers\n",
    "KMER_6MER = \"../../Design/Concatinated_Kmers_Design/Data\"\n",
    "# Location of USCS promoter fasta file\n",
    "USCS_PROMOTER_FASTA_FILE = \"../../Data/External_Data/upstream2000.fa\"\n",
    "HUMAN_GENOME_FILE = \"../../Data/External_Data/human_g1k_v37.fasta\"\n",
    "# Location of PWM to use in site calling\n",
    "PWMS_TABLE_S3A = \"../Table_S3/Table_S3A_PWM_Probabilities.csv\"\n",
    "# Output locations\n",
    "OUTPUT_SX = \"../Figure_SX\"\n",
    "MUTATION_FILE=f\"{OUTPUT_SX}/skcm_mutations.bed\"\n",
    "OUTPUT_3MERS = f\"{OUTPUT_SX}/Promoter_3mers\"\n",
    "OUTPUT_TSX = f\"../Table_SX\"\n",
    "BACKGROUND_INTERSECTION_DIR = f\"{OUTPUT_SX}/Background_Intersection\"\n",
    "PROMOTER_MUTATIONS_DIR_HS = f\"{OUTPUT_SX}/Promoter_Mutations_HS\"\n",
    "\n",
    "# Location of skcm mutations file\n",
    "skcm_mutations_file = f\"{OUTPUT_SX}/skcm_mutations_0base.bed\"\n",
    "\n",
    "################################\n",
    "# Intermediate files generated #\n",
    "################################\n",
    "\n",
    "# Bed file of 2kb promoters\n",
    "promoter_bed_file = f\"{OUTPUT_SX}/Promoter2000.bed\"\n",
    "# Sorted version of the promoter bed file\n",
    "promoter_sorted_bed_file = f\"{OUTPUT_SX}/Promoter2000_Sorted.bed\"\n",
    "# Promoter regions with no overlaps\n",
    "promoter_merged_bed_file = f\"{OUTPUT_SX}/Promoter2000_Merged.bed\"\n",
    "# Mutations within the promoter (1bp)\n",
    "skcm_promoter_mutations_file = f\"{OUTPUT_SX}/skcm_mutations_promoter2000.bed\"\n",
    "# Mutations within the promoter in a trimer context\n",
    "skcm_promoter_mutations_trimer_bed_file = f\"{OUTPUT_SX}skcm_mutations_promoter2000_trimer.bed\"\n",
    "# TSV file of the sequences\n",
    "skcm_mutations_trimer_sequence_tsv_file = f\"{OUTPUT_SX}/skcm_mutations_promoter2000_trimer_sequences.tsv\"\n",
    "# Same as skcm_promoter_mutations_file but the 4th column is the sequence\n",
    "skcm_mutations_trimer_sequence_bed_file = f\"{OUTPUT_SX}/skcm_mutations_promoter2000_trimer_sequences.bed\"\n",
    "promoter_fasta_file = f\"{OUTPUT_SX}/Promoter2000.fasta\"\n",
    "promoter_fasta_merged_file = f\"{OUTPUT_SX}/Promoter2000_Merged.fasta\"\n",
    "promoter_trimer_counts_txt_file = f\"{OUTPUT_SX}/Promoter_Trimer_Counts.txt\"\n",
    "creb1_pwm_promoter_calls_unique_file = f\"{OUTPUT_SX}/CREB1_PWM_promoter_calls_unique.bed\"\n",
    "egr1_pwm_promoter_calls_unique_file = f\"{OUTPUT_SX}/EGR1_PWM_promoter_calls_unique.bed\"\n",
    "\n",
    "##############################\n",
    "# Parameter Global Variables #\n",
    "##############################\n",
    "\n",
    "# Threads to use for multicore parts of the notebook\n",
    "THREADS = 10\n",
    "# Tuple of non-consensus k-mer runs lengths to analyze\n",
    "NC_LENGTH_TUP = (20, 25, 30)\n",
    "# Tuple of transcription factors to analyze\n",
    "TF_TUP = (\"CREB1\", \"EGR1\")\n",
    "# USE PWM CORE ONLY (Othersie CORE + 3bp either side as visualized in paper)\n",
    "PWM_CORE_ONLY = False\n",
    "# Intersect PWM calls with genomic footprints\n",
    "INTERSECT_GENOMIC_FOOTPRINTS = True\n",
    "\n",
    "####################################\n",
    "# Parameter choices from variables #\n",
    "####################################\n",
    "\n",
    "pwm_call_suffix = \"\"\n",
    "if INTERSECT_GENOMIC_FOOTPRINTS:\n",
    "    pwm_call_suffix = \"_InFootprints\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0badf971",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create output folders if not already present\n",
    "os.makedirs(OUTPUT_SX, exist_ok=True)\n",
    "os.makedirs(OUTPUT_TSX, exist_ok=True)\n",
    "os.makedirs(OUTPUT_3MERS, exist_ok=True)\n",
    "os.makedirs(f\"{OUTPUT_SX}/Promoter_Backgrounds_ZM\", exist_ok=True)\n",
    "os.makedirs(f\"{OUTPUT_SX}/Promoter_Backgrounds_HS\", exist_ok=True)\n",
    "os.makedirs(f\"{OUTPUT_SX}/Site_Shuffle\", exist_ok=True)\n",
    "os.makedirs(f\"{OUTPUT_SX}/Site_Shuffle_Intersection\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "986d0c49",
   "metadata": {},
   "source": [
    "# Functions\n",
    "\n",
    "This block is for all functions used in the analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "800176a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uv_preferred_kmer_set(kmer_file):\n",
    "    kmers = pd.read_csv(kmer_file, sep='\\t', skiprows=7)\n",
    "    uvpref = kmers[kmers[\"Escore_UV9\"] > kmers[\"Confidence_Interval_Upper\"]].reset_index(drop=True)\n",
    "    uvpkmers = set(list(uvpref[\"kmerFwd\"]) + list(uvpref[\"kmerRC\"]))\n",
    "    return uvpkmers\n",
    "\n",
    "def count_kmers_sepstrd(sequence, to_dict, k=3):\n",
    "    for i in range(len(sequence) - k + 1):\n",
    "        kmer = sequence[i:i+k]\n",
    "        if \"N\" not in kmer:\n",
    "            to_dict[kmer] += 1\n",
    "\n",
    "def count_kmers(sequence, to_dict, k=6):\n",
    "    for i in range(len(sequence) - k + 1):\n",
    "        kmer = sequence[i:i+k]\n",
    "        if \"N\" not in kmer:\n",
    "            if kmer in to_dict:\n",
    "                to_dict[kmer] += 1\n",
    "            else:\n",
    "                to_dict[uac.reverse_complement(kmer)] += 1\n",
    "            \n",
    "def count_chromosome_kmers(chromosome_file, to_dict, k=6):\n",
    "    prefix_idx = -1 * (k-1)\n",
    "    with open(chromosome_file) as file_obj:\n",
    "        next(file_obj) # Skip header\n",
    "        prefix = \"\" # Previous sequence that overlaps with the next\n",
    "        for i in file_obj:\n",
    "            line = prefix + i.strip().upper()\n",
    "            count_kmers(line, to_dict)\n",
    "            prefix = line[prefix_idx:]\n",
    "            \n",
    "def count_chromosome_kmers_pipeline(input_file, output_file, k=6):\n",
    "    result_dict = defaultdict(int)\n",
    "    count_chromosome_kmers(input_file, result_dict, k)\n",
    "    result_df = pd.DataFrame.from_dict(result_dict,\n",
    "                             orient='index').reset_index()\n",
    "    result_df = result_df.rename(columns={\"index\":\"Kmer\", 0:\"Count\"})\n",
    "    result_df.to_csv(output_file, index=False)\n",
    "    \n",
    "    \n",
    "def get_match_idxs(string, kmers, k=6):\n",
    "    \"\"\"Get indexes of k-mer matches to a string.\n",
    "    \n",
    "    Given a string and set of k-mers, return a list of start\n",
    "    indexes for matches.\n",
    "    \"\"\"\n",
    "    string = string.upper()\n",
    "    matches = []\n",
    "    for idx in range(len(string) - k + 1):\n",
    "        if string[idx:idx+k] in kmers:\n",
    "            matches.append(idx)\n",
    "    return matches\n",
    "\n",
    "\n",
    "def ranges_from_match_idxs(matches, k=6): # Gets range of first, should not start again for already considered.\n",
    "    \"\"\"Given a list of match indexes, return a list of range tuples.\n",
    "    \n",
    "    Converts a list of match indexes into a list of tuples where each\n",
    "    tuple indicates the start (inclusive) and end (exclusive) of consecutuve matches.\n",
    "    \"\"\"\n",
    "    if len(matches) == 0:\n",
    "        return []\n",
    "    elif len(matches) == 1:\n",
    "        return [(matches[0], matches[0] + k)]\n",
    "    regions = []\n",
    "    start = matches[0]\n",
    "    last_seen = start\n",
    "    for i in matches[1:]:\n",
    "        if last_seen + 1 != i:\n",
    "            # Save last result\n",
    "            regions.append((start, last_seen + 1 + k))\n",
    "            # Make new range\n",
    "            start = i\n",
    "        last_seen = i\n",
    "    if regions[-1][0] != start:\n",
    "        regions.append((start, last_seen + 1 + k))\n",
    "    return regions\n",
    "\n",
    "\n",
    "def ranges_to_bed(range_list, chromosome, start, name):\n",
    "    bed_coordinates = []\n",
    "    for i in range_list:\n",
    "        bed_coordinates.append((chromosome, start + i[0], start + i[1], name))\n",
    "    return bed_coordinates\n",
    "\n",
    "\n",
    "def name_from_label(label):\n",
    "    r = label.split(\"_\")\n",
    "    return \"_\".join(r[:2] + r[5:])\n",
    "\n",
    "\n",
    "def max_length_from_range_list(range_list):\n",
    "    if len(range_list) == 0:\n",
    "        return 0\n",
    "    length_list = list(map(lambda x: x[1] - x[0], range_list))\n",
    "    return max(length_list)\n",
    "    \n",
    "\n",
    "def call_nonconsensus_sites(string, kmers, k=6):\n",
    "    matches = get_match_idxs(string, kmers, k)\n",
    "    ranges = ranges_from_match_idxs(matches)\n",
    "    #filtered_ranges = list(filter(lambda x: x[1] - x[0] >= min_size, ranges))\n",
    "    return ranges\n",
    "\n",
    "def call_sites_from_fasta(fasta_file, kmers):\n",
    "    sites = []\n",
    "    with open(fasta_file) as file_obj:\n",
    "        for record in SeqIO.parse(file_obj, \"fasta\"):\n",
    "            site_ranges = call_nonconsensus_sites(str(record.seq), kmers)\n",
    "            genomic_coords = record.description\n",
    "            chromosome = genomic_coords.split(\":\")[0]\n",
    "            start = int(genomic_coords.split(\":\")[1].split(\"-\")[0])\n",
    "            name = name_from_label(record.id)\n",
    "            sites = sites + ranges_to_bed(site_ranges, chromosome, start, name)\n",
    "    sites_df = pd.DataFrame(sites)\n",
    "    return sites_df\n",
    "\n",
    "def call_sites_pipeline(fasta_file,\n",
    "                        kmer_dict,\n",
    "                        tf,\n",
    "                        output_name):\n",
    "    sites = call_sites_from_fasta(fasta_file, kmer_dict[tf])\n",
    "    sites = sites.rename(columns={0: \"Chromosome\", 1: \"Start\", 2: \"End\", 3: \"Name\"})\n",
    "    output_file = f\"{OUTPUT_SX}/{output_name}\"\n",
    "    sites.to_csv(output_file, sep='\\t', header=None, index=False)\n",
    "    \n",
    "def longest_ncsites_from_calls(nc_call_file, tf):\n",
    "    promoter_ncsite_dict = defaultdict(int)\n",
    "    with open(nc_call_file) as file_obj:\n",
    "        for i in file_obj:\n",
    "            c, s, e, promoter = i.strip().split()\n",
    "            site_len = int(e) - int(s)\n",
    "            if site_len > promoter_ncsite_dict[promoter]:\n",
    "                promoter_ncsite_dict[promoter] = site_len\n",
    "    site_df = pd.DataFrame.from_dict(promoter_ncsite_dict, orient='index')\n",
    "    site_df = site_df.reset_index().rename(columns={\"index\":\"Promoter\", 0:f\"Longest_{tf}_NCSite\"})\n",
    "    return site_df\n",
    "\n",
    "\n",
    "def unique_calls_bedfmt_from_fimo(fimo_file, output):\n",
    "    df = pd.read_csv(fimo_file, sep='\\t')\n",
    "    df = df[[\"sequence_name\", \"start\", \"stop\"]].drop_duplicates()\n",
    "    df.to_csv(output,\n",
    "              sep='\\t',\n",
    "              header=None,\n",
    "              index=False)\n",
    "    \n",
    "def genomic_coord_list_from_bed(bed_file):\n",
    "    genomic_coord_list = []\n",
    "    with open(bed_file) as read_obj:\n",
    "        for line in read_obj:\n",
    "            chrom, start, end = line.strip().split()\n",
    "            genomic_coord_list.append(f\"{chrom}:{start}-{end}\")\n",
    "    return genomic_coord_list\n",
    "\n",
    "\n",
    "def add_kmer_positions(string, k, label, to_dict):\n",
    "    \"\"\"Get indexes of k-mer matches to a string.\n",
    "    \n",
    "    Given a string and set of k-mers, return a list of start\n",
    "    indexes for matches.\n",
    "    \"\"\"\n",
    "    string = string.upper()\n",
    "    for idx in range(len(string) - k + 1):\n",
    "        kmer = string[idx:idx+k] \n",
    "        to_dict[kmer].append((label, idx))\n",
    "    \n",
    "def mutation_rate(mutation_count_file):\n",
    "    total_len = 0\n",
    "    mutation_count = 0\n",
    "    with open(mutation_count_file) as read_obj:\n",
    "        for line in read_obj:\n",
    "            chrom, start, end, count = line.strip().split('\\t')\n",
    "            site_len = int(end) - int(start)\n",
    "            total_len += site_len\n",
    "            mutation_count += int(count)\n",
    "    return mutation_count / total_len\n",
    "\n",
    "\n",
    "def mutation_rate_parts(mutation_count_file):\n",
    "    total_len = 0\n",
    "    mutation_count = 0\n",
    "    with open(mutation_count_file) as read_obj:\n",
    "        for line in read_obj:\n",
    "            chrom, start, end, count = line.strip().split('\\t')\n",
    "            site_len = int(end) - int(start)\n",
    "            total_len += site_len\n",
    "            mutation_count += int(count)\n",
    "    return (mutation_count, total_len)\n",
    "\n",
    "\n",
    "def aggregate_bkg_mutation_rate_df(N_bkg,\n",
    "                                   background_intersection_folder=BACKGROUND_INTERSECTION_DIR,\n",
    "                                   label_0base=True,\n",
    "                                   tf_tuple=TF_TUP,\n",
    "                                   min_nc_lengths=NC_LENGTH_TUP):\n",
    "    N = N_bkg\n",
    "    S = 0\n",
    "    if label_0base == False:\n",
    "        N += 1\n",
    "        S += 1\n",
    "    mr_dict = defaultdict(list)\n",
    "    for tf in TF_TUP:\n",
    "        for min_length in NC_LENGTH_TUP:\n",
    "            for i in range(S, N):\n",
    "                mr = mutation_rate(f\"../Figure_SX/Background_Intersection/NC_{tf}_{min_length}_Counts_{i}.bed\")\n",
    "                mr_dict[f\"{tf}_{min_length}\"].append(mr)\n",
    "        for i in range(S, N):\n",
    "            mr = mutation_rate(f\"../Figure_SX/Background_Intersection/{tf}_PWM_Background_Counts_{i}.bed\")\n",
    "            mr_dict[f\"{tf}_pwm\"].append(mr)\n",
    "    # Organize DataFrame of background mutation counts\n",
    "    background_df = pd.DataFrame.from_dict(mr_dict, orient='index').T\n",
    "    return background_df.melt()\n",
    "\n",
    "\n",
    "def measured_mutation_rate_df(tf_tuple=TF_TUP):\n",
    "    mutation_df = []\n",
    "    for tf in tf_tuple:\n",
    "        for min_length in NC_LENGTH_TUP:\n",
    "            mr = mutation_rate(f\"../Figure_SX/NonConsensus_{tf}_in_Promoters_{min_length}_Unique_Mutation_Counts.bed\")\n",
    "            mutation_df.append((f\"{tf}_{min_length}\", mr))\n",
    "        mr = mutation_rate(f\"../Figure_SX/{tf}_PWM_Promoter_Unique_Mutation_Counts.bed\")\n",
    "        mutation_df.append((f\"{tf}_pwm\", mr))\n",
    "    mutation_df = pd.DataFrame(mutation_df).rename(columns={0:\"variable\", 1:\"value\"})    \n",
    "    return mutation_df\n",
    "\n",
    "def plot_mutation_rate(background_mr_df, measured_mr_df):\n",
    "    sns.stripplot(data=background_mr_df, x=\"variable\", y=\"value\", size=3, color='black')\n",
    "    sns.stripplot(data=measured_mr_df, x=\"variable\", y=\"value\", size=6, color='red')\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.xlabel(xlabel='')\n",
    "    plt.ylabel(ylabel='Mutation Rate (sum mutations / total length)')\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ea7025",
   "metadata": {},
   "source": [
    "#### (2) Read UV_Preferred K-mers and organize as a dictionary of sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "74089436",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input files\n",
    "kmer_dict_files = ((\"CREB1\",f\"{KMER_6MER}/CREB1_0_1/OLS_CREB1_0_1_Escore.txt\"),\n",
    "                   (\"EGR1\", f\"{KMER_6MER}/EGR1_2_3/OLS_EGR1_2_3_Escore.txt\"))\n",
    "# From input files, return a dictionary where the key is the TF and value is a set of UV-preferred kmers\n",
    "kmer_dict = {}\n",
    "for tf, file in kmer_dict_files:\n",
    "    pref_kmers = uv_preferred_kmer_set(file)\n",
    "    kmer_dict[tf] = pref_kmers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8666377b",
   "metadata": {},
   "source": [
    "# UCSC File Downloads and Processing\n",
    "\n",
    "**Note:** The file for publication was downloaded on: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4a09d06f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2023-01-12 15:11:53--  https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/upstream2000.fa.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 19674496 (19M) [application/x-gzip]\n",
      "Saving to: ‘../../Data/External_Data/upstream2000.fa.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  0%  325K 59s\n",
      "    50K .......... .......... .......... .......... ..........  0% 49.8M 30s\n",
      "   100K .......... .......... .......... .......... ..........  0%  653K 29s\n",
      "   150K .......... .......... .......... .......... ..........  1%  382M 22s\n",
      "   200K .......... .......... .......... .......... ..........  1%  689K 23s\n",
      "   250K .......... .......... .......... .......... ..........  1% 7.76M 20s\n",
      "   300K .......... .......... .......... .......... ..........  1%  122M 17s\n",
      "   350K .......... .......... .......... .......... ..........  2% 10.9M 15s\n",
      "   400K .......... .......... .......... .......... ..........  2%  731K 16s\n",
      "   450K .......... .......... .......... .......... ..........  2%  194M 14s\n",
      "   500K .......... .......... .......... .......... ..........  2% 7.91M 13s\n",
      "   550K .......... .......... .......... .......... ..........  3%  128M 12s\n",
      "   600K .......... .......... .......... .......... ..........  3% 8.44M 11s\n",
      "   650K .......... .......... .......... .......... ..........  3%  233M 10s\n",
      "   700K .......... .......... .......... .......... ..........  3% 7.57M 10s\n",
      "   750K .......... .......... .......... .......... ..........  4%  236M 9s\n",
      "   800K .......... .......... .......... .......... ..........  4%  237M 9s\n",
      "   850K .......... .......... .......... .......... ..........  4%  701K 10s\n",
      "   900K .......... .......... .......... .......... ..........  4%  277M 9s\n",
      "   950K .......... .......... .......... .......... ..........  5% 7.64M 9s\n",
      "  1000K .......... .......... .......... .......... ..........  5%  290M 8s\n",
      "  1050K .......... .......... .......... .......... ..........  5% 7.76M 8s\n",
      "  1100K .......... .......... .......... .......... ..........  5%  233M 8s\n",
      "  1150K .......... .......... .......... .......... ..........  6%  330K 10s\n",
      "  1200K .......... .......... .......... .......... ..........  6% 74.0M 9s\n",
      "  1250K .......... .......... .......... .......... ..........  6% 9.34M 9s\n",
      "  1300K .......... .......... .......... .......... ..........  7% 73.1M 9s\n",
      "  1350K .......... .......... .......... .......... ..........  7% 90.4M 8s\n",
      "  1400K .......... .......... .......... .......... ..........  7%  119M 8s\n",
      "  1450K .......... .......... .......... .......... ..........  7%  106M 8s\n",
      "  1500K .......... .......... .......... .......... ..........  8%  116M 7s\n",
      "  1550K .......... .......... .......... .......... ..........  8%  111M 7s\n",
      "  1600K .......... .......... .......... .......... ..........  8% 71.2M 7s\n",
      "  1650K .......... .......... .......... .......... ..........  8%  116M 7s\n",
      "  1700K .......... .......... .......... .......... ..........  9% 97.7M 6s\n",
      "  1750K .......... .......... .......... .......... ..........  9%  113M 6s\n",
      "  1800K .......... .......... .......... .......... ..........  9%  128M 6s\n",
      "  1850K .......... .......... .......... .......... ..........  9%  108M 6s\n",
      "  1900K .......... .......... .......... .......... .......... 10%  118M 6s\n",
      "  1950K .......... .......... .......... .......... .......... 10%  109M 6s\n",
      "  2000K .......... .......... .......... .......... .......... 10%  118M 5s\n",
      "  2050K .......... .......... .......... .......... .......... 10%  110M 5s\n",
      "  2100K .......... .......... .......... .......... .......... 11% 81.0M 5s\n",
      "  2150K .......... .......... .......... .......... .......... 11% 47.5M 5s\n",
      "  2200K .......... .......... .......... .......... .......... 11% 1.90M 5s\n",
      "  2250K .......... .......... .......... .......... .......... 11% 95.2M 5s\n",
      "  2300K .......... .......... .......... .......... .......... 12% 3.17M 5s\n",
      "  2350K .......... .......... .......... .......... .......... 12%  204M 5s\n",
      "  2400K .......... .......... .......... .......... .......... 12% 1.62M 5s\n",
      "  2450K .......... .......... .......... .......... .......... 13% 65.4M 5s\n",
      "  2500K .......... .......... .......... .......... .......... 13% 9.32M 5s\n",
      "  2550K .......... .......... .......... .......... .......... 13%  110M 5s\n",
      "  2600K .......... .......... .......... .......... .......... 13% 9.39M 5s\n",
      "  2650K .......... .......... .......... .......... .......... 14% 83.6M 5s\n",
      "  2700K .......... .......... .......... .......... .......... 14% 2.16M 5s\n",
      "  2750K .......... .......... .......... .......... .......... 14% 70.9M 4s\n",
      "  2800K .......... .......... .......... .......... .......... 14% 5.92M 4s\n",
      "  2850K .......... .......... .......... .......... .......... 15% 92.2M 4s\n",
      "  2900K .......... .......... .......... .......... .......... 15% 1.17M 4s\n",
      "  2950K .......... .......... .......... .......... .......... 15%  306M 4s\n",
      "  3000K .......... .......... .......... .......... .......... 15% 7.48M 4s\n",
      "  3050K .......... .......... .......... .......... .......... 16%  326M 4s\n",
      "  3100K .......... .......... .......... .......... .......... 16% 7.39M 4s\n",
      "  3150K .......... .......... .......... .......... .......... 16%  252M 4s\n",
      "  3200K .......... .......... .......... .......... .......... 16% 3.19M 4s\n",
      "  3250K .......... .......... .......... .......... .......... 17%  182M 4s\n",
      "  3300K .......... .......... .......... .......... .......... 17%  229M 4s\n",
      "  3350K .......... .......... .......... .......... .......... 17% 13.8M 4s\n",
      "  3400K .......... .......... .......... .......... .......... 17%  868K 4s\n",
      "  3450K .......... .......... .......... .......... .......... 18%  103M 4s\n",
      "  3500K .......... .......... .......... .......... .......... 18% 8.48M 4s\n",
      "  3550K .......... .......... .......... .......... .......... 18% 98.4M 4s\n",
      "  3600K .......... .......... .......... .......... .......... 18% 46.6M 4s\n",
      "  3650K .......... .......... .......... .......... .......... 19% 9.66M 4s\n",
      "  3700K .......... .......... .......... .......... .......... 19% 94.2M 4s\n",
      "  3750K .......... .......... .......... .......... .......... 19% 2.48M 4s\n",
      "  3800K .......... .......... .......... .......... .......... 20%  169M 4s\n",
      "  3850K .......... .......... .......... .......... .......... 20% 12.6M 4s\n",
      "  3900K .......... .......... .......... .......... .......... 20%  188M 4s\n",
      "  3950K .......... .......... .......... .......... .......... 20%  847K 4s\n",
      "  4000K .......... .......... .......... .......... .......... 21%  119M 4s\n",
      "  4050K .......... .......... .......... .......... .......... 21% 7.42M 4s\n",
      "  4100K .......... .......... .......... .......... .......... 21% 73.7M 4s\n",
      "  4150K .......... .......... .......... .......... .......... 21% 10.4M 4s\n",
      "  4200K .......... .......... .......... .......... .......... 22% 90.7M 4s\n",
      "  4250K .......... .......... .......... .......... .......... 22% 9.17M 4s\n",
      "  4300K .......... .......... .......... .......... .......... 22%  104M 3s\n",
      "  4350K .......... .......... .......... .......... .......... 22% 10.6M 3s\n",
      "  4400K .......... .......... .......... .......... .......... 23%  101M 3s\n",
      "  4450K .......... .......... .......... .......... .......... 23% 1.17M 3s\n",
      "  4500K .......... .......... .......... .......... .......... 23% 4.26M 3s\n",
      "  4550K .......... .......... .......... .......... .......... 23%  181M 3s\n",
      "  4600K .......... .......... .......... .......... .......... 24% 2.22M 3s\n",
      "  4650K .......... .......... .......... .......... .......... 24% 7.45M 3s\n",
      "  4700K .......... .......... .......... .......... .......... 24%  168M 3s\n",
      "  4750K .......... .......... .......... .......... .......... 24% 8.55M 3s\n",
      "  4800K .......... .......... .......... .......... .......... 25%  164M 3s\n",
      "  4850K .......... .......... .......... .......... .......... 25% 8.09M 3s\n",
      "  4900K .......... .......... .......... .......... .......... 25%  209M 3s\n",
      "  4950K .......... .......... .......... .......... .......... 26%  263M 3s\n",
      "  5000K .......... .......... .......... .......... .......... 26% 1.65M 3s\n",
      "  5050K .......... .......... .......... .......... .......... 26% 9.61M 3s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5100K .......... .......... .......... .......... .......... 26% 2.67M 3s\n",
      "  5150K .......... .......... .......... .......... .......... 27% 2.44M 3s\n",
      "  5200K .......... .......... .......... .......... .......... 27% 7.34M 3s\n",
      "  5250K .......... .......... .......... .......... .......... 27%  235M 3s\n",
      "  5300K .......... .......... .......... .......... .......... 27% 7.83M 3s\n",
      "  5350K .......... .......... .......... .......... .......... 28% 7.21M 3s\n",
      "  5400K .......... .......... .......... .......... .......... 28%  158M 3s\n",
      "  5450K .......... .......... .......... .......... .......... 28% 5.06M 3s\n",
      "  5500K .......... .......... .......... .......... .......... 28%  250M 3s\n",
      "  5550K .......... .......... .......... .......... .......... 29% 4.51M 3s\n",
      "  5600K .......... .......... .......... .......... .......... 29%  171M 3s\n",
      "  5650K .......... .......... .......... .......... .......... 29% 4.66M 3s\n",
      "  5700K .......... .......... .......... .......... .......... 29%  641K 3s\n",
      "  5750K .......... .......... .......... .......... .......... 30%  115M 3s\n",
      "  5800K .......... .......... .......... .......... .......... 30% 8.75M 3s\n",
      "  5850K .......... .......... .......... .......... .......... 30% 64.6M 3s\n",
      "  5900K .......... .......... .......... .......... .......... 30% 8.50M 3s\n",
      "  5950K .......... .......... .......... .......... .......... 31% 67.6M 3s\n",
      "  6000K .......... .......... .......... .......... .......... 31% 8.17M 3s\n",
      "  6050K .......... .......... .......... .......... .......... 31%  120M 3s\n",
      "  6100K .......... .......... .......... .......... .......... 32% 7.85M 3s\n",
      "  6150K .......... .......... .......... .......... .......... 32% 77.4M 3s\n",
      "  6200K .......... .......... .......... .......... .......... 32% 1000K 3s\n",
      "  6250K .......... .......... .......... .......... .......... 32% 7.40M 3s\n",
      "  6300K .......... .......... .......... .......... .......... 33%  148M 3s\n",
      "  6350K .......... .......... .......... .......... .......... 33% 8.13M 3s\n",
      "  6400K .......... .......... .......... .......... .......... 33%  201M 3s\n",
      "  6450K .......... .......... .......... .......... .......... 33% 6.21M 3s\n",
      "  6500K .......... .......... .......... .......... .......... 34%  114M 3s\n",
      "  6550K .......... .......... .......... .......... .......... 34% 5.40M 3s\n",
      "  6600K .......... .......... .......... .......... .......... 34% 63.9M 3s\n",
      "  6650K .......... .......... .......... .......... .......... 34% 8.49M 3s\n",
      "  6700K .......... .......... .......... .......... .......... 35% 55.2M 3s\n",
      "  6750K .......... .......... .......... .......... .......... 35%  121M 3s\n",
      "  6800K .......... .......... .......... .......... .......... 35% 1.00M 3s\n",
      "  6850K .......... .......... .......... .......... .......... 35% 5.17M 3s\n",
      "  6900K .......... .......... .......... .......... .......... 36% 7.53M 3s\n",
      "  6950K .......... .......... .......... .......... .......... 36% 87.7M 3s\n",
      "  7000K .......... .......... .......... .......... .......... 36%  175M 3s\n",
      "  7050K .......... .......... .......... .......... .......... 36% 3.21M 3s\n",
      "  7100K .......... .......... .......... .......... .......... 37%  133M 3s\n",
      "  7150K .......... .......... .......... .......... .......... 37% 7.11M 3s\n",
      "  7200K .......... .......... .......... .......... .......... 37% 7.75M 3s\n",
      "  7250K .......... .......... .......... .......... .......... 37%  109M 2s\n",
      "  7300K .......... .......... .......... .......... .......... 38% 13.3M 2s\n",
      "  7350K .......... .......... .......... .......... .......... 38% 1.26M 3s\n",
      "  7400K .......... .......... .......... .......... .......... 38%  119M 2s\n",
      "  7450K .......... .......... .......... .......... .......... 39% 6.43M 2s\n",
      "  7500K .......... .......... .......... .......... .......... 39%  101M 2s\n",
      "  7550K .......... .......... .......... .......... .......... 39% 9.15M 2s\n",
      "  7600K .......... .......... .......... .......... .......... 39% 2.40M 2s\n",
      "  7650K .......... .......... .......... .......... .......... 40% 7.28M 2s\n",
      "  7700K .......... .......... .......... .......... .......... 40% 91.5M 2s\n",
      "  7750K .......... .......... .......... .......... .......... 40% 7.86M 2s\n",
      "  7800K .......... .......... .......... .......... .......... 40% 99.3M 2s\n",
      "  7850K .......... .......... .......... .......... .......... 41% 11.6M 2s\n",
      "  7900K .......... .......... .......... .......... .......... 41% 2.31M 2s\n",
      "  7950K .......... .......... .......... .......... .......... 41% 28.9M 2s\n",
      "  8000K .......... .......... .......... .......... .......... 41% 10.8M 2s\n",
      "  8050K .......... .......... .......... .......... .......... 42% 10.5M 2s\n",
      "  8100K .......... .......... .......... .......... .......... 42% 15.3M 2s\n",
      "  8150K .......... .......... .......... .......... .......... 42% 2.56M 2s\n",
      "  8200K .......... .......... .......... .......... .......... 42% 4.81M 2s\n",
      "  8250K .......... .......... .......... .......... .......... 43%  234M 2s\n",
      "  8300K .......... .......... .......... .......... .......... 43% 10.3M 2s\n",
      "  8350K .......... .......... .......... .......... .......... 43% 94.2M 2s\n",
      "  8400K .......... .......... .......... .......... .......... 43% 7.35M 2s\n",
      "  8450K .......... .......... .......... .......... .......... 44%  105M 2s\n",
      "  8500K .......... .......... .......... .......... .......... 44% 2.36M 2s\n",
      "  8550K .......... .......... .......... .......... .......... 44% 52.3M 2s\n",
      "  8600K .......... .......... .......... .......... .......... 45% 3.98M 2s\n",
      "  8650K .......... .......... .......... .......... .......... 45%  120M 2s\n",
      "  8700K .......... .......... .......... .......... .......... 45% 3.50M 2s\n",
      "  8750K .......... .......... .......... .......... .......... 45% 9.36M 2s\n",
      "  8800K .......... .......... .......... .......... .......... 46% 4.73M 2s\n",
      "  8850K .......... .......... .......... .......... .......... 46% 39.4M 2s\n",
      "  8900K .......... .......... .......... .......... .......... 46% 8.20M 2s\n",
      "  8950K .......... .......... .......... .......... .......... 46% 5.02M 2s\n",
      "  9000K .......... .......... .......... .......... .......... 47%  175M 2s\n",
      "  9050K .......... .......... .......... .......... .......... 47% 9.81M 2s\n",
      "  9100K .......... .......... .......... .......... .......... 47% 3.40M 2s\n",
      "  9150K .......... .......... .......... .......... .......... 47%  156M 2s\n",
      "  9200K .......... .......... .......... .......... .......... 48% 5.31M 2s\n",
      "  9250K .......... .......... .......... .......... .......... 48%  178M 2s\n",
      "  9300K .......... .......... .......... .......... .......... 48% 2.98M 2s\n",
      "  9350K .......... .......... .......... .......... .......... 48% 14.3M 2s\n",
      "  9400K .......... .......... .......... .......... .......... 49% 5.47M 2s\n",
      "  9450K .......... .......... .......... .......... .......... 49% 6.70M 2s\n",
      "  9500K .......... .......... .......... .......... .......... 49%  194M 2s\n",
      "  9550K .......... .......... .......... .......... .......... 49% 5.01M 2s\n",
      "  9600K .......... .......... .......... .......... .......... 50% 10.7M 2s\n",
      "  9650K .......... .......... .......... .......... .......... 50% 4.27M 2s\n",
      "  9700K .......... .......... .......... .......... .......... 50% 10.7M 2s\n",
      "  9750K .......... .......... .......... .......... .......... 51% 36.6M 2s\n",
      "  9800K .......... .......... .......... .......... .......... 51% 6.73M 2s\n",
      "  9850K .......... .......... .......... .......... .......... 51% 17.8M 2s\n",
      "  9900K .......... .......... .......... .......... .......... 51% 4.78M 2s\n",
      "  9950K .......... .......... .......... .......... .......... 52% 3.86M 2s\n",
      " 10000K .......... .......... .......... .......... .......... 52% 4.67M 2s\n",
      " 10050K .......... .......... .......... .......... .......... 52%  141M 2s\n",
      " 10100K .......... .......... .......... .......... .......... 52% 5.50M 2s\n",
      " 10150K .......... .......... .......... .......... .......... 53% 8.33M 2s\n",
      " 10200K .......... .......... .......... .......... .......... 53%  155M 2s\n",
      " 10250K .......... .......... .......... .......... .......... 53% 3.18M 2s\n",
      " 10300K .......... .......... .......... .......... .......... 53%  212M 2s\n",
      " 10350K .......... .......... .......... .......... .......... 54% 6.40M 2s\n",
      " 10400K .......... .......... .......... .......... .......... 54% 12.2M 2s\n",
      " 10450K .......... .......... .......... .......... .......... 54% 12.5M 2s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 10500K .......... .......... .......... .......... .......... 54% 10.3M 2s\n",
      " 10550K .......... .......... .......... .......... .......... 55% 4.06M 2s\n",
      " 10600K .......... .......... .......... .......... .......... 55% 3.77M 2s\n",
      " 10650K .......... .......... .......... .......... .......... 55% 24.2M 2s\n",
      " 10700K .......... .......... .......... .......... .......... 55% 12.4M 2s\n",
      " 10750K .......... .......... .......... .......... .......... 56% 5.82M 2s\n",
      " 10800K .......... .......... .......... .......... .......... 56%  237M 2s\n",
      " 10850K .......... .......... .......... .......... .......... 56% 3.00M 2s\n",
      " 10900K .......... .......... .......... .......... .......... 56%  138M 2s\n",
      " 10950K .......... .......... .......... .......... .......... 57% 10.3M 2s\n",
      " 11000K .......... .......... .......... .......... .......... 57% 11.5M 2s\n",
      " 11050K .......... .......... .......... .......... .......... 57% 15.5M 1s\n",
      " 11100K .......... .......... .......... .......... .......... 58% 7.58M 1s\n",
      " 11150K .......... .......... .......... .......... .......... 58% 4.65M 1s\n",
      " 11200K .......... .......... .......... .......... .......... 58% 2.79M 1s\n",
      " 11250K .......... .......... .......... .......... .......... 58%  141M 1s\n",
      " 11300K .......... .......... .......... .......... .......... 59% 16.5M 1s\n",
      " 11350K .......... .......... .......... .......... .......... 59% 5.76M 1s\n",
      " 11400K .......... .......... .......... .......... .......... 59% 33.4M 1s\n",
      " 11450K .......... .......... .......... .......... .......... 59% 3.16M 1s\n",
      " 11500K .......... .......... .......... .......... .......... 60% 7.18M 1s\n",
      " 11550K .......... .......... .......... .......... .......... 60% 13.0M 1s\n",
      " 11600K .......... .......... .......... .......... .......... 60%  198M 1s\n",
      " 11650K .......... .......... .......... .......... .......... 60% 18.8M 1s\n",
      " 11700K .......... .......... .......... .......... .......... 61% 10.0M 1s\n",
      " 11750K .......... .......... .......... .......... .......... 61% 5.69M 1s\n",
      " 11800K .......... .......... .......... .......... .......... 61% 1.98M 1s\n",
      " 11850K .......... .......... .......... .......... .......... 61%  143M 1s\n",
      " 11900K .......... .......... .......... .......... .......... 62%  172M 1s\n",
      " 11950K .......... .......... .......... .......... .......... 62% 7.51M 1s\n",
      " 12000K .......... .......... .......... .......... .......... 62% 51.3M 1s\n",
      " 12050K .......... .......... .......... .......... .......... 62% 2.74M 1s\n",
      " 12100K .......... .......... .......... .......... .......... 63% 14.4M 1s\n",
      " 12150K .......... .......... .......... .......... .......... 63% 12.5M 1s\n",
      " 12200K .......... .......... .......... .......... .......... 63% 5.34M 1s\n",
      " 12250K .......... .......... .......... .......... .......... 64%  116M 1s\n",
      " 12300K .......... .......... .......... .......... .......... 64%  150M 1s\n",
      " 12350K .......... .......... .......... .......... .......... 64% 10.6M 1s\n",
      " 12400K .......... .......... .......... .......... .......... 64% 2.01M 1s\n",
      " 12450K .......... .......... .......... .......... .......... 65% 8.31M 1s\n",
      " 12500K .......... .......... .......... .......... .......... 65% 10.2M 1s\n",
      " 12550K .......... .......... .......... .......... .......... 65%  203M 1s\n",
      " 12600K .......... .......... .......... .......... .......... 65% 10.9M 1s\n",
      " 12650K .......... .......... .......... .......... .......... 66% 4.29M 1s\n",
      " 12700K .......... .......... .......... .......... .......... 66% 16.4M 1s\n",
      " 12750K .......... .......... .......... .......... .......... 66% 5.58M 1s\n",
      " 12800K .......... .......... .......... .......... .......... 66% 12.0M 1s\n",
      " 12850K .......... .......... .......... .......... .......... 67% 4.48M 1s\n",
      " 12900K .......... .......... .......... .......... .......... 67%  315M 1s\n",
      " 12950K .......... .......... .......... .......... .......... 67%  293M 1s\n",
      " 13000K .......... .......... .......... .......... .......... 67% 5.05M 1s\n",
      " 13050K .......... .......... .......... .......... .......... 68% 2.78M 1s\n",
      " 13100K .......... .......... .......... .......... .......... 68% 10.9M 1s\n",
      " 13150K .......... .......... .......... .......... .......... 68%  329M 1s\n",
      " 13200K .......... .......... .......... .......... .......... 68% 14.8M 1s\n",
      " 13250K .......... .......... .......... .......... .......... 69% 3.59M 1s\n",
      " 13300K .......... .......... .......... .......... .......... 69% 22.1M 1s\n",
      " 13350K .......... .......... .......... .......... .......... 69% 5.94M 1s\n",
      " 13400K .......... .......... .......... .......... .......... 70% 15.2M 1s\n",
      " 13450K .......... .......... .......... .......... .......... 70% 3.68M 1s\n",
      " 13500K .......... .......... .......... .......... .......... 70% 84.4M 1s\n",
      " 13550K .......... .......... .......... .......... .......... 70% 10.3M 1s\n",
      " 13600K .......... .......... .......... .......... .......... 71%  104M 1s\n",
      " 13650K .......... .......... .......... .......... .......... 71% 2.40M 1s\n",
      " 13700K .......... .......... .......... .......... .......... 71% 5.88M 1s\n",
      " 13750K .......... .......... .......... .......... .......... 71%  268M 1s\n",
      " 13800K .......... .......... .......... .......... .......... 72% 8.97M 1s\n",
      " 13850K .......... .......... .......... .......... .......... 72% 13.4M 1s\n",
      " 13900K .......... .......... .......... .......... .......... 72% 7.93M 1s\n",
      " 13950K .......... .......... .......... .......... .......... 72% 7.92M 1s\n",
      " 14000K .......... .......... .......... .......... .......... 73% 14.9M 1s\n",
      " 14050K .......... .......... .......... .......... .......... 73% 7.72M 1s\n",
      " 14100K .......... .......... .......... .......... .......... 73% 5.15M 1s\n",
      " 14150K .......... .......... .......... .......... .......... 73% 10.2M 1s\n",
      " 14200K .......... .......... .......... .......... .......... 74% 8.28M 1s\n",
      " 14250K .......... .......... .......... .......... .......... 74% 6.87M 1s\n",
      " 14300K .......... .......... .......... .......... .......... 74% 6.27M 1s\n",
      " 14350K .......... .......... .......... .......... .......... 74% 6.03M 1s\n",
      " 14400K .......... .......... .......... .......... .......... 75% 7.14M 1s\n",
      " 14450K .......... .......... .......... .......... .......... 75% 33.8M 1s\n",
      " 14500K .......... .......... .......... .......... .......... 75% 7.22M 1s\n",
      " 14550K .......... .......... .......... .......... .......... 75% 8.09M 1s\n",
      " 14600K .......... .......... .......... .......... .......... 76% 19.2M 1s\n",
      " 14650K .......... .......... .......... .......... .......... 76% 10.9M 1s\n",
      " 14700K .......... .......... .......... .......... .......... 76% 2.96M 1s\n",
      " 14750K .......... .......... .......... .......... .......... 77% 9.28M 1s\n",
      " 14800K .......... .......... .......... .......... .......... 77%  278M 1s\n",
      " 14850K .......... .......... .......... .......... .......... 77% 18.0M 1s\n",
      " 14900K .......... .......... .......... .......... .......... 77% 3.61M 1s\n",
      " 14950K .......... .......... .......... .......... .......... 78% 14.1M 1s\n",
      " 15000K .......... .......... .......... .......... .......... 78% 4.98M 1s\n",
      " 15050K .......... .......... .......... .......... .......... 78% 9.45M 1s\n",
      " 15100K .......... .......... .......... .......... .......... 78% 11.6M 1s\n",
      " 15150K .......... .......... .......... .......... .......... 79% 6.46M 1s\n",
      " 15200K .......... .......... .......... .......... .......... 79%  249M 1s\n",
      " 15250K .......... .......... .......... .......... .......... 79% 7.94M 1s\n",
      " 15300K .......... .......... .......... .......... .......... 79% 6.95M 1s\n",
      " 15350K .......... .......... .......... .......... .......... 80% 3.65M 1s\n",
      " 15400K .......... .......... .......... .......... .......... 80% 16.2M 1s\n",
      " 15450K .......... .......... .......... .......... .......... 80% 9.88M 1s\n",
      " 15500K .......... .......... .......... .......... .......... 80% 52.5M 1s\n",
      " 15550K .......... .......... .......... .......... .......... 81% 5.96M 1s\n",
      " 15600K .......... .......... .......... .......... .......... 81% 6.33M 1s\n",
      " 15650K .......... .......... .......... .......... .......... 81% 5.15M 1s\n",
      " 15700K .......... .......... .......... .......... .......... 81% 7.29M 1s\n",
      " 15750K .......... .......... .......... .......... .......... 82%  110M 1s\n",
      " 15800K .......... .......... .......... .......... .......... 82% 5.72M 1s\n",
      " 15850K .......... .......... .......... .......... .......... 82% 5.27M 1s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 15900K .......... .......... .......... .......... .......... 83%  181M 1s\n",
      " 15950K .......... .......... .......... .......... .......... 83% 5.74M 1s\n",
      " 16000K .......... .......... .......... .......... .......... 83% 3.79M 1s\n",
      " 16050K .......... .......... .......... .......... .......... 83% 4.85M 1s\n",
      " 16100K .......... .......... .......... .......... .......... 84%  140M 1s\n",
      " 16150K .......... .......... .......... .......... .......... 84%  181M 0s\n",
      " 16200K .......... .......... .......... .......... .......... 84% 11.7M 0s\n",
      " 16250K .......... .......... .......... .......... .......... 84% 4.33M 0s\n",
      " 16300K .......... .......... .......... .......... .......... 85% 15.1M 0s\n",
      " 16350K .......... .......... .......... .......... .......... 85% 8.55M 0s\n",
      " 16400K .......... .......... .......... .......... .......... 85% 2.11M 0s\n",
      " 16450K .......... .......... .......... .......... .......... 85%  124M 0s\n",
      " 16500K .......... .......... .......... .......... .......... 86% 7.93M 0s\n",
      " 16550K .......... .......... .......... .......... .......... 86%  138M 0s\n",
      " 16600K .......... .......... .......... .......... .......... 86% 27.2M 0s\n",
      " 16650K .......... .......... .......... .......... .......... 86% 4.89M 0s\n",
      " 16700K .......... .......... .......... .......... .......... 87% 4.23M 0s\n",
      " 16750K .......... .......... .......... .......... .......... 87% 7.00M 0s\n",
      " 16800K .......... .......... .......... .......... .......... 87% 11.1M 0s\n",
      " 16850K .......... .......... .......... .......... .......... 87%  136M 0s\n",
      " 16900K .......... .......... .......... .......... .......... 88% 10.2M 0s\n",
      " 16950K .......... .......... .......... .......... .......... 88% 29.6M 0s\n",
      " 17000K .......... .......... .......... .......... .......... 88% 15.9M 0s\n",
      " 17050K .......... .......... .......... .......... .......... 89% 1.94M 0s\n",
      " 17100K .......... .......... .......... .......... .......... 89%  118M 0s\n",
      " 17150K .......... .......... .......... .......... .......... 89% 2.60M 0s\n",
      " 17200K .......... .......... .......... .......... .......... 89% 11.3M 0s\n",
      " 17250K .......... .......... .......... .......... .......... 90%  117M 0s\n",
      " 17300K .......... .......... .......... .......... .......... 90% 11.6M 0s\n",
      " 17350K .......... .......... .......... .......... .......... 90%  175M 0s\n",
      " 17400K .......... .......... .......... .......... .......... 90% 10.5M 0s\n",
      " 17450K .......... .......... .......... .......... .......... 91% 4.84M 0s\n",
      " 17500K .......... .......... .......... .......... .......... 91%  210M 0s\n",
      " 17550K .......... .......... .......... .......... .......... 91% 17.3M 0s\n",
      " 17600K .......... .......... .......... .......... .......... 91% 11.4M 0s\n",
      " 17650K .......... .......... .......... .......... .......... 92% 6.01M 0s\n",
      " 17700K .......... .......... .......... .......... .......... 92% 2.66M 0s\n",
      " 17750K .......... .......... .......... .......... .......... 92% 8.28M 0s\n",
      " 17800K .......... .......... .......... .......... .......... 92% 5.27M 0s\n",
      " 17850K .......... .......... .......... .......... .......... 93% 5.63M 0s\n",
      " 17900K .......... .......... .......... .......... .......... 93%  139M 0s\n",
      " 17950K .......... .......... .......... .......... .......... 93% 14.6M 0s\n",
      " 18000K .......... .......... .......... .......... .......... 93% 15.4M 0s\n",
      " 18050K .......... .......... .......... .......... .......... 94% 3.26M 0s\n",
      " 18100K .......... .......... .......... .......... .......... 94% 7.23M 0s\n",
      " 18150K .......... .......... .......... .......... .......... 94%  211M 0s\n",
      " 18200K .......... .......... .......... .......... .......... 94% 7.97M 0s\n",
      " 18250K .......... .......... .......... .......... .......... 95% 62.6M 0s\n",
      " 18300K .......... .......... .......... .......... .......... 95% 4.37M 0s\n",
      " 18350K .......... .......... .......... .......... .......... 95% 11.0M 0s\n",
      " 18400K .......... .......... .......... .......... .......... 96% 11.8M 0s\n",
      " 18450K .......... .......... .......... .......... .......... 96% 5.88M 0s\n",
      " 18500K .......... .......... .......... .......... .......... 96% 3.13M 0s\n",
      " 18550K .......... .......... .......... .......... .......... 96%  203M 0s\n",
      " 18600K .......... .......... .......... .......... .......... 97% 8.03M 0s\n",
      " 18650K .......... .......... .......... .......... .......... 97%  219M 0s\n",
      " 18700K .......... .......... .......... .......... .......... 97% 5.72M 0s\n",
      " 18750K .......... .......... .......... .......... .......... 97% 5.47M 0s\n",
      " 18800K .......... .......... .......... .......... .......... 98% 22.9M 0s\n",
      " 18850K .......... .......... .......... .......... .......... 98% 6.53M 0s\n",
      " 18900K .......... .......... .......... .......... .......... 98%  206M 0s\n",
      " 18950K .......... .......... .......... .......... .......... 98% 4.51M 0s\n",
      " 19000K .......... .......... .......... .......... .......... 99% 19.5M 0s\n",
      " 19050K .......... .......... .......... .......... .......... 99% 16.8M 0s\n",
      " 19100K .......... .......... .......... .......... .......... 99% 3.29M 0s\n",
      " 19150K .......... .......... .......... .......... .......... 99% 3.61M 0s\n",
      " 19200K .......... ...                                        100% 5.90M=3.1s\n",
      "\n",
      "2023-01-12 15:11:56 (6.12 MB/s) - ‘../../Data/External_Data/upstream2000.fa.gz’ saved [19674496/19674496]\n",
      "\n",
      "gzip: ../../Data/External_Data/upstream2000.fa already exists;\tnot overwritten\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\nwget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/upstream2000.fa.gz -O ../../Data/External_Data/upstream2000.fa.gz\\ngunzip ../../Data/External_Data/upstream2000.fa.gz\\n'' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mwget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/upstream2000.fa.gz -O ../../Data/External_Data/upstream2000.fa.gz\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mgunzip ../../Data/External_Data/upstream2000.fa.gz\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/UV_Project/uvbind_env/lib/python3.9/site-packages/IPython/core/interactiveshell.py:2347\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2345\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2346\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2347\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2348\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/Documents/UV_Project/uvbind_env/lib/python3.9/site-packages/IPython/core/magics/script.py:153\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    151\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    152\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/UV_Project/uvbind_env/lib/python3.9/site-packages/IPython/core/magics/script.py:305\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    301\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    302\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    303\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    304\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 305\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\nwget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/upstream2000.fa.gz -O ../../Data/External_Data/upstream2000.fa.gz\\ngunzip ../../Data/External_Data/upstream2000.fa.gz\\n'' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/upstream2000.fa.gz -O ../../Data/External_Data/upstream2000.fa.gz\n",
    "gunzip ../../Data/External_Data/upstream2000.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e078fa8",
   "metadata": {},
   "source": [
    "#### (4) Process the UCSC FASTA Promoter File\n",
    "\n",
    "\n",
    "The promoter fasta file has the sequences relative to the strand of the gene. For this analysis, having all sequences on the plus strand simplifies the process. Using the FASTA labels, a BED file is generates from the promoter FASTA file.\n",
    "\n",
    "**File Output:**\n",
    "\n",
    "1. **Promoter2000.bed** - Bed coordinates of upstream2000.fa from USCS\n",
    "2. **Promoter2000_Sorted.bed** - Sorted version of Promoter2000.bed \n",
    "3. **Promoter2000_Merged.bed** - Merged version of Promoter2000_Sorted.bed \n",
    "4. **Promoter2000.fasta** - Fasta file from Promoter2000.bed\n",
    "5. **Promoter2000_Merged.fasta** - Fasta file from Promoter2000_Merged.bed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cfbeaa56",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = []\n",
    "# Read promoter file and parse genomic coordinates\n",
    "with open(USCS_PROMOTER_FASTA_FILE) as file_obj:\n",
    "    for record in SeqIO.parse(file_obj, \"fasta\"):\n",
    "        genomic_coords = record.description.split()[1].strip()\n",
    "        if not any([i in genomic_coords for i in [\"alt\", \"fix\", \"random\", \"chrUn\", \"hap\"]]):\n",
    "            chrom = genomic_coords.split(\":\")[0]\n",
    "            start = int(genomic_coords.split(\":\")[1].split(\"-\")[0]) - 1\n",
    "            end = genomic_coords.split(\":\")[1].split(\"-\")[1]\n",
    "            bed.append((chrom, start, end))\n",
    "beddf = pd.DataFrame(bed)\n",
    "beddf = beddf.rename(columns={0:\"Chr\", 1:\"Start\", 2:\"End\"})\n",
    "beddf = beddf.drop_duplicates().reset_index(drop=True)\n",
    "beddf.to_csv(promoter_bed_file, sep=\"\\t\", header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94679bf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$promoter_bed_file\" \"$promoter_sorted_bed_file\" \"$promoter_merged_bed_file\"\n",
    "# Sort and merge the promoter file to create a merged promoter file with no overlaps\n",
    "bedtools sort -i $1 > $2\n",
    "bedtools merge -i $2 > $3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "027e70d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$HUMAN_GENOME_FILE\" \"$promoter_bed_file\" \"$promoter_fasta_file\" \"$promoter_merged_bed_file\" \"$promoter_fasta_merged_file\"\n",
    "# Run bedtools getfasta to convert the non-merged and merged promoters to fasta\n",
    "bedtools getfasta -fi $1 -bed $2 -fo $3\n",
    "bedtools getfasta -fi $1 -bed $4 -fo $5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99bb13f9",
   "metadata": {},
   "source": [
    "# Calling Binding Sites\n",
    "\n",
    "#### (6) Call Non-Consensus Sites in the Promoter\n",
    "\n",
    "**Output files:**\n",
    "1. **NonConsensus_{tf}_promoter_calls.bed** - non-consensus site calls at any length\n",
    "2. **NonConsensus_{tf}_promoter_calls_\\{min_length}.bed** - non-consensus site calls at least {min_length} long\n",
    "3. **NonConsensus_{tf}_in_Promoters_\\{min_length}_unique.bed** - unique non-consensus site calls at least {min_length} long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "640aac4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calling Non-Consensus sites for CREB1...\n",
      "Calling Non-Consensus sites for EGR1...\n"
     ]
    }
   ],
   "source": [
    "for tf in TF_TUP:\n",
    "    print(f\"Calling Non-Consensus sites for {tf}...\")\n",
    "    # Create non-consensus site calls at any length\n",
    "    call_sites_pipeline(promoter_fasta_file,\n",
    "                        kmer_dict,\n",
    "                        tf,\n",
    "                        f\"NonConsensus_{tf}_promoter_calls.bed\")\n",
    "    for min_length in NC_LENGTH_TUP:\n",
    "        # Create BED file with minimum k-mer runs\n",
    "        with open(f\"{OUTPUT_SX}/NonConsensus_{tf}_promoter_calls.bed\") as read_obj, \\\n",
    "             open(f\"{OUTPUT_SX}/NonConsensus_{tf}_promoter_calls_{min_length}.bed\", 'w') as write_obj:\n",
    "            for line in read_obj:\n",
    "                chrom, start, end, name = line.strip().split()\n",
    "                site_len = int(end) - int(start)\n",
    "                if site_len >= min_length:\n",
    "                    write_obj.write(f\"{chrom}\\t{start}\\t{end}\\n\")\n",
    "        # Create a BED file with unique sites\n",
    "        df = pd.read_csv(f\"{OUTPUT_SX}/NonConsensus_{tf}_promoter_calls_{min_length}.bed\",\n",
    "                         sep='\\t',\n",
    "                         header=None)\n",
    "        df = df.drop_duplicates().reset_index(drop=True)\n",
    "        df.to_csv(f\"{OUTPUT_SX}/NonConsensus_{tf}_promoter_calls_{min_length}_unique.bed\",\n",
    "                  sep='\\t',\n",
    "                  header=None,\n",
    "                  index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a256cdca",
   "metadata": {},
   "source": [
    "#### (7) Call Canonical Sites in the Promoter\n",
    "\n",
    "Call sites using the Non-UV PWMs for CREB1 and EGR1 to use as a comparison for the non-consensus sites.\n",
    "\n",
    "**Output files:**\n",
    "1. {protein}_NonUV_pwm.txt - PWMs for CREB1 and EGR1 in text format\n",
    "2. {protein}_NonUV_pwm.meme - PWMs for CREB1 and EGR1 in meme format\n",
    "3. {tf}_PWM_promoter_calls.txt - FIMO calls in tabular output\n",
    "4. {tf}_PWM_promoter_calls_unique.bed - Unique FIMO calls in BED format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4519b266",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query PWMs from Supplementary Table 3A and save to files\n",
    "pwm_df = pd.read_csv(PWMS_TABLE_S3A)\n",
    "pwm_percentages = []\n",
    "for protein in TF_TUP:\n",
    "    pwm = pwm_df[pwm_df[\"Experiment\"] == f\"{protein}_NonUV\"]\n",
    "    # Transpose to wide format and convert to numpy array\n",
    "    pwm = pwm[[\"A\", \"C\", \"G\", \"T\"]]\n",
    "    pwm = pwm.rename(columns={\"A\":\"A:\", \"C\":\"C:\", \"G\":\"G:\",\"T\":\"T:\"})\n",
    "    pwm = pwm.T\n",
    "    if PWM_CORE_ONLY:\n",
    "        pwm = pwm[pwm.columns[3:-3]]\n",
    "    with open(f\"{OUTPUT_SX}/{protein}_NonUV_pwm.txt\", 'w') as write_obj:\n",
    "        write_obj.write(f\"{protein}\\n\")\n",
    "    pwm.to_csv(f\"{OUTPUT_SX}/{protein}_NonUV_pwm.txt\", sep='\\t', header=None, mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "348ce8c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +CREB1 of width 14.\n",
      "Using motif -CREB1 of width 14.\n",
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +EGR1 of width 15.\n",
      "Using motif -EGR1 of width 15.\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$promoter_fasta_file\"\n",
    "\n",
    "# Convert the pwm files to meme format using uniprobe2meme\n",
    "# pwm text file, pwm meme output\n",
    "\n",
    "# Call sites using FIMO\n",
    "# --parse-genomic-coord - Used to get coordinates as genomic locations\n",
    "# -bfile - Background file for calls\n",
    "# --text - Output is in tabular format\n",
    "\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    # Convert the pwm files to meme format using uniprobe2meme\n",
    "    uniprobe2meme ../Figure_SX/${tf}_NonUV_pwm.txt > ../Figure_SX/${tf}_NonUV_pwm.meme\n",
    "    # Call sites using FIMO\n",
    "    fimo --parse-genomic-coord --bfile ../Figure_SX/Promoter2000_Merged_markov_bkg.txt --text  ../Figure_SX/${tf}_NonUV_pwm.meme  ${1} > ../Figure_SX/${tf}_PWM_promoter_calls.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "856dab50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create unique bed file versions of the FIMO calls\n",
    "unique_calls_bedfmt_from_fimo(f\"{OUTPUT_SX}/CREB1_PWM_promoter_calls.txt\",\n",
    "                              creb1_pwm_promoter_calls_unique_file)\n",
    "unique_calls_bedfmt_from_fimo(f\"{OUTPUT_SX}/EGR1_PWM_promoter_calls.txt\",\n",
    "                              egr1_pwm_promoter_calls_unique_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dbf304f",
   "metadata": {},
   "source": [
    "# Generate Background Mutations - ZM\n",
    "\n",
    "#### (1) Determine background mutations frequency - Trimers\n",
    "\n",
    "For each 3-mer with orientations treated seperatly, find the positions of all 3-mers. Create a seperate file for each. \n",
    "\n",
    "This is done on an intersection of regions, k-mer counts need to be unique. \n",
    "\n",
    "**Output files:**\n",
    "\n",
    "1. {kmer}_Promoter2000_Merged.bed - Positions of a given k-mer in the promoter as BED file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d00ac1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary of list and specify k\n",
    "kmer_map_dict = defaultdict(list)\n",
    "k = 3\n",
    "# For each sequence, add k-mer positions to kmer_map_dict\n",
    "with open(f\"{OUTPUT_SX}/{promoter_fasta_merged_file}\") as file_obj:\n",
    "    for record in SeqIO.parse(file_obj, \"fasta\"):\n",
    "        add_kmer_positions(str(record.seq), 3, record.description, kmer_map_dict)\n",
    "# Convert the kmer_map_dict to bed files of where k-mers are in the promoter\n",
    "for kmer in kmer_map_dict:\n",
    "    with open(f\"{OUTPUT_3MERS}/{kmer}_Promoter2000_Merged.bed\", 'w') as write_obj:\n",
    "        for match in kmer_map_dict[kmer]:\n",
    "            genomic_coords, idx = match\n",
    "            chromosome = genomic_coords.split(\":\")[0]\n",
    "            start = int(genomic_coords.split(\":\")[1].split(\"-\")[0])\n",
    "            bed_label = f\"{chromosome}\\t{start + idx}\\t{start + idx + k}\\n\"\n",
    "            write_obj.write(bed_label)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2b0e931",
   "metadata": {},
   "source": [
    "#### (2) Intersect mutations file with promoter\n",
    "\n",
    "Intersects the mutations file with the merged promoter file, returning the skcm mutations that are located within the promoter. \n",
    "\n",
    "\n",
    "**File Output**\n",
    "1. skcm_mutations_promoter2000.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d66feb3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$skcm_mutations_file\" \"$promoter_merged_bed_file\" \"$skcm_promoter_mutations_file\"\n",
    "\n",
    "# -a: 0-base coordinates for skcm mutations\n",
    "# -b: Merged promoter file\n",
    "# -u: Returns entries from -a if there is an intersection with -b\n",
    "\n",
    "bedtools intersect -a ${1} -b ${2} -u > ${3}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f54405",
   "metadata": {},
   "source": [
    "#### (3) Convert SKCM_Promoter SNP Mutations to a Trimer Context\n",
    "\n",
    "**File Output:**\n",
    "1. skcm_mutations_promoter2000_trimer.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "43ebf1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjust intersected coordinates to trimers\n",
    "with open(skcm_promoter_mutations_file) as read_obj, \\\n",
    "     open(skcm_promoter_mutations_trimer_bed_file, 'w') as write_obj:\n",
    "    for line in read_obj:\n",
    "        chrom, start, end, wt, snp = line.strip().split()\n",
    "        start = int(start) - 1\n",
    "        end = int(end) + 1\n",
    "        write_obj.write(f\"{chrom}\\t{start}\\t{end}\\t{wt}\\t{snp}\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c930ff0",
   "metadata": {},
   "source": [
    "#### (4) Get sequences for all mutant trimers and convert to BED format\n",
    "\n",
    "**Output files:**\n",
    "1. skcm_mutations_promoter2000_trimer_sequences.tsv\n",
    "2. skcm_mutations_promoter2000_trimer_sequences.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "fa38bf58",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$HUMAN_GENOME_FILE\" \"$skcm_promoter_mutations_trimer_bed_file\" \"$skcm_mutations_trimer_sequence_tsv_file\"\n",
    "\n",
    "bedtools getfasta -fi ${1} -bed ${2} -fo ${3} -tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c44f6435",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(skcm_mutations_trimer_sequence_tsv_file) as read_obj, \\\n",
    "     open(skcm_mutations_trimer_sequence_bed_file, 'w') as write_obj:\n",
    "    for line in read_obj:\n",
    "        genomic_position, seq = line.strip().split()\n",
    "        chrom = genomic_position.split(':')[0]\n",
    "        start, end = genomic_position.split(\":\")[1].split('-')\n",
    "        write_obj.write(f\"{chrom}\\t{start}\\t{end}\\t{seq}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6f7b557",
   "metadata": {},
   "source": [
    "#### (5) Get counts of all mutation trimers in the promoter\n",
    "\n",
    "**Output Files:**\n",
    "1. Promoter_Trimer_Total_and_Mutation_Count_Table.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "912a18fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Trimer</th>\n",
       "      <th>Mutation_Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCC</td>\n",
       "      <td>30436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGA</td>\n",
       "      <td>30127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCC</td>\n",
       "      <td>15712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GGG</td>\n",
       "      <td>15560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGA</td>\n",
       "      <td>10841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TGC</td>\n",
       "      <td>496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>CGT</td>\n",
       "      <td>496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ATC</td>\n",
       "      <td>493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>GAT</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>GCA</td>\n",
       "      <td>444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Trimer  Mutation_Count\n",
       "0     TCC           30436\n",
       "1     GGA           30127\n",
       "2     CCC           15712\n",
       "3     GGG           15560\n",
       "4     TGA           10841\n",
       "..    ...             ...\n",
       "59    TGC             496\n",
       "60    CGT             496\n",
       "61    ATC             493\n",
       "62    GAT             482\n",
       "63    GCA             444\n",
       "\n",
       "[64 rows x 2 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter_trimers = pd.read_csv(skcm_mutations_trimer_sequence_bed_file,\n",
    "                               sep='\\t',\n",
    "                               header=None)\n",
    "promoter_trimers = promoter_trimers.rename(columns={0:\"Chromosome\", 1:\"Start\", 2:\"End\", 3:\"Trimer\"})\n",
    "trimer_count_df = pd.DataFrame(promoter_trimers.value_counts(\"Trimer\"))\n",
    "trimer_count_df = trimer_count_df.reset_index()\n",
    "trimer_count_df = trimer_count_df.rename(columns={0:\"Mutation_Count\"})\n",
    "# Display trimer_count_df\n",
    "trimer_count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "67e83994",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$promoter_trimer_counts_txt_file\"\n",
    "# Get Counts for all trimers in promoter\n",
    "wc -l ../Figure_SX/Promoter_3mers/* > ../Figure_SX/Promoter_Trimer_Counts.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "c441d86f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read trimer count file and process\n",
    "promoter_trimer_counts = pd.read_csv(promoter_trimer_counts_txt_file,\n",
    "                                     delim_whitespace=True,\n",
    "                                     header=None)\n",
    "promoter_trimer_counts = promoter_trimer_counts.rename(columns={0:\"Promoter_Count\", 1:\"File\"})\n",
    "promoter_trimer_counts = promoter_trimer_counts[promoter_trimer_counts[\"File\"] != \"total\"]\n",
    "promoter_trimer_counts[\"Trimer\"] = promoter_trimer_counts[\"File\"].apply(lambda x: x.split('/')[-1].split('_')[0])\n",
    "promoter_trimer_counts = promoter_trimer_counts[[\"Trimer\", \"Promoter_Count\"]]\n",
    "# Merge the mutation and total trimer counts into a single table\n",
    "all_trimer_counts = pd.merge(promoter_trimer_counts, trimer_count_df, on=\"Trimer\")\n",
    "# Save the results\n",
    "all_trimer_counts.to_csv(f\"{OUTPUT_SX}/Promoter_Trimer_Total_and_Mutation_Count_Table.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faa2bf1d",
   "metadata": {},
   "source": [
    "#### (6) Generate background distribution of mutation trimers\n",
    "\n",
    "**Output Files:**\n",
    "1. {OUTPUT_SX}/Promoter_Backgrounds/Background_{n}.bed - Shuffled mutations for use in background comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ee85d136",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate M selections from range P 1000+ times\n",
    "# Set Seed\n",
    "random_gen = np.random.default_rng(0)\n",
    "# Set N samples\n",
    "n_samples = 100\n",
    "\n",
    "# For each trimer and its associated Total Promter Count and Mutant Promoter Count\n",
    "for row in all_trimer_counts.itertuples():\n",
    "    # Read the Total trimer count into memory as a list of lines\n",
    "    trimer_file = f\"{OUTPUT_3MERS}/{row.Trimer}_Promoter2000_Merged.bed\"\n",
    "    with open(trimer_file) as read_obj:\n",
    "        trimer_lines = read_obj.readlines()\n",
    "    # For each backgound file to make\n",
    "    for n in range(n_samples):\n",
    "        # Randomly select trimer locations with no replacement from uniform distribution\n",
    "        trimer_idxs = random_gen.choice(a=row.Promoter_Count, size=row.Mutation_Count, replace=False)\n",
    "        # Subset by index the trimer locations in memory\n",
    "        selected_lines = [trimer_lines[idx] for idx in trimer_idxs]\n",
    "        # Save the subset and an append\n",
    "        output_file = f\"{OUTPUT_SX}/Promoter_Backgrounds_ZM/Background_{n}.bed\"\n",
    "        with open(output_file, 'a') as append_obj:\n",
    "            for line in selected_lines:\n",
    "                append_obj.write(line)\n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e94af7d6",
   "metadata": {},
   "source": [
    "# Generate Mutations - HS\n",
    "\n",
    "Pipeline for generating mutations using Harshit's method. \n",
    "\n",
    "Files are stored in folders with a suffix \\_HS. \n",
    "\n",
    "Note-ZM: The changes I made are just:\n",
    "\n",
    "1. Update file locations\n",
    "2. More comments to help illustrate the process for me\n",
    "3. If a duplicate file is made that I verified is 100% the same, I indicated so and used my previously generated file. \n",
    "4. I refactored some stuff only when there was a quick way to save memory or time to help me troubleshoot better.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3cab9318",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a list of all trimers\n",
    "bases=[\"A\",\"T\",\"C\",\"G\"]\n",
    "trimers=[]\n",
    "for base1 in bases:\n",
    "    for base2 in bases:\n",
    "        for base3 in bases:\n",
    "            trimers.append(base1+base2+base3)\n",
    "            \n",
    "# Read in all mutations as a CSV file, then extend it to trimers\n",
    "## Original code\n",
    "#mutations=pd.read_csv(\"skcm_mutations.bed\",sep=\"\\t\",header=None)\n",
    "#tmp=mutations\n",
    "#tmp[1]=tmp[1]-2\n",
    "#tmp[2]=tmp[2]+1\n",
    "#tmp.to_csv(\"skcm_mutations_extended.bed\",sep=\"\\t\",index=None,header=None)\n",
    "\n",
    "## Updated Code: Change is the whole df is not stored in memory\n",
    "with open(f\"{OUTPUT_SX}/skcm_mutations.bed\") as read_obj, \\\n",
    "     open(f\"{OUTPUT_SX}/skcm_mutations_extended_HS.bed\", 'w') as write_obj:\n",
    "    for line in read_obj:\n",
    "        chrom, start, end, wt, snp = line.strip().split()\n",
    "        start = int(start) - 2\n",
    "        end = int(end) + 1\n",
    "        write_obj.write(f\"{chrom}\\t{start}\\t{end}\\t{wt}\\t{snp}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e20f3b65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Figure_SX\n",
      "../../Data/External_Data/human_g1k_v37.fasta\n",
      "../Figure_SX/Promoter2000.bed\n",
      ".....Original Mutations File.....\n",
      "chr8\t109768883\t109768883\tA\tG\n",
      "chr4\t3643290\t3643290\tC\tA\n",
      "chr8\t7655500\t7655500\tC\tT\n",
      "chr10\t2126670\t2126670\tC\tT\n",
      "chr8\t80248082\t80248082\tA\tT\n",
      "chr10\t2543554\t2543554\tG\tA\n",
      "chr10\t71999467\t71999467\tG\tA\n",
      "chrX\t125008199\t125008199\tA\tC\n",
      "chr5\t152513013\t152513013\tT\tC\n",
      "chr13\t40707667\t40707667\tT\tG\n",
      ".....Trimer Mutations File.....\n",
      "chr8\t109768881\t109768884\tA\tG\n",
      "chr4\t3643288\t3643291\tC\tA\n",
      "chr8\t7655498\t7655501\tC\tT\n",
      "chr10\t2126668\t2126671\tC\tT\n",
      "chr8\t80248080\t80248083\tA\tT\n",
      "chr10\t2543552\t2543555\tG\tA\n",
      "chr10\t71999465\t71999468\tG\tA\n",
      "chrX\t125008197\t125008200\tA\tC\n",
      "chr5\t152513011\t152513014\tT\tC\n",
      "chr13\t40707665\t40707668\tT\tG\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$OUTPUT_SX\" \"$HUMAN_GENOME_FILE\" \"$promoter_bed_file\"\n",
    "\n",
    "echo ${1}\n",
    "echo ${2}\n",
    "echo ${3}\n",
    "echo \".....Original Mutations File.....\"\n",
    "head \"../Figure_SX/skcm_mutations.bed\"\n",
    "echo \".....Trimer Mutations File.....\"\n",
    "head \"../Figure_SX/skcm_mutations_extended_HS.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "331f2503",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr8\t109768881\t109768884\tA\tG\tchr8:109768881-109768884\tAAA\n",
      "chr4\t3643288\t3643291\tC\tA\tchr4:3643288-3643291\tACA\n",
      "chr8\t7655498\t7655501\tC\tT\tchr8:7655498-7655501\tCCG\n",
      "chr10\t2126668\t2126671\tC\tT\tchr10:2126668-2126671\tTCC\n",
      "chr8\t80248080\t80248083\tA\tT\tchr8:80248080-80248083\tTAT\n",
      "chr10\t2543552\t2543555\tG\tA\tchr10:2543552-2543555\tCGG\n",
      "chr10\t71999465\t71999468\tG\tA\tchr10:71999465-71999468\tCGG\n",
      "chrX\t125008197\t125008200\tA\tC\tchrX:125008197-125008200\tGAC\n",
      "chr5\t152513011\t152513014\tT\tC\tchr5:152513011-152513014\tTTA\n",
      "chr13\t40707665\t40707668\tT\tG\tchr13:40707665-40707668\tCTG\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$OUTPUT_SX\" \"$HUMAN_GENOME_FILE\" \"$promoter_bed_file\"\n",
    "\n",
    "# Remove any MT genome information\n",
    "grep -v MT \"${1}/skcm_mutations_extended_HS.bed\" > \"${1}/skcm_mutations_extended_NoMT_HS.bed\"\n",
    "\n",
    "# Get fasta sequences for the 3-mer skin mutations\n",
    "bedtools getfasta -tab -fi ${2} -bed \"${1}/skcm_mutations_extended_NoMT_HS.bed\" > \"${1}/skcm_mutations_extended_HS.fa\"\n",
    "# Merge sequences horizontally\n",
    "paste \"${1}/skcm_mutations_extended_NoMT_HS.bed\" \"${1}/skcm_mutations_extended_HS.fa\" > \"${1}/skcm_mutations_extended_with_sequence_HS.bed\"\n",
    "# Display start of file\n",
    "head \"${1}/skcm_mutations_extended_with_sequence_HS.bed\"\n",
    "# This file has all mutations along with a trimer context. \n",
    "# This will help with the intersection amongst trimers.\n",
    "# We might wanna only shuffle all promoter mutations, so first we intersect this with the promoter file.\n",
    "# Flag -u is to only keep mutations once, even if found in multiple promoters (they can overlap)\n",
    "bedtools intersect -wa -wb -u -a \"${1}/skcm_mutations_extended_with_sequence_HS.bed\" -b ${3} \\\n",
    "> \"${1}/skcm_mutations_extended_with_sequence_promoters_only_HS.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2333b820",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>130517603</td>\n",
       "      <td>130517606</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:130517603-130517606</td>\n",
       "      <td>CAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>162531408</td>\n",
       "      <td>162531411</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:162531408-162531411</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>144639613</td>\n",
       "      <td>144639616</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr8:144639613-144639616</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>54156084</td>\n",
       "      <td>54156087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3:54156084-54156087</td>\n",
       "      <td>CTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13714186</td>\n",
       "      <td>13714189</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:13714186-13714189</td>\n",
       "      <td>GAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2  3  4                         5    6\n",
       "0  chr9  130517603  130517606  A  C  chr9:130517603-130517606  CAC\n",
       "1  chr1  162531408  162531411  T  G  chr1:162531408-162531411  GTG\n",
       "2  chr8  144639613  144639616  C  A  chr8:144639613-144639616  ACG\n",
       "3  chr3   54156084   54156087  T  C    chr3:54156084-54156087  CTT\n",
       "4  chr1   13714186   13714189  A  G    chr1:13714186-13714189  GAG"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read skin trimer mutations in promoters\n",
    "prom_mutations=pd.read_csv(f\"{OUTPUT_SX}/skcm_mutations_extended_with_sequence_promoters_only_HS.bed\",\n",
    "                           sep=\"\\t\",\n",
    "                           header=None)\n",
    "prom_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "322450d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>130517603</td>\n",
       "      <td>130517606</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr9:130517603-130517606</td>\n",
       "      <td>CAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>162531408</td>\n",
       "      <td>162531411</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:162531408-162531411</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>144639613</td>\n",
       "      <td>144639616</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr8:144639613-144639616</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>54156084</td>\n",
       "      <td>54156087</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3:54156084-54156087</td>\n",
       "      <td>CTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13714186</td>\n",
       "      <td>13714189</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:13714186-13714189</td>\n",
       "      <td>GAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2  3  4                         5    6\n",
       "0  chr9  130517603  130517606  A  C  chr9:130517603-130517606  CAC\n",
       "1  chr1  162531408  162531411  T  G  chr1:162531408-162531411  GTG\n",
       "2  chr8  144639613  144639616  C  A  chr8:144639613-144639616  ACG\n",
       "3  chr3   54156084   54156087  T  C    chr3:54156084-54156087  CTT\n",
       "4  chr1   13714186   13714189  A  G    chr1:13714186-13714189  GAG"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Also stores which promoter the mutation was in, but we don't necessarily need that. \n",
    "#column 6 is the trimer, column 3 and 4 are reference and mutant alleles respectively.\n",
    "##converting all bases to upper-case \n",
    "prom_mutations[6]=prom_mutations[6].str.upper()\n",
    "prom_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d2bc2222",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(PROMOTER_MUTATIONS_DIR_HS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c0d5c714",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAA Done\n",
      "AAT Done\n",
      "AAC Done\n",
      "AAG Done\n",
      "ATA Done\n",
      "ATT Done\n",
      "ATC Done\n",
      "ATG Done\n",
      "ACA Done\n",
      "ACT Done\n",
      "ACC Done\n",
      "ACG Done\n",
      "AGA Done\n",
      "AGT Done\n",
      "AGC Done\n",
      "AGG Done\n",
      "TAA Done\n",
      "TAT Done\n",
      "TAC Done\n",
      "TAG Done\n",
      "TTA Done\n",
      "TTT Done\n",
      "TTC Done\n",
      "TTG Done\n",
      "TCA Done\n",
      "TCT Done\n",
      "TCC Done\n",
      "TCG Done\n",
      "TGA Done\n",
      "TGT Done\n",
      "TGC Done\n",
      "TGG Done\n",
      "CAA Done\n",
      "CAT Done\n",
      "CAC Done\n",
      "CAG Done\n",
      "CTA Done\n",
      "CTT Done\n",
      "CTC Done\n",
      "CTG Done\n",
      "CCA Done\n",
      "CCT Done\n",
      "CCC Done\n",
      "CCG Done\n",
      "CGA Done\n",
      "CGT Done\n",
      "CGC Done\n",
      "CGG Done\n",
      "GAA Done\n",
      "GAT Done\n",
      "GAC Done\n",
      "GAG Done\n",
      "GTA Done\n",
      "GTT Done\n",
      "GTC Done\n",
      "GTG Done\n",
      "GCA Done\n",
      "GCT Done\n",
      "GCC Done\n",
      "GCG Done\n",
      "GGA Done\n",
      "GGT Done\n",
      "GGC Done\n",
      "GGG Done\n"
     ]
    }
   ],
   "source": [
    "#-ZM Save promoter mutations of each trimer to individual files\n",
    "for trimer in trimers:\n",
    "    df=prom_mutations[prom_mutations[6]==trimer]\n",
    "    df.to_csv(f\"{PROMOTER_MUTATIONS_DIR_HS}/skcm_promoter_mutations_{trimer}_HS\",\n",
    "              sep=\"\\t\",\n",
    "              header=None,\n",
    "              index=False)\n",
    "    #df.to_csv(\"skcm_promoter_mutations_\"+trimer,sep=\"\\t\",header=None,index=False)\n",
    "    print(trimer,\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ccaf9616",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9\t90531559\t90531562\tA\tT\tchr9:90531559-90531562\tAAA\n",
      "chr7\t48962406\t48962409\tA\tG\tchr7:48962406-48962409\tAAA\n",
      "chr1\t153960002\t153960005\tA\tG\tchr1:153960002-153960005\tAAA\n",
      "chr14\t60338971\t60338974\tA\tT\tchr14:60338971-60338974\tAAA\n",
      "chr3\t111696171\t111696174\tA\tC\tchr3:111696171-111696174\tAAA\n",
      "chr22\t32336400\t32336403\tA\tT\tchr22:32336400-32336403\tAAA\n",
      "chr19\t56348422\t56348425\tA\tT\tchr19:56348422-56348425\tAAA\n",
      "chr5\t108746291\t108746294\tA\tG\tchr5:108746291-108746294\tAAA\n",
      "chr11\t17099993\t17099996\tA\tG\tchr11:17099993-17099996\tAAA\n",
      "chr2\t130736436\t130736439\tA\tC\tchr2:130736436-130736439\tAAA\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$PROMOTER_MUTATIONS_DIR_HS\"\n",
    "\n",
    "head \"${1}/skcm_promoter_mutations_AAA_HS\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3ee4b935",
   "metadata": {},
   "outputs": [],
   "source": [
    "ANNOTATED_PROMOTER_MUTATIONS_DIR = f\"{OUTPUT_SX}/Promoter_Mutations_Annotated_HS\"\n",
    "os.makedirs(ANNOTATED_PROMOTER_MUTATIONS_DIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d1531972",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2334 2339 AAA\n",
      "1404 1405 AAT\n",
      "2365 2371 AAC\n",
      "1315 1316 AAG\n",
      "943 940 ATA\n",
      "1391 1392 ATT\n",
      "493 494 ATC\n",
      "716 716 ATG\n",
      "768 764 ACA\n",
      "1279 1277 ACT\n",
      "2814 2809 ACC\n",
      "501 500 ACG\n",
      "10251 10081 AGA\n",
      "1287 1283 AGT\n",
      "1671 1670 AGC\n",
      "9776 9600 AGG\n",
      "1228 1227 TAA\n",
      "975 970 TAT\n",
      "561 559 TAC\n",
      "567 565 TAG\n",
      "1251 1254 TTA\n",
      "2371 2365 TTT\n",
      "973 973 TTC\n",
      "715 714 TTG\n",
      "10718 10485 TCA\n",
      "10493 10341 TCT\n",
      "30464 29251 TCC\n",
      "7427 7098 TCG\n",
      "10857 10572 TGA\n",
      "759 757 TGT\n",
      "497 495 TGC\n",
      "7560 7538 TGG\n",
      "769 772 CAA\n",
      "721 717 CAT\n",
      "2017 2019 CAC\n",
      "940 940 CAG\n",
      "542 540 CTA\n",
      "1295 1286 CTT\n",
      "1418 1416 CTC\n",
      "972 975 CTG\n",
      "7425 7387 CCA\n",
      "9835 9653 CCT\n",
      "15722 15561 CCC\n",
      "3970 3925 CCG\n",
      "7563 7175 CGA\n",
      "500 493 CGT\n",
      "925 922 CGC\n",
      "4022 3983 CGG\n",
      "1048 1048 GAA\n",
      "483 481 GAT\n",
      "853 851 GAC\n",
      "1499 1502 GAG\n",
      "590 591 GTA\n",
      "2414 2405 GTT\n",
      "898 899 GTC\n",
      "1910 1911 GTG\n",
      "445 443 GCA\n",
      "1788 1772 GCT\n",
      "2949 2933 GCC\n",
      "935 938 GCG\n",
      "30156 29009 GGA\n",
      "2864 2853 GGT\n",
      "3027 3011 GGC\n",
      "15567 15399 GGG\n"
     ]
    }
   ],
   "source": [
    "##For each trimer, read in the bed file of promoters and the bed files of mutations\n",
    "##This returns a dataframe that annotates the original promoter 3 mers with a new column, containing 0 for no mutations\n",
    "##or 1 for a mutation. The idea is to shuffle this column to generate different labels\n",
    "\n",
    "for trimer in trimers: \n",
    "    # Read mutation data of the trimer\n",
    "    #mutation_file=pd.read_csv(\"skcm_promoter_mutations_\"+trimer,sep=\"\\t\",header=None)\n",
    "    mutation_file=pd.read_csv(f\"{PROMOTER_MUTATIONS_DIR_HS}/skcm_promoter_mutations_{trimer}_HS\",sep=\"\\t\",header=None)\n",
    "    # Read promoter data of the trimer\n",
    "    #promoter_3mer=pd.read_csv(\"Genome_Figure_Data/Promoter_3mers/\"+trimer+\"_2000Promoter.bed\",sep=\"\\t\",header=None)\n",
    "    promoter_3mer=pd.read_csv(f\"{OUTPUT_3MERS}/{trimer}_Promoter2000_Merged.bed\",sep=\"\\t\",header=None)\n",
    "    #-ZM ???? What is all of this doing?\n",
    "    promoter_3mer[3]=(promoter_3mer[1].isin(mutation_file[1]) & promoter_3mer[2].isin(mutation_file[2]))\n",
    "    promoter_3mer[3]=promoter_3mer[3]*1 # Typecast boolean to int\n",
    "    promoter_3mer[4]=trimer # Make 4th column the trimer\n",
    "    promoter_3mer=promoter_3mer.drop_duplicates()\n",
    "    # Save annotated trimer\n",
    "    #promoter_3mer.to_csv(\"skcm_promoter_mutations_\"+trimer+\"_annotated\",index=False,sep=\"\\t\",header=None)\n",
    "    promoter_3mer.to_csv(f\"{ANNOTATED_PROMOTER_MUTATIONS_DIR}/skcm_promoter_mutations_{trimer}_annotated_HS\",\n",
    "                         index=False,\n",
    "                         sep=\"\\t\",\n",
    "                         header=None)\n",
    "    print(len(mutation_file),promoter_3mer[3].sum(),trimer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d1a4d947",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t857420\t857423\t0\tAAA\n",
      "chr1\t857512\t857515\t0\tAAA\n",
      "chr1\t857620\t857623\t0\tAAA\n",
      "chr1\t858235\t858238\t0\tAAA\n",
      "chr1\t858742\t858745\t0\tAAA\n",
      "chr1\t858799\t858802\t0\tAAA\n",
      "chr1\t858800\t858803\t0\tAAA\n",
      "chr1\t858927\t858930\t0\tAAA\n",
      "chr1\t859012\t859015\t0\tAAA\n",
      "chr1\t859274\t859277\t0\tAAA\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$ANNOTATED_PROMOTER_MUTATIONS_DIR\"\n",
    "##eg of such a file\n",
    "##column 3 is mutation status\n",
    "##column 4 is original trimer, with central base mutated\n",
    "head \"${1}/skcm_promoter_mutations_AAA_annotated_HS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1b3f01b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n",
      "/tmp/ipykernel_10955/3472069091.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  mut_df=mut_df.append(df, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "#-ZM ? So the annotated files have 1 if it is part of the original mutation set? Otherwise 0?\n",
    "# Is this just a control? Or to format easily? \n",
    "\n",
    "###regenerate original mutation set\n",
    "mut_df= pd.DataFrame()\n",
    "\n",
    "for trimer in trimers:\n",
    "    #df=pd.read_csv(\"skcm_promoter_mutations_\"+trimer+\"_annotated\",sep=\"\\t\",header=None)\n",
    "    df=pd.read_csv(f\"{ANNOTATED_PROMOTER_MUTATIONS_DIR}/skcm_promoter_mutations_{trimer}_annotated_HS\",sep=\"\\t\",header=None)\n",
    "    ##keep rows with original mutations\n",
    "    df=df[df[3]==1]\n",
    "    mut_df=mut_df.append(df, ignore_index=True)\n",
    "    #print(trimer)\n",
    "\n",
    "#mut_df.to_csv(\"skcm_promoter_mutations_reconstructed_from_trimers\",sep=\"\\t\",header=None,index=None)\n",
    "mut_df.to_csv(f\"{OUTPUT_SX}/skcm_promoter_mutations_reconstructed_from_trimers_HS\",\n",
    "              sep=\"\\t\",\n",
    "              header=None,\n",
    "              index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4b0befb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t1294741\t1294744\t1\tAAA\n",
      "chr1\t2937165\t2937168\t1\tAAA\n",
      "chr1\t3667760\t3667763\t1\tAAA\n",
      "chr1\t3713118\t3713121\t1\tAAA\n",
      "chr1\t6105305\t6105308\t1\tAAA\n",
      "chr1\t6422497\t6422500\t1\tAAA\n",
      "chr1\t7763772\t7763775\t1\tAAA\n",
      "chr1\t8940068\t8940071\t1\tAAA\n",
      "chr1\t9710826\t9710829\t1\tAAA\n",
      "chr1\t12892554\t12892557\t1\tAAA\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$OUTPUT_SX\"\n",
    "# Visualize output file\n",
    "head \"${1}/skcm_promoter_mutations_reconstructed_from_trimers_HS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "18a18ffe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>857420</td>\n",
       "      <td>857423</td>\n",
       "      <td>0</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>857512</td>\n",
       "      <td>857515</td>\n",
       "      <td>0</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>857620</td>\n",
       "      <td>857623</td>\n",
       "      <td>0</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>858235</td>\n",
       "      <td>858238</td>\n",
       "      <td>0</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>858742</td>\n",
       "      <td>858745</td>\n",
       "      <td>0</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2  3    4\n",
       "0  chr1  857420  857423  0  AAA\n",
       "1  chr1  857512  857515  0  AAA\n",
       "2  chr1  857620  857623  0  AAA\n",
       "3  chr1  858235  858238  0  AAA\n",
       "4  chr1  858742  858745  0  AAA"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(f\"{ANNOTATED_PROMOTER_MUTATIONS_DIR}/skcm_promoter_mutations_AAA_annotated_HS\",\n",
    "               sep=\"\\t\",\n",
    "               header=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "a5da503d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1887</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2936703</td>\n",
       "      <td>2936706</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1929</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3370162</td>\n",
       "      <td>3370165</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2344</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6094343</td>\n",
       "      <td>6094346</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2641</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6420138</td>\n",
       "      <td>6420141</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6421867</td>\n",
       "      <td>6421870</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322932</th>\n",
       "      <td>chrY</td>\n",
       "      <td>9364170</td>\n",
       "      <td>9364173</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1324039</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20707481</td>\n",
       "      <td>20707484</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1324155</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20936478</td>\n",
       "      <td>20936481</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1324257</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21156035</td>\n",
       "      <td>21156038</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1324361</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21438866</td>\n",
       "      <td>21438869</td>\n",
       "      <td>1</td>\n",
       "      <td>AAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2339 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0         1         2  3    4\n",
       "1887     chr1   2936703   2936706  1  AAA\n",
       "1929     chr1   3370162   3370165  1  AAA\n",
       "2344     chr1   6094343   6094346  1  AAA\n",
       "2641     chr1   6420138   6420141  1  AAA\n",
       "2707     chr1   6421867   6421870  1  AAA\n",
       "...       ...       ...       ... ..  ...\n",
       "1322932  chrY   9364170   9364173  1  AAA\n",
       "1324039  chrY  20707481  20707484  1  AAA\n",
       "1324155  chrY  20936478  20936481  1  AAA\n",
       "1324257  chrY  21156035  21156038  1  AAA\n",
       "1324361  chrY  21438866  21438869  1  AAA\n",
       "\n",
       "[2339 rows x 5 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(0) \n",
    "df[3]=np.random.permutation(df[3].values)\n",
    "df[df[3]==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "fee911db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         0        1        2  3    4\n",
      "317   chr1  1165697  1165700  1  AAA\n",
      "483   chr1  1261874  1261877  1  AAA\n",
      "2574  chr1  6297730  6297733  1  AAA\n",
      "4023  chr1  8879327  8879330  1  AAA\n",
      "4489  chr1  9597532  9597535  1  AAA\n",
      "         0        1        2  3    4\n",
      "974   chr1  1411791  1411794  1  AAA\n",
      "1401  chr1  1712463  1712466  1  AAA\n",
      "2047  chr1  3667658  3667661  1  AAA\n",
      "2700  chr1  6421570  6421573  1  AAA\n",
      "2745  chr1  6422496  6422499  1  AAA\n"
     ]
    }
   ],
   "source": [
    "##now we will randomly permute the 3rd column (which contains 0 or 1). For 100 permutations, \n",
    "##we will store 100 different combined mutation files\n",
    "\n",
    "np.random.seed(0)\n",
    "\n",
    "for i in range(0,2):\n",
    "    df[3]=np.random.permutation(df[3].values)\n",
    "    print(df[df[3]==1].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "170c7c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "db5d5884",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n"
     ]
    }
   ],
   "source": [
    "# Set random seed to 0\n",
    "np.random.seed(0)\n",
    "# For 100 backgroud files\n",
    "for i in range(0,100):\n",
    "    #trimer=\"AAA\"\n",
    "    mut_df= pd.DataFrame()\n",
    "    # For each trimer: (AAA, AAT, etc.)\n",
    "    for trimer in trimers:\n",
    "        # Read the trimer annotated file\n",
    "        #df=pd.read_csv(\"skcm_promoter_mutations_\"+trimer+\"_annotated\",sep=\"\\t\",header=None)\n",
    "        df=pd.read_csv(f\"{ANNOTATED_PROMOTER_MUTATIONS_DIR}/skcm_promoter_mutations_{trimer}_annotated_HS\",sep=\"\\t\",header=None)\n",
    "        ##permute mutated or not labels\n",
    "        df[3]=np.random.permutation(df[3].values)\n",
    "        df=df[df[3]==1]\n",
    "        mut_df=mut_df.append(df, ignore_index=True)\n",
    "    # i=i+1 # Convert to 1-base labels?\n",
    "    #mut_df.to_csv(\"skcm_promoter_mutations_reconstructed_from_trimers_shuffle\"+str(i),sep=\"\\t\",header=None,index=None)\n",
    "    mut_df.to_csv(f\"{OUTPUT_SX}/Promoter_Backgrounds_HS/skcm_promoter_mutations_reconstructed_from_trimers_shuffle_HS{i+1}\",sep=\"\\t\",header=None,index=None)\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "47d986e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pipeline for Harshit's Promoter Backgrounds\n",
    "os.makedirs(f\"{OUTPUT_SX}/Promoter_Backgrounds\", exist_ok=True)\n",
    "import glob\n",
    "\n",
    "files = glob.glob(f\"{OUTPUT_SX}/Promoter_Backgrounds_HS/*\")\n",
    "for read_file in files:\n",
    "    fnumber = read_file.split('_shuffle_HS')[1]\n",
    "    with open(read_file) as read_obj, \\\n",
    "         open(f\"../Figure_SX/Promoter_Backgrounds/Background_{fnumber}.bed\", 'w') as write_obj:\n",
    "        for line in read_obj:\n",
    "            components = line.split()\n",
    "            chromosome = components[0]\n",
    "            start = int(components[1])\n",
    "            end = int(components[2])\n",
    "            write_obj.write(f\"{chromosome}\\t{start + 1}\\t{end - 1}\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e99cce",
   "metadata": {},
   "source": [
    "# Generate Background Site choices\n",
    "\n",
    "TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4eba8d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Shuffle sites \n",
    "\n",
    "for i in {0..99}\n",
    "do\n",
    "    for tf in CREB1 EGR1\n",
    "    do\n",
    "        for min_length in 20 25 30\n",
    "        do\n",
    "            bedtools shuffle -seed 0 -i \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -g \"../../Data/External_Data/hg19.chrom.sizes\" -incl \"../Figure_SX/Promoter2000_Merged.bed\" > \"../Figure_SX/Site_Shuffle/${tf}_${min_length}_Shuffle_${i}.bed\" \n",
    "            bedtools intersect -a \"../Figure_SX/Site_Shuffle/${tf}_${min_length}_Shuffle_${i}.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/Site_Shuffle_Intersection/NC_${tf}_${min_length}_Counts_${i}.bed\"\n",
    "        done\n",
    "        bedtools shuffle -seed 0 -i \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -g \"../../Data/External_Data/hg19.chrom.sizes\" -incl \"../Figure_SX/Promoter2000_Merged.bed\" > \"../Figure_SX/Site_Shuffle/${tf}_PWM_Shuffle_${i}.bed\" \n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dd1b364",
   "metadata": {},
   "source": [
    "#### (9) Get mutation Counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a45adb9",
   "metadata": {},
   "source": [
    "# Pipeline - Flanks NoFootprints - Harshit Backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ab09a348",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"{OUTPUT_SX}/Background_Intersection\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e2a27380",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# GENERATE FLANK FILES for PWM CALLS\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    bedtools slop -i \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -g \"../../Data/External_Data/hg19.chrom.sizes\" -b 5 > \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\"\n",
    "done\n",
    "\n",
    "# BACKGROUND INTERSECTIONS FOR FLANK FILES\n",
    "for i in {1..100}\n",
    "do\n",
    "    for tf in CREB1 EGR1\n",
    "    do\n",
    "        for min_length in 20 25 30\n",
    "        do\n",
    "            bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/NC_${tf}_${min_length}_Counts_${i}.bed\"\n",
    "        done\n",
    "        bedtools intersect -a \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/${tf}_PWM_Background_Counts_${i}.bed\"\n",
    "    done\n",
    "done\n",
    "\n",
    "# Intersect PWM (No-Footprint - FLANK) and NonConsensus Calls with Mutations\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    for min_length in 20 25 30\n",
    "    do\n",
    "        bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/NonConsensus_${tf}_in_Promoters_${min_length}_Unique_Mutation_Counts.bed\"\n",
    "    done\n",
    "    bedtools intersect -a \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/${tf}_PWM_Promoter_Unique_Mutation_Counts.bed\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ecbd4b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "background_mr_df = aggregate_bkg_mutation_rate_df(100, label_0base=False)\n",
    "measured_mr_df = measured_mutation_rate_df()\n",
    "plot_mutation_rate(background_mr_df, measured_mr_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1b61249",
   "metadata": {},
   "source": [
    "# Pipeline - Flanks YesFootprints - Harshit Backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5e418203",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "os.makedirs(BACKGROUND_INTERSECTION_DIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3f8adc78",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# GENERATE FLANK FILES for PWM CALLS in footptints\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    bedtools intersect -u -a \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -b \"../Figure_SX/consensus_footprints_and_collapsed_motifs_simple_hg19.bed\" > \"../Figure_SX/${tf}_PWM_promoter_calls_infootprints_unique.bed\"\n",
    "    bedtools slop -i \"../Figure_SX/${tf}_PWM_promoter_calls_infootprints_unique.bed\" -g \"../../Data/External_Data/hg19.chrom.sizes\" -b 5 > \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\"\n",
    "done\n",
    "\n",
    "# BACKGROUND INTERSECTIONS FOR FLANK FILES\n",
    "for i in {1..100}\n",
    "do\n",
    "    for tf in CREB1 EGR1\n",
    "    do\n",
    "        for min_length in 20 25 30\n",
    "        do\n",
    "            bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/NC_${tf}_${min_length}_Counts_${i}.bed\"\n",
    "        done\n",
    "        bedtools intersect -a \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/${tf}_PWM_Background_Counts_${i}.bed\"\n",
    "    done\n",
    "done\n",
    "\n",
    "# Intersect PWM (No-Footprint - FLANK) and NonConsensus Calls with Mutations\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    for min_length in 20 25 30\n",
    "    do\n",
    "        bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/NonConsensus_${tf}_in_Promoters_${min_length}_Unique_Mutation_Counts.bed\"\n",
    "    done\n",
    "    bedtools intersect -a \"../Figure_SX/${tf}_PWM_Promoter_Calls_Unique_Flanks.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/${tf}_PWM_Promoter_Unique_Mutation_Counts.bed\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "0ec75e8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEvCAYAAABSRCY3AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAA4EUlEQVR4nO3deXwV1d3H8c8vIRAI+2LYRBZRW5VFwiJKRVQUFNFSRbTFuiEqtS4VrdoWebRVCrZuVRRReKxbXVERtIoLKo+yiaCiARSDuCCryJbk9/wxc+NNuMskZu7MJL/363VfmeVM7tfrkHNnzplzRFUxxhhjvMoKOoAxxphosYrDGGNMpVjFYYwxplKs4jDGGFMpVnEYY4ypFKs4jDHGVIpVHMYYYyqlTroCIpIFdAfaAjuAFar6td/BjDHGhJMkewBQRLoAVwPHAp8C3wK5wAHAD8BUYIaqlmYmatW1bNlSO3bsGHQMY4yJlEWLFm1Q1VYVt6eqOB4B7gbe1AqFRGQf4Exgk6rO8CFvtSooKNCFCxcGHcMYYyJFRBapakHF7UlvVanqqBT7vgH+WT3RjDHGREnaNg4AEekPdIwvr6ozfcpkjDEmxLw0jv8v0AVYCpS4mxWwisMYY2ohL1ccBcDPK7ZzGGOMqZ28PMexHGjtdxBjjDHRkLTiEJHnRGQW0BL4UETmisis2MvLLxeRE0RkpYgUisg1CfaLiNzu7l8mIofF7ZsuIt+IyPIKxzQXkZdF5FP3ZzPv/7nGGFMDlZbC7Nlw443Oz1J/n5JIdatq8k/5xSKSDdwFHAcUAe+JyCxV/TCu2BCgq/vqi9P9t6+770HgTvZuS7kGeEVVb3Yro2twnjcxxpjap7QUTj0VZsV9nz/5ZHj6acjyZ3CQpL9VVV9X1deBobHl+G0efncfoFBVV6vqbuBRYHiFMsOBmepYADQVkTbu+78BbEzwe4cDsWdHZgCneMhijDE105w55SsNcNbnzPHtLb1UR8cl2DbEw3HtgC/i1ovcbZUtU1G+qq4HcH/uk6iQiIwRkYUisvDbb7/1ENcYYyJo8eLE25cs8e0tU7VxXCQiHwAHuu0PsdcaYJmH3y0JtlXsmeWlTJWo6r2qWqCqBa1a7fXEvDHG1AyHHZZ4e8+evr1lqiuOh4FhwCz3Z+zVS1V/7eF3FwH7xq23B76sQpmKvo7dznJ/fuMhizHG1EwnnOC0acQ7+WRnu09SDTmyBdgiIpdU3CciOaq6J83vfg/oKiKdgHXAGTjjW8WbBYwTkUdxGsW3xG5DpTALOBu42f35bJryxhhTc2VlOQ3hc+Y4t6d69nQqDZ8axsHbA4CLca4KNuHcWmoKrBeRb4ALVHVRooNUtVhExgFzgWxguqquEJGx7v57gNk4De2FOCPunhM73h1kcSDQUkSKgL+o6v04FcbjInIesBY4rbL/0cYYU6NkZcHQoc4rA5KOjltWQOQe4GlVneuuDwZOAB4HblPVvqmODwMbHdcYYyov2ei4Xq5lCmKVBoCqvgT8wu0+W68aMxpjjIkAL7eqNorI1TjPYQCMBDa5D/iFfhInY4wx1cvLFceZOL2dnsFpiO7gbssGTvctmTHGmFBKe8WhqhuA3yXZXVi9cYwxxoSdl/k4DgD+wN4TOQ3yL5Yxxpiw8tLG8R/gHmAaP07kZIwxppbyUnEUq+rdvicxxhgTCV4ax58TkYtFpI07F0ZzEWnuezJjjDGh5OWK42z351Vx2xToXP1xjDHGhJ2XXlWdMhHEGGNMNKS9VSUiDUTkehG5113vKiIn+R/NGGNMGHlp43gA2A30d9eLgBt9S2SMMSbUvFQcXVR1ErAHQFV3kHgCJmOMMbWAl4pjt4jUx52ZT0S6ALt8TWWMMSa0vPSq+gswB9hXRP4NHAH81s9QxhhjwstLr6qXRWQx0A/nFtXv3fGrjDHG1EJJKw4RqTgDemxK1w4i0kFVF/sXyxhjTFiluuKYkmKfAjbIoTHG1EJJKw5VPTqTQYwxxkSDl15VxhhjTBmrOIwxxlSKVRzGGGMqpTK9qsqxXlXGGFM7Wa8qY4wxlWK9qowxxlSKlyFHEJFDgJ8DubFtqjrTr1DGGGPCK23FISJ/AQbiVByzgSHAfMAqDmOMqYW89Kr6FXAM8JWqngN0B+r5msoYY0xoeak4dqhqKVAsIo2Bb7D5xo0xptby0saxUESaAvcBi4DvgXf9DGWMMSa80l5xqOrFqrpZVe8BjgPOdm9ZGWOMCdDEiRPp378/L730UkbfV1Q1dQGRV1T1mHTbwqygoEAXLlwYdAxjjKk269ato3379gAcdNBB1KtXjyZNmvDUU0/RokWLankPEVmkqgUVtye94hCRXBFpDrQUkWYi0tx9dQTaVksqY4wxVdKyZUsOPvhgAPLy8nj//fd54403eOaZZ3x/71RtHBcCl+FUEvHDi2wF7vIxkzHGmBS2b9/O5MmTGTt2LIMHD2b9+vUMGTKEhg0bctRRR/n+/qmeHL8NuE1Efqeqd/iexBhjjCcTJ05k0qRJAMyZM4fjjz+eDRs2sH37drKzs31/fy/dcaeKyKUi8oT7GiciOb4nM8YYk1CTJk3KlidPngzADTfcQLt27WjdujUvv/yyr+/vpXF8GpADzHA3/QYoUdXzfU1Wjaxx3BhTk2zbto3GjRuXrU+ZMoUrr7yybL1Ro0Z88skntG7d+ie9T6Ubx+P0VtWzVfVV93UO0PsnpTHGGFNlWVlZ1KnzY0vDlVdeiYiUrW/bto0JEyb49/4eypSISJfYioh0Bkp8S2SMMSapHTt2MGbMmHIVBUDFu0dTp07l3Xf9eVbbS8VxFTBPRF4TkdeBV4E/ePnlInKCiKwUkUIRuSbBfhGR2939y+Inj0p2rIj0EJEFIrJURBaKSB8vWYwxpib4z3/+w8MPP8yePXvSln3ggQd8yeCl4pgPdAUudV8HAm+lO0hEsnG67Q7BGVl3lIj8vEKxIe7v7gqMAe72cOwk4AZV7QH82V03xphaoUePHuXWe/XqRU5O4v5KY8aM8SWDl4rjHVXdparLVPV9Vd0FvOPhuD5AoaquVtXdwKPA8AplhgMz1bEAaCoibdIcq0CsVagJ8KWHLMYYUyMccsgh5SqKRYsWJb36ePXVV33JkGrO8dZAO6C+iPQEYjfUGgMNPPzudsAXcetFQF8PZdqlOfYyYK6ITMap+PonyT8G5yqGDh06eIhrjDHh980333DAAQewdu1atm3blrasH1I9OX488FugPc7847GKYytwrYffLQm2Vez7m6xMqmMvAi5X1SdF5HTgfuDYvQqr3gvcC053XA95jTEm9C644AJWrFiRtlzTpk25+uqrfcmQ6snxGcAMERmhqk9W4XcXAfvGrbdn79tKycrUTXHs2cDv3eX/ANOqkM0YYyKpUaNGnsr179+f5s2b+5LBy7DqVak0AN4DuopIJxGpC5wBzKpQZhYw2u1d1Q/Yoqrr0xz7JRAbjGUQ8GkV8xljTOQMH16xqTixjRs3+pbBy0ROVaKqxSIyDpgLZAPTVXWFiIx199+DM4f5UKAQ+AE4J9Wx7q++AGcMrTrATtx2DGOMqQ28DpleVFTkW4akQ46ISBv323/k2ZAjxpiaYteuXeTm5u61vU6dOuy3336sWrUKgIYNG6ZtPE8n2ZAjqa44potIM+A1YA4wX1WLf1IKY4wxP8nUqVMTbheRskoDYNy4cb5lSNU4PkREcoGBwKnAZBFZi1OJzFHVtb6lMsYYk9DmzZsTbo9/lqNu3bpcddVVvmVI2cahqjtxKwoAEemE8zT3nSLSWlVtuA9jjMmg+vXrpy1Tr169coMgVrdK/WZVXQP8C/iX29vJGGNMBvXp04fs7GxUlb59+/LOO85AHo0aNaJv374cf/zxHHTQQb5WHF6GHEnIHQrEGGNMBokIJSUllJaWsmnTprLt2dnZHHHEEXz88ccMGzaMww47jB07dviSwb8qyRhjTLXbvn172fLHH39ctrx582ZuuOGGsnGsVq5cyZdffkmXLl32+h0/VaWuOESkmYh0q/YUxhhjPBkyZAjTpk3jZz/7WcL9zZs35/DDD+faa6/1pdIAb1PHvgacjHN1shT4FnhdVa/wJZEP7DkOY0xNs27dOi699FJefPHFsltSbdu25Y033qi2CqMqz3HENFHVrSJyPvCAqv5FRJZVSypjjDFV0q5dOyZMmMDIkSNZs2YNS5cu5Z///Cf5+fm+v7eXiqOOO0fG6cB1Pucxxhjjwdq1a+nTpw87d+5k3LhxPPLIIxl7by9tHBNxxowqVNX33DnHbWBBY4wJ0NatW9m5cyfgzLvxwQcfcM8995TraeWXtG0cNYG1cZgqKS2FOXNg8WI47DA44QTIqnIPdmOq3cyZM1myZAkXXXQRBQUFbNu2jaFDh/LCCy9Uy++vchuHiLTCGZG2Y3x5VT23WpIZE0alpXDqqTArbiaAk0+Gp5+2ysOExujRoxk9ejQbN25k927n0br47rp+8dLG8SzwJvBfoMTfOMaExJw55SsNcNbnzIGhQ4PJZEwSzZs3Z/bs2cybN48xY/yfacJLxdFAVf2Zf9CYsFq8OPH2JUus4jChNGjQIAYNGpSR9/Jyzf28iNi/FFO7HHZY4u09e2Y2hzEh5KXi+D1O5bFTRLa5r61+BzMmUCec4LRpxDv5ZGe7MbVc2ltVquptZvTayHrd1FxZWU5D+Jw5zu2pnj3t/68xLk+DHIrIycAv3NXXVPV5/yJFhPW6qfmyspz2DGvTMKactH/hRORmnNtVH7qv37vbardUvW6MMaYG8/LVeChwnKpOV9XpwAnuttotVa8bY4ypwbzeU2kat9zEhxzRY71ujDG1lJeK42/AEhF5UERmAIuAv/obKwKs140xppby0qvqEXdOjt6AAFer6ld+Bws963VjjKmlklYcInKQqn4sIrF7MkXuz7Yi0lZVk9zkr0Ws140xphZKdcVxBTAGmJJgnwKZebbdGGNMqCStOFQ1NlLWEFXdGb9PRHJ9TWWMMSa0vNyQf9vjNmOMMbVAqjaO1kA7oL6I9MRpGAdoDDTIQDZjjDEhlKqN43jgt0B74Na47duAa33MZIwxJsRStXHMAGaIyAhVfTKDmYwxxoSYl+c4nhSRE4GDgdy47RP9DGaMMSacvMw5fg9Om8bRwDTgV8C7Pucyxphg2HQJaXkZVr2/qnYTkWWqeoOITAGe8juYMcZknE2X4ImXT2KH+/MHEWkL7AE6+RfJGGMCYtMleOJ1zvGmwN+BxcBnwKM+ZjLGmGDYdAmeeLlVNUlVdwFPisjzOA3kO9McY4wx0WPTJXji5YrjndiCqu5S1S3x24wxpsaw6RI8SVpxiEhrEemF++S4iBzmvgbi8clxETlBRFaKSKGIXJNgv4jI7e7+ZXEj8aY8VkR+5+5bISKTvP7HGmNMSrHpEl54AW680flpDeN78e3JcRHJBu4CjsMZkv09EZmlqh/GFRsCdHVffYG7gb6pjhWRo4HhQDdV3SUi+3j6LzXGGC9suoS0/HxyvA9QqKqrAUTkUZw/+PEVx3BgpqoqsEBEmopIG6BjimMvAm52211Q1W+qkM0YY0wVeWkcP0REDq640cOT4+2AL+LWi3CuKtKVaZfm2AOAASJyE04j/R9U9b2Kby4iY3DmE6FDhw5pohpjjPHKy42774Ht7qsE5/ZSRw/HSYJt6rFMqmPrAM2AfsBVwOMisld5Vb1XVQtUtaBVq1Ye4hpjjPHCy1hV5WYAFJHJwKwkxeMVAfvGrbcHvvRYpm6KY4uAp9zbW++KSCnQEvjWQyZjjDE/UVW6CjQAOnso9x7QVUQ6iUhd4Az2rnBmAaPd3lX9gC2quj7Nsc/gTlsrIgfgVDIbqvDfYYwxpgq8DHL4AT/eJsoGWgFpR8ZV1WIRGQfMdY+brqorRGSsu/8eYDYwFCgEfgDOSXWs+6unA9NFZDmwGzjbvfowxhiTAZLub66I7Be3Wgx8rarFvqaqZgUFBbpw4cKgYxhjTKSIyCJVLai43Usbx+ci0gynzaEOkC8iqGqSQV2MMcbUZF5uVf0PzoOAq/jxlpXitjMYY4ypXbw8x3E60EVVd/sdxhhjTPh5qTiWA00Be0LbGFN9bKa9yPJScfwNWOL2YtoV26iqJyc/xBhjUrCZ9iLNS8UxA7gF+AAo9TeOMaZWSDXTng0uGHpeKo4Nqnq770nMT2OX/SZKUs20ZxVH6HmpOBaJyN9wntyOv1Vl3XHDwi77TdTYTHuR5qXiiP2f7Be3zbrjhklUL/vtKqn2is20V/HLjs20FwleHgA8OhNBzE8Qxcv+iFwlqSrz58+nc+fOtGvXLug4qUWpIo7NtDdnjnOe9uwZ7rymHC9XHCbsonjZH5GrpGuvvZabb76ZZs2a8dFHH5Gfnx90pMQiUhGXYzPtRVZIzyhTKbHL/nhhv+xPdZUUEs8//zy33XYbAJs2baJTp04MGDCAzZs3BxsskVQVsTHVzCqOmiB22f/CC3Djjc7PMH/ThEhcJU2ZMoUdO3aUre/YsYP58+cze/bsAFMlEYGK2NQcXsaqygZOxJn1r6y8qt7qXyxTaVG77I9A42j37t157bXXytZFhNatWzNgwIDgQiUTgYrY1BxevpI+hzPIYQugUdzLmKqLwFXSGWecUbZ83XXXsX37dr744gv23XffFEcFJIq3K01keZmPY5mqdstQHl/YfBymql555RWKioo466yzqFMn5H1JYr2qrJeSqSbJ5uPwcla9KCKDfchkTKht3LiRlStXsn79el566aWg46QXu1153XXOT6s0jE+8fIVaADwtIlnAHkAAVdXGviYzJmCjRo0qV2G89NJLHHHEETRo0CDAVMYEz8tXkinA4UADVW2sqo2s0jC1wfbt28utT5s2jby8PI4//nhKS228T1N7eak4PgWWa7rGEGNqkClTptC0aVN69OhB586d6dOnDy+//DLgXHl07dqVlStXBpzSmGB4aRx/EOgMvEj5QQ4j0x3XGsdNZbz//vv06NEj4b6cnBz27NkDQLdu3Xj//fczmMyYzPopjeNrgFeAulh3XFMLtG7dmmbNmu21PatCY/OHH37I+vXrMxXLmNDwMsjhDZkIYkxY5Ofnc80113D11VcD0KVLF/r168czzzxTrt2juLiY7777jjZt2gQV1ZhAeHlyfB7OMOrlqKoNq25qrOnTp5ctT5o0iREjRgBQr149evXqxYYNGxg9ejSHHHJIUBGNCYyX7rh/iFvOBUYAxf7EMSZ427ZtY9WqVQA0atSIo446iqysLEpLS8nLy+Odd95BVXnrrbcCTmpMMLzcqlpUYdNbIvK6T3mMCVxOTg7NmjXj22+/pV+/fuTl5dG8eXM2bNjAli1biHUoadGiRcBJjQlG2sZxEWke92opIscDrTOQzZhA5ObmsmDBAmbOnMnjjz9etj548GBKSkoA50rksssuCzaoMQHxNOc4ThuH4NyiWgOc52coY4LWuXNnRIQrrriCdu3a0a1bN95+++2y/du2beNf//oX999/f4ApjQmGl1tVnTIRJKx27NjBAw88wIEHHsgxxxwTdByTIaWlpQwcOJC1a9cm3J+VlUW/fv1QVUQkw+mMCZaXW1WniUgjd/l6EXlKRJIM/l/zjB8/nksuuYTBgwezfPnysu27du3iqaeeYs2aNQGmM35QVY455piklUb9+vU57rjjGDNmDGeffXaG0xkTPC8PAP5JVbeJyJHA8cAM4G5/Y4XH7t27AeePSUlJCXv27OGUU06hQ4cOjBgxgoKCArZs2RJwSlOdtmzZUm4Cp4r27NnD3LlzAXj22WczlMqY8PDSxlHi/jwRuFtVnxWRCf5FCpfJkyfTokULGjRoQH5+PmeeeWa5PxYbN26kc+fOdO/enRdeeIH69esHmNZUh6ZNm3Lqqafy9NNPJ9xfXPxjb/TevXtnKpYxoeHlimOdiEwFTgdmi0g9j8fVCO+++y4333wzf/rTn2jTpg2z4qY6bdTIGXll48aNzJs3jxdeeCGomKaaeR06fezYsT4nMSZ8vFQApwNzgRNUdTPQHLjKz1Bh8p///If4gSB3795Nbm4uOTk5bNu2rWy7iFBQsNdYYCainnnmmbLlrKyshM9sHH744fzqV7/KYCpjwiFpxSEiDQFU9QdVfUpVP3XX16vqS/FlarLf/e53e/3R2LlzZ9kIqTGqSocOHTIZzfhk06ZN5cakKi0tZceOHXuV69q1ayZjGRMaqa44nhWRKSLyCxHJi20Ukc4icp6IzAVO8D9isA4++GD69u3rqezOnTt9TmMyoVmzZnu1Vf3www97lZs5cyaFhYWZimVMaCStOFT1GJzh1C8EVojIFhH5DngI58nxs1X1iczEDM6qVauYPXu2p7K5ubk+p6k5SkpKePzxx3nnnXeCjpLQsccem3B7/NDqDRo0oEmTJpmKZExopGzjUNXZqnqWqnZU1Saq2kJV+6vqTar6VaZCBunrr7/2XDbR7QyT2I033sjIkSM58sgjWbJkSdBx9nLppZcm3B4/Zezll19Oq1atMhXJmNCoNb2jqqpevXqeymVlZVG3bl2f09QcGzduBJw/xJs3bw42TAI9evRgwIABKcucdtppGUpjTLj4WnGIyAkislJECkXkmgT7RURud/cvi38i3cOxfxARFZGWfv439OrVy1OFUFpaWqmrk9pu4sSJXHvttUydOpWjjz466Dh7ufjii3nzzTdTlrn++uszlMaYcPHyAGCViEg2cBdwHFAEvCcis1T1w7hiQ4Cu7qsvzhPpfdMdKyL7uvsSjwlRjXbt2lU2ImoqdevWpX379n7HqTGaNGnCTTfdVG7bvHnzaNu2LQceeGBAqX7k5cuCtWmZ2srTFYeIHCki57jLrUTEy8CHfYBCVV2tqruBR4HhFcoMB2aqYwHQVETaeDj2H8B4EsxMWN2eeeYZTxXH7t27KSoq8jtOlYwZM4ZWrVpxzz33BB0lqUmTJjFo0CC6d+/Oxx9/HHQc7r77bm6//faE+7KyssjLy+PCCy/McCpjwsHLIId/Aa4G/uhuysHpWZVOO+CLuPUid5uXMkmPFZGTgXWq+n6a3GNEZKGILPz22289xE0sXRfb2MioPXv2ZJ999qny+/hlw4YN3HfffWzYsIFbb7016DhJxbq17tq1KxQVcKNGjWjZsmW59cMPPxxwbktu3749aQO6MTWdl1tVpwI9gcUAqvplbLTcNBKNNV3xCiFZmYTbRaQBcB0wON2bq+q9wL0ABQUFVb4yGTZsGHl5eWzfvp327dvTpUsXhgwZQseOHVm7di3HHnssLVu2JD8/P5SN4y1atGDYsGE8//zzjB49Oug4SU2cOJFFixaxdetWmjZtGnQcwBkZOeb777/n/PPPL9d9+KOPPuLpp5/m1FNPDSKeMYHxUnHsVlUVEQWIfxgwjSJg37j19sCXHsvUTbK9C9AJeN/9pt8eWCwiffzqHty8eXO+//57SktLy/XhD5t169YxadIk+vfvz8iRIwGne/Add9zBxo0bOf/887n88ssDTpncokWLWLx4MQBXXXUV8+bNCzSPqpZ7elxV+eqrvU8xGxnZ1EqqmvIF/AGYCqwGLgDeAS71cFwd95hOOBXB+8DBFcqcCLyIc4XRD3jX67Fuuc+Alumy9OrVS2s696pKAc3OztbevXtrkyZNyrYBOmTIkKBjaklJiY4fP15HjRqlRUVFqqo6Y8aMcjnHjx8fcErHggULtHXr1mW5PvroI73rrrv0xBNP1EGDBunNN9+sJSUlQcc0xjfAQk3wN9XLDICTReQ4YCtwIPBnVX3Zw3HFIjIOZ4DEbGC6qq4QkbHu/nuA2cBQoBD4ATgn1bHp3rM2+/7778uWS0pKeO+99/Yqs2zZssCvnGbPns2kSZMAp/2lSZMmzJ8/v2x/48aNueWWW4KKV6a4uJiVK1cycuRIHn30UQYOHEjXrl056KCDuPjii4OOZ0ywEtUmWv5b/S1etoX5VRuuOD766CP92c9+Vu6be6LXgw8+GHjOunXrKqBZWVkKaF5enuK0YemMGTO0sLBQS0tLA8155plnln1mv//97wPNYkxQSHLF4eWr53EJtg35adWVqW4iwoknnpj2SffYjIZBOeCAA8queGI90mKZ8/LyGDNmDPvvvz/nnXdeYBkBVq5cmXDZGJOicVxELgIuBjqLyLK4XY2At/wOZipn1KhRnsZ88jpBkV+ysrLo0qULK1asIDc3l+3bt5cNPxJ/u+2///1vUBEBZ+Tb008/nY0bN5YNLVJSUsJtt93Gnj17OPzww3nuuecYPXo0hx56aKBZjcm4RJchzhUKTYCOwCPAfnGv5smOCeurNtyqGjBgQNrbVID27t076Ki6ceNGPe2005JmFBF97LHHgo6pgwYNKutssHTpUp0+fXpZxnr16imgXbp0CTqmMb6hsreqVHWLqn6mqqNU9XNgh/uPpqGI2IxFIeN1rvMwzB+xdevWpPN5g/NlJt04UZkQm6yrpKSE3r17s2HDhrJ9sSu3sDxzYkwmeXlyfJiIfAqsAV7H6QL7os+5TCV5fTp+06ZNLFu2LH1Bn2zfvp2xY8dSXFycstydd94Z+BPkxx33Y/Penj17uOuuu7jlllt45plnWLx4MdOmTbN55k2t5KVx/EacZyw+UdVOwDFYG0fo3HfffZ7KiQgNGwY34+/EiROZM2dO2nJZWVmer6L8cvHFF9OzZ08aN24MwOeff87cuXMZPnw4HTt25LzzziM/Pz/QjMYEwUvFsUdVvwOyRCRLVecBPfyNZSqrUycv4046FUfnzp19TpNc7I9wOl26dNlrrvdMe/rpp1myZAlbt24tqyAGD0472o0xNZ6XimOziDQE3gD+LSK3AanvM5iM++tf/+qp3KBBg3xOkto111zD5ZdfnnZI8sLCwsDncI/vuhyrbN9/P+XYmsbUCl7GqhoO7AQuB87C6W11g5+hTOWlu+LIz89n0qRJnq9M/DJ//nz+8Y9/pC2Xm5tb9pxHUC688EJUlZycHK644goAnnzyyUAzGRMKibpapXoBBwH3Vfa4IF+1oTuuquqYMWM0JycnZXdcEdG33347sIytWrVK22X4oosu0g8++CCwjIlMmTJF27Ztq+edd17QUYzJGCrbHVdEuonISyKyXERuFJF8EXkS+C/wYbLjTDC+++473n77bbKzs8u2tWjRgq5du9KxY8eybarKpk2bAkjo8DKa7PTp0z1NnpVJvXr1Yv369dx///1MmzYt6DjGBCpVG8d9wMPACOBbnPk4VgP7q2r6ew0mo958802WL19erl3gu+++49NPPy3XID1y5EiGDh0aREQAfv3rX6cts2vXLm688cYMpPFu3bp1sStu1q1bF3AaY4KVqo2jnqo+6C6vFJE/ANeoari+ChoAjjrqKAoKCli7di1XX301++23H5deeilffvkl55xzDgUFBWzZsoUTTzwx0Jx33nknjz76KD/88EPKcr17985QIm9GjhzJ6tWr2bJlC1deeWXQcYwJVKqKI1dEevLjbHzfA93EbbFU1cV+hzPeNWvWjPfee4958+YxfPhwWrZsyeOPP07btm0DbxCP52U+8fbt23PZZZf5H6YSsrOzuf7664OOYUwopKo41gPxk1R/FbeuQLD9Ok1CjzzyCNu2bWPbtm0cffTRLFiwIFQVx+233572aqOoqIjTTjuNZ599NkOpjDGVkWqsqqNTvKzSCKnf/va3ZU+G79mzJ3TPHZx00knUqVMn7WRSn3zySYYSGWMqK7yTaJsq6d+/P4WFhYwYMYJzzjmHM844I+hI5Zx00kncdNNNZY3kBx10ULmeYDFhmAXQGJOYlwcATcTk5+fzxBNPBB0joYEDB7JgwQLAua22evVqrrvuOgA6duzIL37xC8aMGcMRRxwRZExjTApWcZiM+uyzz8qWv/rqq7L5x7OysnjllVcCHUfLGOONp1tVItJORPqLyC9iL7+DmZrpkUceoWvXrgwdOpSLLrqIAQMGANChQwfOPfdc5s6dG3BCY0w6EnuoKWkBkVuAkThPi8ee4VBVPdnnbNWmoKBAFy5cGHQMk0BxcTFPPPEEo0aNApzRe19//fWyCsUYExwRWaSqBRW3e7lVdQpwoKruqvZUptarU6cORx99NPXr12fHjh2oKsuWLbOKw5gQ81JxrAZyAKs4jC/y8/NZunQpl112Ga1bt+bss88OOpIxJgUvFccPwFIReYW4ykNVL/Utlal19uzZQ6tWrRg4cGCgMxQaY9Lz0saR8Oufqs7wJZEPrI0j/I488kjeeustRISioiLatm0bdCRjar0qt3Go6gwRqQsc4G5aqap7qjugqd06d+7MW2+9RatWrWjUqFHQcYwxKaTtjisiA4FPgbuAfwGfWHdcU92mTZvG7NmzWbx4sVUcxoSclzaOKcBgVV0JICIHAI8AvfwMZmqXunXrMmTIkKBjGGM88PIAYE6s0gBQ1U9welkZY0Li66+/ZvXq1QCMHz+eAw88kMceeyzgVMmVlpYGHSGtOXPm8MQTT5CuHbg28lJxLBSR+0VkoPu6D1jkdzBjTHo7d+7kxRdfZL/99qNLly6ccsop/P3vf+eTTz5hwoQJQcfbi6oyfPhwcnJyuPnmm4OOk9Ts2bMZMmQIp512GlOnTuXjjz/mN7/5DVOnTg06Wih4qTguAlYAlwK/x3mCfKyfoYwx6e3cuZNDDjmEoUOHsmuX01P+2WefpU4d5w50fn4+w4YN48knnwwyZjlr1qxh1qxZlJaW8uCDDwYdJ6ktW7aUWx43bhwPPfQQY8eOZeXKlSmOrB289KrahTOB063pyhpjMufjjz9m1apVe20vLi4G4PXXXwdg7ty57Nq1C3fyzsDMnz+/rB0rLy8vdLM8xnz++efcf//99OjRg127djFhwgRycn68O//aa69x4IEHBpgweEkrDhF5XFVPF5EPcGb8K0dVu/mazBiT0vbt2z2VKy0tZevWrTRp0sTnRIkVFxdzwQUXMHPmzLK2jby8PC655BLeeecdJk2axKhRo2jRogWPPfZY2km+/LR582aGDRvGBx98UG77zp07y5bbtGmT6Vjho6oJX0Ab9+d+iV7Jjgvjq1evXmpMTbNq1SrF+VKX9jV+/PjAcv73v/8tlyU7O7vcuoiULV9yySWB5VRVnTBhQlmWrKyscjlvuOEGff755wPNl2nAQk3wNzXV1LHr3cWLVfXz+BdwcbXUWsaYKsvNzfVcNsheTJs3by633r1793LrGtdrKcipjhcvXswdd9xRtl5aWlruFtWECROYO3cuL7/8chDxQsXLNeFxCbZZh3tjAvanP/3Jc9nWrVv7mCS1jRs3lltfvHhx0rJffvml33GSGjduHN999125bXv2/DhIhqpyxx13MHjwYB5++OFMxwuVpBWHiFzktm8cKCLL4l5rgGWZi2iMSeSss87yXPbuu+/2MUlqgwYN8lw2Pz/fxySpxV/5pPPss8/6mCT8UvWqehh4EfgbcE3c9m2qujHxIcaYTOnbty8i4ukP3tdff52BRIndd999nsv269fPxySpjR07lgULFngqO3r0aJ/ThFuqNo4tqvqZqo5y2zV24DQSNRSRDhlLaIxJKC8vj6VLl5KXl7fXvoo9kzp0CO6f7ODBgz2Xfffdd31MktqSJUs8lWvZsiVHHnmkz2nCzcsgh8NE5FNgDfA68BnOlUhaInKCiKwUkUIRuSbBfhGR2939y0TksHTHisjfReRjt/zTItLUSxZjaqJu3brx1FNP0bdvX0aMGEFOTg4///nPWbFiRdl0vLm5uTz00EOBZRw0aBB//OMfPZXt27evz2mSO/744z2V27BhQ62/VZW2KyvwPtACWOKuHw3c6+G4bGAV0Bmo6/6en1coMxSnEhKgH/B/6Y4FBgN13OVbgFvSZbHuuKa2KC4uLlveuXOnzpo1S7/44osAEzmGDh3qqdtw9+7dA81Zp04dTznPOuusQHNmCpXtjhtnj6p+B2SJSJaqzgN6eDiuD1CoqqtVdTfwKDC8QpnhwEw34wKgqYi0SXWsqr6kqsXu8QuA9h6yGFMrZGdnly3Xq1ePYcOG0b598P9ECgsLy63Xq1cvYbmgHlIEZ1DD2FP36QwdOtTnNOHmpeLYLCINgTeAf4vIbYCXT7cd8EXcepG7zUsZL8cCnEuS22YiMkZEForIwm+//dZDXGOMX6644opy63Xr1k1YbsSIEZmIk9Crr77quez+++/vY5Lw81JxDMeZd/xyYA7OLaSTPByXaGCcit0/kpVJe6yIXIdTgf070Zur6r2qWqCqBa1atfIQ1xjjl3322YdDDjkEgJycHJ577rmEz5b06dMn09HKXHjhhZ7L3nLLLT4mCT8vFcefVbVUVYtVdYaq3g5c7eG4ImDfuPX2QMWne5KVSXmsOw/6ScBZ7n04Y0xILVy4kF/+8pcsX76ccePGsXv3bo466ig+++wzCgp+nM56n3324bDDDkvxm/zVpk2bhONkJdo2fHjFu+61i59Pjr8HdBWRTu6c5WcAsyqUmQWMdntX9QO2qDPUSdJjReQEnIrrZFX9wUMOY0yA6tSpUzYyb/wQHvXq1ePdd9/lyiuv5Mwzz+STTz5JegsrExo0aMDFF5cfTalp06aUlpbSsGFDsrKy6NmzJ88991ytf44j1ei4F+GMSdVZROKfFG8EvJXuF6tqsYiMA+bi9JKarqorRGSsu/8eYDZOz6pCnNth56Q61v3VdwL1gJfdk3GBqtr8IMaEVI8ePZgzZw6FhYWce+655faJCJMnTw4o2d4OPfTQsuUWLVqUDUHSpEkTvv/+e5YsWRJo5RYWkuxOj4g0AZpRA54cLygo0IULFwYdwxgTciUlJdx+++18/fXXZe0YIsJZZ53FQw89RG5uLkuXLq0183GIyCJVLdhre7omgmRPiavq2mrK5jurOIwxlbF7924OOOAAPv/8cwBuvfVWCgoKaNOmTa3qUZWs4kg7AyDwAj/2dMoFOgErgYOrNaExxoRE3bp1eeihhzj22GPJysqif//+gT7VHjZepo49NH7dHRbEe781Y4yJoCOPPJKioiJEhBYtWgQdJ1S8XHGUo6qLRaS3H2GMMSZMWrZsGXSEUEpbcYhI/COfWcBhgD2KbYwxtZSXK45GccvFOG0eT/oTxxhjTNh5aeO4IRNBjDHGREOqBwArPuVdjqqeXP1xjDHGhF2qK47DcUaofQT4PxIPPGiMMaaWSVVxtMYZp2oUcCZO28YjcUN/GGOMqYVSzTleoqpzVPVsnNn5CoHXROR3GUtnjDEmdFIOOSIi9YATca46OuKMUDtdVddlJF01EZFvgc+r+de2BDZU8+/0g+WsPlHICJazutXmnPup6l4TGqUa5HAGcAjODHuPquryag4UaSKyMNEYLmFjOatPFDKC5axulnNvqdo4fgNsBw4ALo2Np4/TSK6q2tjnbMYYY0IoacWhql4meTLGGFPLWOVQdfcGHcAjy1l9opARLGd1s5wVpJ2PwxhjjIlnVxzGGGMqxSoOY4wxlWIVhzHGmEqp9EROtZmI5APtcKbS/VJVvw44Uloi0lxVNwadI5Eofp5hFsXPM6znZxQ/y0yyxnEPRKQHcA/QBIg9Nd8e2AxcrKqLg0lWnohcr6o3uss/B54BcnCevRmpqv8XYLwyUfk8AUQkG2f0hI7EfdFS1VuDylRRVD7PKJyfUfksY4I6P63i8EBElgIXVjyxRaQfMFVVuwcSrAIRWayqh7nLLwB3quqLItIH+Keq9g82oSMqnyeAiMwGdgIfAKWx7WGapyYqn2cUzs+ofJYxQZ2fdqvKm7xE34ZUdYGI5AURyIO2qvoigKq+KyL1gw4UJ0qfZ3tV7RZ0iDSi9HnGhPX8jNpnGcj5aRWHNy+635Bm4sxRArAvMBqYE1iqvXV2J+ASoL2INFDVH9x9OQHmqigqnyc4WQer6ktBB0khKp9nFM7PqHyWMYGcn3aryiMRGQIMx2kwE6AImKWqswMNFkdEjqqwaZGqfu829P1KVe8KIlciUfg8AUTkVOAhnB6IewjpWG1R+Dyjcn5G4bOMCer8tIrDmBREZDVwCvCB2j8WEzJBnZ/2HIcHIpItIheKyP+ISP8K+64PKldFFXIeUWFfmHI2EJHxInKViOSKyNkiMktEJolIw6DzVfApsDzMlYadn9UnYucmBHR+2hWHByIyDWgAvIsz3PzrqnqFu6+sp0jQIpTzcZz7x/WBA4GPgMeBYUBrVf1NgPHKEZEHgc4489Lsim0PWXfcqPx/D33OKJ2bENz5aRWHByKyLNZzQUTqAP/CmW1rFLBAVXsGmS8mQjmXqmoPcSZ5WQ+0UVV1198PUy8mEflLou0h644blf/voc8ZpXMTgjs/rVeVN3VjC6paDIwRkT8DrwJhunyNSk7AacETkdmxy2x3PWzfZP5XVVcHHSKNqPx/j0rOqJybEND5aW0c3iwUkRPiN6jqROABnCc2wyJKORsCqOq5sY0i0gXYFliqxB4UkVUi8qiIXCwihwYdKIEo/X8Pe84onZsQ0Plpt6qqkYgcp6ovB50jnTDnFBGJfcsLS04RqQv0BgYCFwINVbV5oKGqICyfZzphzRnGcxOCOT+t4qhGYWngS8dyVirDkcAA99UUWAq8qaqPBBirSsLweXoRhZxhyRjU+WltHNVLgg7gkeX07nVgIfA3YLaq7g44z08Rhs/TiyjkDEvGQM5PqziqV1Qu3yyndy2AI4BfAJeKSCnwjqr+KdhYVRKGz9OLKOQMS8ZAzk+rOIxJQVU3u0/n7oszvHZ/wjOukqnlgjo/reKoXp8FHcCjz4IO4NFnQQcQkVXASmA+zjwN50T4dtVnQQfw6LOgA3jwWdABILjz0xrHf6Iw9a5IxXJWjYhkqWpp+pLhFLbPM5ko5AxjxqDOT3uO46e7P+gAHlnOqukoIs+JyLci8o2IPCsinYMOVQlh+zyTiULOMGYM5Py0W1UeiDOHQMJdOI1ToWA5ffEwcBdwqrt+BvAI0DewRBVE5fOMQs4oZKwgkPPTblV5ICKbgF8D31fcBTymqvmZT7U3y1n9ROT/VLVvhW0LVLVfUJkqisrnGYWcUcgYL6jz0644vFkA/KCqr1fcISIrA8iTjOWsfvNE5BrgUZwumCOBF0SkOYCqbgwynCsqn2cUckYhY7xAzk+74jAmBRFZk2K3qmqU2jtMDRPU+WkVRxWJSEvguzBP8AOW029h7GkTIyLZwBmq+u+gs6QShZxRyJiIX+en9aryQET6ichrIvKUiPQUkeXAcuDriqN9BslyBuKWoAOISGMR+aOI3Ckig8XxO2A1cHrQ+WKikDMKGSvJl/PTrjg8EJGFwLVAE+BeYIiqLhCRg4BHwjABDVjOIIjIkqDzisizwCbgHeAYoBnO3Be/V9WlAUYrJwo5o5CxMvw6P63i8EDcWcHc5Y9U9Wdx+wL/wxFjOTNPQjBKqoh8oKqHusvZwAagg6qGav6IKOSMQsbK8Ov8tFtV3sQ/mbmjwr4w1byWs3baE1tQ1RJgTUj/0EUhZxQyBs6uODwQkRJgO05f7vrAD7FdQK6qhmLQO8uZeSLylKr+MuAMsc8Tyn+mgtOzpnFQ2eJFIWcUMlaGX+enVRzGVFKYe1IZk4nz025VeSAig+KWO1XYF+i3zXiWM2PCOGZRGRFpJyId3FdoH/KNQs4oZEzA9/PTrjg8iG9gqtjYFIbG0RjLWX0k9ZhFg1Q1L5N5UhGRPwI5qjrRXV8LbMGZl2GGqv4tyHwxUcgZhYwQ/PkZlRo0aJJkOdF6kCxn9RlA8jGL+mQ+Tkqn4eSN+U5Ve7q9gl7HmVY0DKKQMwoZIeDz0yoObzTJcqL1IFnO6hOpMYtUdXvc6m3uthIRqR9QpISikDMKGQn4/LRbVR6IyGbgDZzafIC7jLt+pKo2CyhaOZazdhKRT4CDVXVPhe31gOWq2jWYZOVFIWcUMoaBVRweiMhRqfYnqvWDYDn9JSEdT0tE/gq0Bsap6g/utjzgTuArVf1jkPliopAzChlTkQyNqWUVx08kIkeo6ltB50jHclY6Rz/gZmAj8D/A/wItcXoijlbVOQHGK8f9Y3ETcD7wubu5A07vmutVtTiobPGikDMKGQFEpDFwCdAOmAW8DIwD/gAsVdXhvr6/VRzpuSfT6Tj/k+ao6nIROQlnvKX6YRkiw3JWnyiOp+Xeg9/fXS1U1YpP5YdCFHKGPWPgY2qpqr3SvIAHgVdwelS8CjwAfAycEnQ2y+lbxqVxyx9V2Lck6HwV8oyPWz6twr6/Bp0vSjmjkNHN8kHccjZOJdIoU+9vVxweiDPsdzdVLRWRXJyBz/ZX1a8CjlaO5aw+UXjWJCYqWaOQMwoZIfhs1h3Xm92qWgqgqjtF5JMw/ZGLYzmrT3cR2Yo7XpG7jLueG1yshKLwXAxEI2cUMsKP5yeUP0czMqaWVRzeHCQiy9xlAbrEraOq3YKJtRfLWU1UNTvoDJUQhediIBo5o5Ax8PPTblV5ICJdgXzgiwq79gO+VNXCzKfam+WsPiIySFVfdZc7qeqauH2/VNWngktXnkRktOEo5IxCxkREpB1OWwc4/4Z87f1lgxx68w9gq6p+Hv/COan+EXC2eJaz+kyOW36ywr7rMxkkHVXNVtXGqtpIVeu4y7H10Pyhi0LOKGQEZ0wtEflz3KZ3gBeAl4Cr/H5/qzi86aiqyypuVNWFQMfMx0nKclafqNzrNrXTacCUuPXv1Jm58GDgRL/f3CoOb1I1hoZp/BrLWX0ica/b1F6aZEwtMvBvyCoOb94TkQsqbhSR84BFAeRJxnJWn84iMktEnotbjq13SnewMT5rKCJlt85U9UEoG1PL91kKrXHcAxHJB54GdvPjH7YCnCc1Tw1LV1LLWX2iOp6WqR2CHlPLKo5KEJGjgUPc1RWxXjdhYzn9FZbxtEztFfSYWlZxGJNAFMbTMiaoMbWsjcOYxO7H+TbXArhdRB7A6aI7ySoNEzQRGQ/gVhQHqeoHsUrDvY3l7/vbFYcxe4vCeFqm9gp6TC274jAmsXLjaQFhHE/L1F6BPmdkY1UZk1jox9MytVqgzxnZrSpjEojCeFqm9gp6TC274jAmsX8A17pjaJURkVbuvmGBpDKG4EfHtTYOYxKLwnhaxgTCKg5jEovCeFrGBMIqDmMSi8J4WsYEwhrHjUkgCuNpGRMUqziMSSGq42kZ4yerOIwxxlSKtXEYY4ypFKs4jDHGVIpVHMYYYyrFKg5jjDGV8v8TOAtETYYhfQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "background_mr_df = aggregate_bkg_mutation_rate_df(100, label_0base=False)\n",
    "measured_mr_df = measured_mutation_rate_df()\n",
    "plot_mutation_rate(background_mr_df, measured_mr_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd3ed41",
   "metadata": {},
   "source": [
    "# Pipeline - NoFlanks NoFootprints - Harshit Backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e5253e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"{OUTPUT_SX}/Background_Intersection\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3377a626",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# BACKGROUND INTERSECTIONS FOR No Flank FILES\n",
    "for i in {1..100}\n",
    "do\n",
    "    for tf in CREB1 EGR1\n",
    "    do\n",
    "        for min_length in 20 25 30\n",
    "        do\n",
    "            bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/NC_${tf}_${min_length}_Counts_${i}.bed\"\n",
    "        done\n",
    "        bedtools intersect -a \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/${tf}_PWM_Background_Counts_${i}.bed\"\n",
    "    done\n",
    "done\n",
    "\n",
    "# Intersect PWM (No-Footprint - No Flank) and NonConsensus Calls with Mutations\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    for min_length in 20 25 30\n",
    "    do\n",
    "        bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/NonConsensus_${tf}_in_Promoters_${min_length}_Unique_Mutation_Counts.bed\"\n",
    "    done\n",
    "    bedtools intersect -a \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/${tf}_PWM_Promoter_Unique_Mutation_Counts.bed\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "8059c1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregate counts - 1 index\n",
    "N = 100\n",
    "mr_dict = defaultdict(list)\n",
    "for tf in (\"CREB1\", \"EGR1\"):\n",
    "    for min_length in NC_LENGTH_TUP:\n",
    "        for i in range(1, N+1):\n",
    "            mr = mutation_rate(f\"../Figure_SX/Background_Intersection/NC_{tf}_{min_length}_Counts_{i}.bed\")\n",
    "            mr_dict[f\"{tf}_{min_length}\"].append(mr)\n",
    "    for i in range(1, N+1):\n",
    "        mr = mutation_rate(f\"../Figure_SX/Background_Intersection/{tf}_PWM_Background_Counts_{i}.bed\")\n",
    "        mr_dict[f\"{tf}_pwm\"].append(mr)\n",
    "# Organize DataFrame of background mutation counts\n",
    "background_df = pd.DataFrame.from_dict(mr_dict, orient='index').T\n",
    "background_df.melt().to_csv(f\"{OUTPUT_SX}/Background_Rates.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08c68f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "background_mr_df = aggregate_bkg_mutation_rate_df(100, label_0base=False)\n",
    "measured_mr_df = measured_mutation_rate_df()\n",
    "plot_mutation_rate(background_mr_df, measured_mr_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6d2e69b",
   "metadata": {},
   "source": [
    "# No Flanks, Yes Footprints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "658be810",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"{OUTPUT_SX}/Background_Intersection\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0cd35c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# GENERATE FLANK FILES for PWM CALLS in footptints\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    bedtools intersect -u -a \"../Figure_SX/${tf}_PWM_promoter_calls_unique.bed\" -b \"../Figure_SX/consensus_footprints_and_collapsed_motifs_simple_hg19.bed\" > \"../Figure_SX/${tf}_PWM_promoter_calls_infootprints_unique.bed\"\n",
    "done\n",
    "\n",
    "# BACKGROUND INTERSECTIONS FOR FLANK FILES\n",
    "for i in {1..100}\n",
    "do\n",
    "    for tf in CREB1 EGR1\n",
    "    do\n",
    "        for min_length in 20 25 30\n",
    "        do\n",
    "            bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/NC_${tf}_${min_length}_Counts_${i}.bed\"\n",
    "        done\n",
    "        bedtools intersect -a \"../Figure_SX/${tf}_PWM_promoter_calls_infootprints_unique.bed\" -b \"../Figure_SX/Promoter_Backgrounds/Background_${i}.bed\" -c > \"../Figure_SX/Background_Intersection/${tf}_PWM_Background_Counts_${i}.bed\"\n",
    "    done\n",
    "done\n",
    "\n",
    "# Intersect PWM (No-Footprint - FLANK) and NonConsensus Calls with Mutations\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    for min_length in 20 25 30\n",
    "    do\n",
    "        bedtools intersect -a \"../Figure_SX/NonConsensus_${tf}_promoter_calls_${min_length}_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/NonConsensus_${tf}_in_Promoters_${min_length}_Unique_Mutation_Counts.bed\"\n",
    "    done\n",
    "    bedtools intersect -a \"../Figure_SX/${tf}_PWM_promoter_calls_infootprints_unique.bed\" -b \"../Figure_SX/skcm_mutations_promoter2000.bed\" -c > \"../Figure_SX/${tf}_PWM_Promoter_Unique_Mutation_Counts.bed\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85c0f3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "background_mr_df = aggregate_bkg_mutation_rate_df(100, label_0base=False)\n",
    "measured_mr_df = measured_mutation_rate_df()\n",
    "plot_mutation_rate(background_mr_df, measured_mr_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
