{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bbe840ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "PWM_CORE_ONLY = False\n",
    "PWMS_TABLE_S3A = \"../Table_S3/Table_S3A_PWM_Probabilities.csv\"\n",
    "OUTPUT_SX = \"../Figure_S5H\" \n",
    "TF_TUP = (\"CREB1\", \"EGR1\", \"MITF\", \"ETS1\")\n",
    "GENOMIC_INIT_DIR = f\"{OUTPUT_SX}/Genomic_Init_Files\"\n",
    "promoter_fasta_file = f\"{OUTPUT_SX}/Promoter2000.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7489e0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(OUTPUT_SX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ea4b15dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query PWMs from Supplementary Table 3A and save to files\n",
    "pwm_df = pd.read_csv(PWMS_TABLE_S3A)\n",
    "pwm_percentages = []\n",
    "for protein in TF_TUP:\n",
    "    pwm = pwm_df[pwm_df[\"Experiment\"] == f\"{protein}_NonUV\"]\n",
    "    # Transpose to wide format and convert to numpy array\n",
    "    pwm = pwm[[\"A\", \"C\", \"G\", \"T\"]]\n",
    "    pwm = pwm.rename(columns={\"A\":\"A:\", \"C\":\"C:\", \"G\":\"G:\",\"T\":\"T:\"})\n",
    "    pwm = pwm.T\n",
    "    if PWM_CORE_ONLY:\n",
    "        pwm = pwm[pwm.columns[3:-3]]\n",
    "    with open(f\"{OUTPUT_SX}/{protein}_NonUV_pwm.txt\", 'w') as write_obj:\n",
    "        write_obj.write(f\"{protein}\\n\")\n",
    "    pwm.to_csv(f\"{OUTPUT_SX}/{protein}_NonUV_pwm.txt\", sep='\\t', header=None, mode='a')\n",
    "    \n",
    "# Query PWMs from Supplementary Table 3A and save to files\n",
    "pwm_df = pd.read_csv(PWMS_TABLE_S3A)\n",
    "pwm_percentages = []\n",
    "for protein in TF_TUP:\n",
    "    pwm = pwm_df[pwm_df[\"Experiment\"] == f\"{protein}_UV\"]\n",
    "    # Transpose to wide format and convert to numpy array\n",
    "    pwm = pwm[[\"A\", \"C\", \"G\", \"T\"]]\n",
    "    pwm = pwm.rename(columns={\"A\":\"A:\", \"C\":\"C:\", \"G\":\"G:\",\"T\":\"T:\"})\n",
    "    pwm = pwm.T\n",
    "    if PWM_CORE_ONLY:\n",
    "        pwm = pwm[pwm.columns[3:-3]]\n",
    "    with open(f\"{OUTPUT_SX}/{protein}_UV_pwm.txt\", 'w') as write_obj:\n",
    "        write_obj.write(f\"{protein}\\n\")\n",
    "    pwm.to_csv(f\"{OUTPUT_SX}/{protein}_UV_pwm.txt\", sep='\\t', header=None, mode='a')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5407ae1",
   "metadata": {},
   "source": [
    "### UV called sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "443097ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +CREB1 of width 14.\n",
      "Using motif -CREB1 of width 14.\n",
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +EGR1 of width 15.\n",
      "Using motif -EGR1 of width 15.\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$promoter_fasta_file\" \"$OUTPUT_SX\" \"$GENOMIC_INIT_DIR\"\n",
    "\n",
    "# Convert the pwm files to meme format using uniprobe2meme\n",
    "# pwm text file, pwm meme output\n",
    "\n",
    "# Call sites using FIMO\n",
    "# --parse-genomic-coord - Used to get coordinates as genomic locations\n",
    "# -bfile - Background file for calls\n",
    "# --text - Output is in tabular format\n",
    "\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    # Convert the pwm files to meme format using uniprobe2meme\n",
    "    uniprobe2meme \"${2}/${tf}_UV_pwm.txt\" > \"${2}/${tf}_UV_pwm.meme\"\n",
    "    # Call sites using FIMO\n",
    "    fimo --parse-genomic-coord --bfile \"${3}/Promoter2000_Merged_markov_bkg.txt\" --text  \"${2}/${tf}_UV_pwm.meme\"  ${1} > \"${2}/${tf}_PWM_UV_promoter_calls.txt\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64415a72",
   "metadata": {},
   "source": [
    "### Non-UV Called Sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aea76d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +CREB1 of width 14.\n",
      "Using motif -CREB1 of width 14.\n",
      "Converted 1 motifs.\n",
      "Skipped 0 motifs.\n",
      "Warning: text mode turns off computation of q-values\n",
      "Using motif +EGR1 of width 15.\n",
      "Using motif -EGR1 of width 15.\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$promoter_fasta_file\" \"$OUTPUT_SX\" \"$GENOMIC_INIT_DIR\"\n",
    "\n",
    "# Convert the pwm files to meme format using uniprobe2meme\n",
    "# pwm text file, pwm meme output\n",
    "\n",
    "# Call sites using FIMO\n",
    "# --parse-genomic-coord - Used to get coordinates as genomic locations\n",
    "# -bfile - Background file for calls\n",
    "# --text - Output is in tabular format\n",
    "\n",
    "for tf in CREB1 EGR1\n",
    "do\n",
    "    # Convert the pwm files to meme format using uniprobe2meme\n",
    "    uniprobe2meme \"${2}/${tf}_NonUV_pwm.txt\" > \"${2}/${tf}_NonUV_pwm.meme\"\n",
    "    # Call sites using FIMO\n",
    "    fimo --parse-genomic-coord --bfile \"${3}/Promoter2000_Merged_markov_bkg.txt\" --text  \"${2}/${tf}_NonUV_pwm.meme\"  ${1} > \"${2}/${tf}_PWM_promoter_calls.txt\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88032979",
   "metadata": {},
   "source": [
    "### Get UV PWM Calls for all Non-UV sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cce0e8ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66998092</td>\n",
       "      <td>66998105</td>\n",
       "      <td>-</td>\n",
       "      <td>10.87</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATATGACGTACTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66999224</td>\n",
       "      <td>66999237</td>\n",
       "      <td>+</td>\n",
       "      <td>12.47</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGTGACGTCAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66998092</td>\n",
       "      <td>66998105</td>\n",
       "      <td>-</td>\n",
       "      <td>10.87</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATATGACGTACTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66998092</td>\n",
       "      <td>66998105</td>\n",
       "      <td>-</td>\n",
       "      <td>10.87</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATATGACGTACTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66999224</td>\n",
       "      <td>66999237</td>\n",
       "      <td>+</td>\n",
       "      <td>12.47</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGTGACGTCAGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4435</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>45772988</td>\n",
       "      <td>45773001</td>\n",
       "      <td>-</td>\n",
       "      <td>10.80</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAGTGACGCAGTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4436</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>46002011</td>\n",
       "      <td>46002024</td>\n",
       "      <td>+</td>\n",
       "      <td>11.59</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGATGACGTCATAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4437</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>46002011</td>\n",
       "      <td>46002024</td>\n",
       "      <td>-</td>\n",
       "      <td>11.50</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATATGACGTCATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4438</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>45874430</td>\n",
       "      <td>45874443</td>\n",
       "      <td>+</td>\n",
       "      <td>10.33</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAATGACGCAGGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4439</th>\n",
       "      <td>CREB1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>47741634</td>\n",
       "      <td>47741647</td>\n",
       "      <td>-</td>\n",
       "      <td>10.19</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAGTGACGCAGGTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4440 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     motif_id motif_alt_id sequence_name     start      stop strand  score  \\\n",
       "0       CREB1        CREB1          chr1  66998092  66998105      -  10.87   \n",
       "1       CREB1        CREB1          chr1  66999224  66999237      +  12.47   \n",
       "2       CREB1        CREB1          chr1  66998092  66998105      -  10.87   \n",
       "3       CREB1        CREB1          chr1  66998092  66998105      -  10.87   \n",
       "4       CREB1        CREB1          chr1  66999224  66999237      +  12.47   \n",
       "...       ...          ...           ...       ...       ...    ...    ...   \n",
       "4435    CREB1        CREB1         chr21  45772988  45773001      -  10.80   \n",
       "4436    CREB1        CREB1         chr21  46002011  46002024      +  11.59   \n",
       "4437    CREB1        CREB1         chr21  46002011  46002024      -  11.50   \n",
       "4438    CREB1        CREB1         chr21  45874430  45874443      +  10.33   \n",
       "4439    CREB1        CREB1         chr21  47741634  47741647      -  10.19   \n",
       "\n",
       "       p-value  q-value matched_sequence  \n",
       "0     0.000061      NaN   ATATGACGTACTGG  \n",
       "1     0.000026      NaN   ATGTGACGTCAGTC  \n",
       "2     0.000061      NaN   ATATGACGTACTGG  \n",
       "3     0.000061      NaN   ATATGACGTACTGG  \n",
       "4     0.000026      NaN   ATGTGACGTCAGTC  \n",
       "...        ...      ...              ...  \n",
       "4435  0.000063      NaN   CAGTGACGCAGTCT  \n",
       "4436  0.000041      NaN   GGATGACGTCATAT  \n",
       "4437  0.000043      NaN   ATATGACGTCATCC  \n",
       "4438  0.000080      NaN   AAATGACGCAGGCA  \n",
       "4439  0.000085      NaN   AAGTGACGCAGGTC  \n",
       "\n",
       "[4440 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(f\"{OUTPUT_SX}/CREB1_PWM_promoter_calls.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fccc771",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
