{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48adf50a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#uvddb_r1 = uac.process_alldata_file(UVDDB_FILES[0])\n",
    "#uvddb_r2 = uac.process_alldata_file(UVDDB_FILES[1])\n",
    "#uvddb = pd.merge(uvddb_r1, uvddb_r2, on=[\"Name\", \"Sequence\", \"Has_PyDi\"], suffixes=(\"_r1\", \"_r2\"))\n",
    "\n",
    "# Calculate R2 \n",
    "regression = stats.linregress(uvddb[\"Signal_r1\"], uvddb[\"Signal_r2\"])\n",
    "rsquared = regression.rvalue ** 2\n",
    "\n",
    "# Plot scatterplot\n",
    "source = ColumnDataSource(uvddb)\n",
    "p = figure(plot_width=800, plot_height=800,\n",
    "           x_range=F5B_RANGE, y_range=F5B_RANGE)\n",
    "p.circle(\"Signal_r1\",\n",
    "         \"Signal_r2\",\n",
    "         source=source,\n",
    "         color = \"black\",\n",
    "         size=F5B_CIRCLE_SIZE)\n",
    "p.xaxis.ticker = F5B_TICKS\n",
    "p.yaxis.ticker = F5B_TICKS\n",
    "p.xgrid.grid_line_color = None\n",
    "p.ygrid.grid_line_color = None\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "p.toolbar_location = None\n",
    "#p.output_backend = \"svg\"\n",
    "show(p)\n",
    "#export_png(p, filename=\"../Scratch/UVDDB_PLOTS/UVDDB_scatter.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e37dbcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "uvddb_r1 = uac.process_alldata_file(UVDDB_FILES[0])\n",
    "uvddb_r2 = uac.process_alldata_file(UVDDB_FILES[1])\n",
    "uvddb = pd.merge(uvddb_r1, uvddb_r2, on=[\"Name\", \"Sequence\", \"Has_PyDi\"], suffixes=(\"_r1\", \"_r2\"))\n",
    "\n",
    "# Calculate R2 \n",
    "regression = stats.linregress(uvddb[\"Signal_r1\"], uvddb[\"Signal_r2\"])\n",
    "rsquared = regression.rvalue ** 2\n",
    "\n",
    "# Plot scatterplot\n",
    "source = ColumnDataSource(uvddb)\n",
    "p = figure(plot_width=800, plot_height=800,\n",
    "           x_range=F5B_RANGE, y_range=F5B_RANGE)\n",
    "p.circle(\"Signal_r1\",\n",
    "         \"Signal_r2\",\n",
    "         source=source,\n",
    "         color = \"black\",\n",
    "         size=F5B_CIRCLE_SIZE)\n",
    "p.xaxis.ticker = F5B_TICKS\n",
    "p.yaxis.ticker = F5B_TICKS\n",
    "p.xgrid.grid_line_color = None\n",
    "p.ygrid.grid_line_color = None\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "p.toolbar_location = None\n",
    "#p.output_backend = \"svg\"\n",
    "show(p)\n",
    "#export_png(p, filename=\"../Scratch/UVDDB_PLOTS/UVDDB_scatter.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b93db04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot data\n",
    "uvddb_results = result_medians[result_medians[\"Experiment\"] == \"UVDDB_r1\"].reset_index(drop=True)\n",
    "source = ColumnDataSource(uvddb_results)\n",
    "figures = []\n",
    "colors = [\"#1b9e77\",\n",
    "          '#d95f02',\n",
    "          '#7570b3',\n",
    "          '#e7298a']\n",
    "pydi = (\"TT\", \"TC\", \"CT\", \"CC\")\n",
    "p = figure(plot_width=600, plot_height=800)\n",
    "for color, py in zip(colors, pydi) :\n",
    "    plot_df = uvddb_results[uvddb_results[\"Group\"] == pydi]\n",
    "    p.circle(plot_df[py],\n",
    "             plot_df[\"Median_Signal_ln\"],\n",
    "             color=color,\n",
    "             size=CIRCLE_SIZE)\n",
    "    p.line(plot_df[py],\n",
    "           plot_df[\"Median_Signal_ln\"],\n",
    "           color=color,\n",
    "           line_width=5)\n",
    "p.xgrid.visible = False\n",
    "#p.ygrid.visible = False\n",
    "p.yaxis.ticker = \n",
    "p.xaxis.major_tick_line_color = None  # turn off x-axis major ticks\n",
    "p.xaxis.minor_tick_line_color = None  # turn off x-axis minor ticks\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "show(p)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0619169",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot data\n",
    "CIRCLE_SIZE=25\n",
    "df = plot_df\n",
    "source = ColumnDataSource(df)\n",
    "figures = []\n",
    "colors = ['black', 'cyan', 'gold']\n",
    "pydi = (\"TT\", \"TC\", \"CT\", \"CC\")\n",
    "seqs = (\"MITF\", \"TBP\", \"p53\")\n",
    "for dinuc in pydi:\n",
    "    p = figure(plot_width=600, plot_height=800, y_range=CPD_RANGE)\n",
    "    for seq, color in zip(seqs, colors):\n",
    "        pdf = df[(df[\"Sequence\"] == seq) & (df[\"PyDi\"] == dinuc)].reset_index(drop=True)\n",
    "        p.circle(pdf[\"Count\"],\n",
    "                 pdf[\"Signal\"],\n",
    "                 color=color,\n",
    "                 size=CIRCLE_SIZE)\n",
    "        p.line(pdf[\"Count\"],\n",
    "                 pdf[\"Signal\"],\n",
    "                 color=color,\n",
    "                 line_width=5)\n",
    "    p.xaxis.ticker = []\n",
    "    p.yaxis.ticker = [np.log(2000), np.log(10000), np.log(20000)]\n",
    "    p.xgrid.grid_line_color = None\n",
    "    p.ygrid.grid_line_color = None\n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.toolbar_location = None\n",
    "    show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bbb2b5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "uvddb_r1 = uac.process_alldata_file(UVDDB_FILES[0])\n",
    "uvddb_r2 = uac.process_alldata_file(UVDDB_FILES[1])\n",
    "uvddb = pd.merge(uvddb_r1, uvddb_r2, on=[\"Name\", \"Sequence\", \"Has_PyDi\"], suffixes=(\"_r1\", \"_r2\"))\n",
    "\n",
    "# Calculate R2 \n",
    "regression = stats.linregress(uvddb[\"Signal_r1\"], uvddb[\"Signal_r2\"])\n",
    "rsquared = regression.rvalue ** 2\n",
    "\n",
    "# Plot scatterplot\n",
    "source = ColumnDataSource(uvddb)\n",
    "p = figure(plot_width=800, plot_height=800,\n",
    "           x_range=F5B_RANGE, y_range=F5B_RANGE)\n",
    "p.circle(\"Signal_r1\",\n",
    "         \"Signal_r2\",\n",
    "         source=source,\n",
    "         color = \"black\",\n",
    "         size=F5B_CIRCLE_SIZE)\n",
    "p.xaxis.ticker = F5B_TICKS\n",
    "p.yaxis.ticker = F5B_TICKS\n",
    "p.xgrid.grid_line_color = None\n",
    "p.ygrid.grid_line_color = None\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "p.toolbar_location = None\n",
    "#p.output_backend = \"svg\"\n",
    "show(p)\n",
    "#export_png(p, filename=\"../Scratch/UVDDB_PLOTS/UVDDB_scatter.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ff3c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "TFS = (\"MITF\", \"TBP\", \"p53\")\n",
    "MITF_TICKS = [np.log(20000), np.log(25000), np.log(35000)]\n",
    "TBP_TICKS = [np.log(25000), np.log(30000), np.log(40000)]\n",
    "P53_TICKS = [np.log(20000), np.log(30000), np.log(50000)]\n",
    "TICK_TUP = (MITF_TICKS, TBP_TICKS, P53_TICKS)\n",
    "\n",
    "for TF, TICK in zip(TFS, TICK_TUP):\n",
    "    df = process_file(FILE, False, False)\n",
    "    mitf_probes = df[(df[\"Name\"].str.contains('P6')) &\n",
    "                   (df[\"Name\"].str.startswith(TF))]\n",
    "    mitf_probes = mitf_probes.reset_index(drop=True)\n",
    "    features = (\"TT\", \"TC\", \"CT\", \"CC\")\n",
    "    for i in features:\n",
    "        result = []\n",
    "        for seq in mitf_probes[\"Sequence\"]:\n",
    "            result.append(dimer_count(seq, i))\n",
    "        mitf_probes[i] = result\n",
    "    mitf_probes\n",
    "    only_one = []\n",
    "    for row in mitf_probes.itertuples():\n",
    "        if [row.TT, row.TC, row.CT, row.CC].count(0) >= 3:\n",
    "            only_one.append(1)\n",
    "        else:\n",
    "            only_one.append(0)\n",
    "    mitf_probes[\"Select\"] = only_one\n",
    "    mitf_probes = mitf_probes[mitf_probes[\"Select\"] == 1]\n",
    "    mitf_probes = mitf_probes.reset_index(drop=True)\n",
    "    mitf_probes = mitf_probes[[\"TT\", \"TC\", \"CT\", \"CC\", \"Alexa488\"]]\n",
    "    mitf_probes = mitf_probes.groupby(by=[\"TT\", \"TC\", \"CT\", \"CC\"]).aggregate(np.median)\n",
    "    mitf_probes = mitf_probes.reset_index()\n",
    "    mitf_probes[\"Alexa488\"] = mitf_probes[\"Alexa488\"].apply(lambda x: np.log(x))\n",
    "    mitf_probes\n",
    "    # Plot data\n",
    "    CIRCLE_SIZE=25\n",
    "    dataframe = mitf_probes\n",
    "    source = ColumnDataSource(dataframe)\n",
    "    figures = []\n",
    "    colors = [\"#1b9e77\",\n",
    "        '#d95f02',\n",
    "        '#7570b3',\n",
    "        '#e7298a']\n",
    "    pydi = (\"TT\", \"TC\", \"CT\", \"CC\")\n",
    "    p = figure(plot_width=600, plot_height=800)\n",
    "    for color, py in zip(colors, pydi) :\n",
    "        df = query_group(dataframe, py)\n",
    "        p.circle(df[py],\n",
    "                  df[\"Alexa488\"],\n",
    "                  color = color,\n",
    "                  size=CIRCLE_SIZE)\n",
    "        p.line(df[py],\n",
    "                  df[\"Alexa488\"],\n",
    "                  color = color,\n",
    "              line_width=5)\n",
    "    p.xgrid.visible = False\n",
    "    #p.ygrid.visible = False\n",
    "    p.yaxis.ticker = TICK\n",
    "    p.xaxis.major_tick_line_color = None  # turn off x-axis major ticks\n",
    "    p.xaxis.minor_tick_line_color = None  # turn off x-axis minor ticks\n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.major_label_text_font_size = '0pt'\n",
    "    show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2fc4a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "regress_df = pd.DataFrame(regress_results)\n",
    "regress_df = regress_df.rename(columns={0:\"Experiment\", 1:\"Sequence_Set\", 2:\"Group\", 3:\"P_Value\"})\n",
    "regress_df[\"P_Value\"] = regress_df[\"P_Value\"].apply(lambda x: '{:.2E}'.format(x))\n",
    "regress_df = regress_df.sort_values(by=[\"Experiment\", \"Sequence_Set\", \"Group\"]).reset_index(drop=True)\n",
    "regress_df.to_csv(f\"{OUTPUT_FOLDER_T2}/Linear_Regression.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50bfe790",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_results = []\n",
    "for experiment in list(set(result[\"Experiment\"])):\n",
    "    for sequence_set in list(set(result[\"Sequence_Set\"])):\n",
    "        for group in list(set(result[\"Group\"])):\n",
    "            # Get data subset\n",
    "            comparison_df = result[(result[\"Experiment\"] == experiment) &\n",
    "                                   (result[\"Sequence_Set\"] == sequence_set) &\n",
    "                                   (result[\"Group\"] == group)].reset_index(drop=True)\n",
    "            count_a_values = comparison_df[comparison_df[\"Has_PyDi\"] == \"No\"][\"Signal\"]\n",
    "            count_b_values = comparison_df[comparison_df[\"Has_PyDi\"] == \"Yes\"][\"Signal\"]\n",
    "            p_value = stats.ranksums(count_a_values, count_b_values).pvalue\n",
    "            group_results.append((experiment, sequence_set, group, p_value))\n",
    "group_df = pd.DataFrame(group_results)\n",
    "group_df = group_df.rename(columns={0:\"Experiment\", 1:\"Sequence_Set\", 2:\"Group\", 3:\"P_Value\"})\n",
    "#group_df[\"FDR\"] = multipletests(group_df[\"P_Value\"], method=\"fdr_bh\")[1]\n",
    "group_df.to_csv(f\"{OUTPUT_FOLDER_T2}/YesNo.csv\", index=None)\n",
    "statistics_results = []\n",
    "for experiment in list(set(result[\"Experiment\"])):\n",
    "    for sequence_set in list(set(result[\"Sequence_Set\"])):\n",
    "        for group in list(set(result[\"Group\"])):\n",
    "            # Get data subset\n",
    "            comparison_df = result[(result[\"Experiment\"] == experiment) &\n",
    "                                   (result[\"Sequence_Set\"] == sequence_set) &\n",
    "                                   (result[\"Group\"] == group)].reset_index(drop=True)\n",
    "            group_results = []\n",
    "            for count_a, count_b in itertools.combinations((0, 1, 2, 3), 2):\n",
    "                count_a_values = comparison_df[comparison_df[\"Count\"] == count_a][\"Signal\"]\n",
    "                count_b_values = comparison_df[comparison_df[\"Count\"] == count_b][\"Signal\"]\n",
    "                p_value = stats.ranksums(count_a_values, count_b_values).pvalue\n",
    "                group_results.append((experiment, sequence_set, group, count_a, count_b, p_value))\n",
    "            group_df = pd.DataFrame(group_results)\n",
    "            group_df = group_df.rename(columns={0:\"Experiment\", 1:\"Sequence_Set\", 2:\"Group\", 3:\"Count_A\", 4:\"Count_B\", 5:\"P_Value\"})\n",
    "            group_df[\"FDR\"] = multipletests(group_df[\"P_Value\"], method=\"fdr_bh\")[1]\n",
    "            statistics_results.append(group_df)\n",
    "statistics_df = pd.concat(statistics_results).reset_index(drop=True)\n",
    "statistics_df = statistics_df.sort_values(by=[\"Experiment\", \"Sequence_Set\", \"Group\", \"Count_A\", \"Count_B\"]).reset_index(drop=True)\n",
    "statistics_df[\"P_Value\"] = statistics_df[\"P_Value\"].apply(lambda x: \"{:.2e}\".format(x))\n",
    "statistics_df[\"FDR\"] = statistics_df[\"FDR\"].apply(lambda x: \"{:.2e}\".format(x))\n",
    "statistics_df.to_csv(f\"{OUTPUT_FOLDER_T2}/Table_S2_Pairwise.csv\")\n",
    "statistics_df.query(\"Count_A == 0\").to_csv(f\"{OUTPUT_FOLDER_T2}/Table_S2_Pairwise_nonevsany.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "564d4c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "result[\"Has_PyDi\"] = result[\"Count\"].apply(lambda x: \"Yes\" if x > 0 else \"No\")\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c74b13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redo to accompdate for result_medians ####\n",
    "def plot_by_dinucleotide(df: pd.DataFrame,\n",
    "                         output: str,\n",
    "                         y_range: tuple,\n",
    "                         circle_size: int,\n",
    "                         ticker: list,\n",
    "                         palette: tuple):\n",
    "    \"\"\"Generates a set of line plots \"\"\"\n",
    "    figures = []\n",
    "    colors = palette\n",
    "    pydi = (\"TT\", \"TC\", \"CT\", \"CC\")\n",
    "    seqs = (\"MITF\", \"p53\", \"TBP\")\n",
    "    tickers = [ticker, [], [], []]\n",
    "    plots = []\n",
    "    # For each set of dinucleotide and y ticker to plot\n",
    "    for dinuc, y_ticks in zip(pydi, tickers):\n",
    "        # Create figure object\n",
    "        p = figure(plot_width=150, plot_height=200, y_range=y_range)\n",
    "        # For each sequence and color set\n",
    "        for seq, color in zip(seqs, colors):\n",
    "            # Filter the data for the pyrimidine dinucleotide and sequence\n",
    "            pdf = df[(df[\"Sequence\"] == seq) & (df[\"PyDi\"] == dinuc)].reset_index(drop=True)\n",
    "            # Draw the circles and lines for that dataframe\n",
    "            p.circle(pdf[\"Count\"],\n",
    "                     pdf[\"Signal\"],\n",
    "                     color=color,\n",
    "                     size=circle_size)\n",
    "            p.line(pdf[\"Count\"],\n",
    "                   pdf[\"Signal\"],\n",
    "                   color=color,\n",
    "                   line_width=2)\n",
    "        # Set an empty x axsis ticker\n",
    "        p.xaxis.ticker = []\n",
    "        # Set y axis ticker\n",
    "        p.yaxis.ticker = y_ticks\n",
    "        # Remove grid lines, label text, and toolbar\n",
    "        p.xgrid.grid_line_color = None\n",
    "        p.ygrid.grid_line_color = None\n",
    "        p.xaxis.major_label_text_font_size = '0pt'\n",
    "        p.yaxis.major_label_text_font_size = '0pt'\n",
    "        p.toolbar_location = None\n",
    "        # Settings for the border\n",
    "        p.outline_line_width = 1\n",
    "        p.outline_line_color = 'black'\n",
    "        # Set backend to svg\n",
    "        p.output_backend = 'svg'\n",
    "        # Add to list of plots\n",
    "        plots.append(p)\n",
    "    # Create a grid of plots from the plots list with 1 row\n",
    "    grid = gridplot([plots])\n",
    "    # Export the grid of plots\n",
    "    export_svg(grid, filename=output)\n",
    "\n",
    "\n",
    "# Plot CPD and 6-4PP data\n",
    "plot_by_dinucleotide(cpd_results,\n",
    "                     f\"{OUTPUT_FOLDER_F1}/CPD.svg\",\n",
    "                     CPD_RANGE,\n",
    "                     F1E_CIRCLE_SIZE,\n",
    "                     CPD_Y_TICKER,\n",
    "                     F1E_COLOR_PALETTE)\n",
    "plot_by_dinucleotide(sixfour_results,\n",
    "                     f\"{OUTPUT_FOLDER_F1}/64PP.svg\",\n",
    "                     SIXFOUR_RANGE,\n",
    "                     F1E_CIRCLE_SIZE,\n",
    "                     SIXFOUR_Y_TICKER,\n",
    "                     F1E_COLOR_PALETTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29395beb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kl_divergence_list(pwm_a, pwm_b) -> list[float]:\n",
    "    \"\"\"Returns the sum of divergences\"\"\"\n",
    "    kl_list = []\n",
    "    for i in range(len(pwm_a)):\n",
    "        np_matrix_a = np.array(list(pwm_a.T[i]))\n",
    "        np_matrix_b = np.array(list(pwm_b.T[i]))\n",
    "        kl_list.append(kl_divergence(np_matrix_a, np_matrix_b))\n",
    "    return kl_list\n",
    "\n",
    "def matrix_dictionary_from_meta_data(meta_dataframe: pd.DataFrame,\n",
    "                                     matrix_name: str,\n",
    "                                     trim_matrix=True) -> dict:\n",
    "    \"\"\"Given a meta data file, reads all matricies from a given name\n",
    "\n",
    "    Given the name of a type of matrix (Enrichment, Probability, etc),\n",
    "    reads all matricies as defined in the meta data file and returns a \n",
    "    dictionary where the keys are the experiment ID and values are the\n",
    "    matricies in long form.\n",
    "    \"\"\"\n",
    "    matrix_dict = {}\n",
    "    for row in meta_dataframe.itertuples():\n",
    "        prefix = f\"{DATA_FOLDER}/{row.Protein}_{row.Condition}_ID{row.ID}\"\n",
    "        if row.File == \"Default\":\n",
    "            file_path = f\"{prefix}_7mers_pwm.txt\"\n",
    "        else:\n",
    "            file_path = f\"{prefix}_pwm_from_top_canonical.txt\"\n",
    "        pwm = read_pwm_matrix(file_path, matrix_name)\n",
    "        if row.Reverse_Complement == 1:\n",
    "            pwm = reverse_complement_matrix(pwm)\n",
    "        if trim_matrix:\n",
    "            # Trim PWM\n",
    "            pwm = pwm.iloc[row.Start_Inclusive:row.End_Exclusive, :]\n",
    "        pwm = pwm.reset_index(drop=True)\n",
    "        matrix_dict[row.ID] = pwm\n",
    "    return matrix_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4932c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "ic_plotted_data[\"Experiment\"] = ic_plotted_data[\"Experiment\"].apply(lambda x: \"_\".join(x.split('_')[:2]))\n",
    "ic_plotted_data[\"Experiment\"] = ic_plotted_data[\"Experiment\"].apply(lambda x: x.replace(\"_WC\", \"_NonUV\"))\n",
    "ic_plotted_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c846dba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.figure(figsize=(10,10))\n",
    "plt.scatter(plot_df[\"Escore_WC8\"],\n",
    "            plot_df[\"Escore_UV9\"],\n",
    "            c=plot_df[\"Preference_Score\"],\n",
    "            cmap=\"bwr\")\n",
    "plt.plot(plot_df[\"Escore_WC8\"],\n",
    "         plot_df[\"Confidence_Interval_Upper\"],\n",
    "         color=\"black\")\n",
    "plt.plot(plot_df[\"Escore_WC8\"],\n",
    "         plot_df[\"Confidence_Interval_Lower\"],\n",
    "         color=\"black\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb05151",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Plot - WT Comparison\n",
    "plot_df = df.sort_values(by=\"Escore_WC8\").reset_index()\n",
    "plot_range = (-0.52, 0.52)\n",
    "p = figure(x_range=plot_range, y_range=plot_range)\n",
    "p.circle(x=plot_df[\"Escore_WC8\"],\n",
    "         y=plot_df[\"Escore_WC9\"],\n",
    "         color='black',\n",
    "        size=6)\n",
    "p.line(x=plot_df[\"Escore_WC8\"],\n",
    "       y=plot_df[\"Confidence_Interval_Upper\"],\n",
    "       color='black',\n",
    "       line_dash='dashed',\n",
    "       line_width=2)\n",
    "p.line(x=plot_df[\"Escore_WC8\"],\n",
    "       y=plot_df[\"Confidence_Interval_Lower\"],\n",
    "       color='black',\n",
    "       line_dash='dashed',\n",
    "       line_width=2)\n",
    "p.toolbar_location = None\n",
    "p.xgrid.visible = False\n",
    "p.ygrid.visible = False\n",
    "p.xaxis.minor_tick_line_color = None\n",
    "p.yaxis.minor_tick_line_color = None\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "show(p)\n",
    "p.output_backend='svg'\n",
    "export_svg(p, filename=f\"{OUTPUT}/Replicate_6mer_OLS.svg\")\n",
    "\n",
    "\n",
    "# Plot - WT Comparison\n",
    "plot_df = df.sort_values(by=\"Escore_WC8\").reset_index()\n",
    "plot_range = (-0.52, 0.52)\n",
    "p = figure(x_range=plot_range, y_range=plot_range)\n",
    "p.circle(x=plot_df[\"Escore_WC8\"],\n",
    "         y=plot_df[\"Escore_UV9\"],\n",
    "         color=plot_df[\"Color\"],\n",
    "         size=6)\n",
    "p.line(x=plot_df[\"Escore_WC8\"],\n",
    "       y=plot_df[\"Confidence_Interval_Upper\"],\n",
    "       color='black',\n",
    "       line_dash='dashed',\n",
    "       line_width=2)\n",
    "p.line(x=plot_df[\"Escore_WC8\"],\n",
    "       y=plot_df[\"Confidence_Interval_Lower\"],\n",
    "       color='black',\n",
    "       line_dash='dashed',\n",
    "       line_width=2)\n",
    "p.toolbar_location = None\n",
    "p.xgrid.visible = False\n",
    "p.ygrid.visible = False\n",
    "p.xaxis.minor_tick_line_color = None\n",
    "p.yaxis.minor_tick_line_color = None\n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.yaxis.major_label_text_font_size = '0pt'\n",
    "show(p)\n",
    "p.output_backend='svg'\n",
    "export_svg(p, filename=f\"{OUTPUT}/Comparison_6mer_OLS.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ef84ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the meta data and make output directories if not present already\n",
    "\n",
    "plot_dataframes = []\n",
    "# For each row in the meta data table\n",
    "for row in kmer_meta.itertuples():\n",
    "    # Read dataframes, merge, assign colors\n",
    "    non_uv_file = f\"{DATA_FOLDER_KMERS}/{row.Non_UV_File}\"\n",
    "    uv_file = f\"{DATA_FOLDER_KMERS}/{row.UV_File}\"\n",
    "    plot_dataframe = uac.read_kmer_file_pair(non_uv_file,\n",
    "                                             uv_file,\n",
    "                                             (\"_NUV\", \"_UV\"))\n",
    "    plot_dataframe = color_on_threshold(plot_dataframe)\n",
    "    # Plot E-Scores\n",
    "    p = kmer_scatterplot(plot_dataframe,\n",
    "                         \"E-score_NUV\",\n",
    "                         \"E-score_UV\",\n",
    "                         ESCORE_PLOT_RANGE,\n",
    "                         ticker_list=ESCORE_TICKER)\n",
    "    p.output_backend = 'svg'\n",
    "    export_svg(p, filename=f\"{FIGURE_S3_OUTPUT}/{row.Protein}_Escore.svg\")\n",
    "    # Plot Z-scores\n",
    "    plot_range = uac.plot_range_from_x_y(plot_dataframe[\"Z-score_NUV\"],\n",
    "                                        plot_dataframe[\"Z-score_UV\"])\n",
    "    zscore_ticker = []\n",
    "    p = kmer_scatterplot(plot_dataframe,\n",
    "                         \"Z-score_NUV\",\n",
    "                         \"Z-score_UV\",\n",
    "                         plot_range,\n",
    "                         ticker_list=zscore_ticker)\n",
    "    p.output_backend = 'svg'\n",
    "    export_svg(p, filename=f\"{FIGURE_2_OUTPUT}/{row.Protein}_Zscore.svg\")\n",
    "    plot_dataframe[\"Protein\"] = row.Protein\n",
    "    plot_dataframes.append(plot_dataframe)\n",
    "plotted_data = pd.concat(plot_dataframes)\n",
    "#plotted_data.to_csv(f\"{FIGURE_2_OUTPUT}/{row.Protein}_plotted_kmer_data.csv\")\n",
    "#plotted_data.to_csv(f\"{FIGURE_S3_OUTPUT}/{row.Protein}_plotted_kmer_data.csv\")\n",
    "plotted_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac06f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_meta = pd.read_csv(META_DATA_KMERS)\n",
    "plot_dfs= []\n",
    "# For each row in the meta data table\n",
    "for row in kmer_meta.itertuples():\n",
    "    # Read dataframes, merge, assign colors\n",
    "    non_uv_file = f\"{DATA_FOLDER_KMERS}/{row.Non_UV_File}\"\n",
    "    uv_file = f\"{DATA_FOLDER_KMERS}/{row.UV_File}\"\n",
    "    plot_dataframe = uac.read_kmer_file_pair(non_uv_file,\n",
    "                                             uv_file,\n",
    "                                             (\"_NUV\", \"_UV\"))\n",
    "    plot_dataframe[\"Experiment\"] = row.Protein\n",
    "    plot_dataframe = group_on_threshold(plot_dataframe)\n",
    "    plot_dfs.append(plot_dataframe)\n",
    "result = pd.concat(plot_dfs).reset_index(drop=True)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe4c89a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, axes = plt.subplots(ncols=5, sharex=True, sharey=True)\n",
    "for ax in axes:\n",
    "    ax.plot([1, 2, 3], [1, 2, 3])\n",
    "    ax.set(adjustable='box-forced', aspect='equal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1358cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "sns.set_theme(style=\"white\")\n",
    "\n",
    "FIRST_ROW = (\"MYC\", \"CREB1\", \"EGR1\", \"ELK1\", \"ETS1\")\n",
    "\n",
    "\n",
    "plt.show()\n",
    "fig, ax_array = plt.subplots(1, 5)\n",
    "fig.set_size_inches(25, 5)\n",
    "for idx, protein in enumerate(FIRST_ROW):\n",
    "    df = result[result[\"Experiment\"] == protein]\n",
    "    sns.scatterplot(x=\"E-score_NUV\",\n",
    "                    y=\"E-score_UV\",\n",
    "                    data=df,\n",
    "                    ax=ax_array[idx],\n",
    "                   hue=\"Color\",\n",
    "                    hue_order=GROUPS,\n",
    "                   palette=COLOR_PALETTE,\n",
    "                   linewidth=0,\n",
    "                   legend=False)\n",
    "    ax_array[idx].set(adjustable='box', aspect='equal')\n",
    "    #plt.axline((0, 0), slope=1, color=\"black\", ax=ax_array[idx])\n",
    "plt.subplots_adjust(wspace=0.0001)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e3e6ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "Draw scatterplots for both E-score and Z-score metrics. \n",
    "\n",
    "1. Read in meta data and ensure output location exists.\n",
    "2. Use that data to read pairs of k-mer files and combine as one dataframe.\n",
    "3. Determine the color for each datapoint and add to the dataframe.\n",
    "4. Save the plots for both E-score and Z-score metrics.\n",
    "5. Save the dataframe used to plot the data as a table for both metrics.\n",
    "\n",
    "**Input**: META_DATA_KMERS file and the 7of9mer files it references.\n",
    "\n",
    "**Output**: Scatterplots comparing non-UV and UV conditions across 10 transcription factors for both E-scores and Z-scores saved in svg format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d861fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def kmer_scatterplot(dataframe: pd.DataFrame,\n",
    "                     column_a: str,\n",
    "                     column_b: str,\n",
    "                     plot_range: tuple,\n",
    "                     ticker_list: list,\n",
    "                     font: str = 'ariel',\n",
    "                     font_size: str = '0pt',\n",
    "                     circle_size: int = 10,\n",
    "                     line_width: int = 2,\n",
    "                     outline_width: int = 2,\n",
    "                     outline_color: str = \"black\"):\n",
    "    \"\"\"Return a bokeh figure object of a scatterplot.\n",
    "\n",
    "    Given a dataframe, two columns of data to plot as column a\n",
    "    and b, a tuple of two values for the plot min and max values,\n",
    "    and settings for font, font_size, circle_size, and color, \n",
    "    returns a bokeh figure object of a scatterplot. \n",
    "\n",
    "    The plot contains a black border with the diagonal drawn. \n",
    "    \"\"\"\n",
    "    source = ColumnDataSource(dataframe)\n",
    "    p = figure(x_range=plot_range, y_range=plot_range)\n",
    "    # Draw the data points\n",
    "    p.circle(column_a,\n",
    "             column_b,\n",
    "             color=\"Color\",\n",
    "             size=circle_size,\n",
    "             source=source)\n",
    "    # Draw a diagonal line\n",
    "    p.line(plot_range, plot_range, color=\"black\", width=line_width)\n",
    "    # Settings for label text\n",
    "    p.xaxis.major_label_text_font_size = font_size\n",
    "    p.yaxis.major_label_text_font_size = font_size\n",
    "    p.xaxis.major_label_text_font = font\n",
    "    p.yaxis.major_label_text_font = font\n",
    "    # Settings for the ticker and tick marks\n",
    "    p.xaxis.ticker = ticker_list\n",
    "    p.yaxis.ticker = ticker_list\n",
    "    # Settings for the border\n",
    "    p.outline_line_width = outline_width\n",
    "    p.outline_line_color = outline_color\n",
    "    # Remove toolbar and grids\n",
    "    p.toolbar_location = None\n",
    "    p.xgrid.visible = False\n",
    "    p.ygrid.visible = False\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd28e12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_df_from_var_df2(tf, var_id, yticker, output):\n",
    "    t = variation_dict[tf].copy()\n",
    "    t = t[t[\"Variant_ID\"] == var_id]\n",
    "    # Remove WT \n",
    "    wt = t[t[\"Position\"] < 0]\n",
    "    wt_len = len(wt)\n",
    "    if wt_len != 1:\n",
    "        raise ValueError(\"WT Search found {wt_len} entries.\")\n",
    "    wt_nuv = float(wt.Signal_Non_UV)\n",
    "    wt_uv = float(wt.Signal_UV)\n",
    "    var = t[t[\"Position\"] > 0].reset_index(drop=True)\n",
    "    # Calculate relative plot positions \n",
    "    min_pos = min(var[\"Position\"])\n",
    "    nuv_plot_positions = []\n",
    "    for row in var.itertuples():\n",
    "        # If single mutant\n",
    "        if len(row.Variant) == 1:\n",
    "            pos = row.Position - min_pos + 1\n",
    "        # If double mutant\n",
    "        elif len(row.Variant) == 2:\n",
    "            pos = row.Position - min_pos + 1.5\n",
    "        else:\n",
    "            raise ValueError(\"Variant length not 1 or 2.\")\n",
    "        nuv_plot_positions.append(pos)\n",
    "    var[\"Plot_Positions_NUV\"] = nuv_plot_positions\n",
    "    max_pos = max(nuv_plot_positions)\n",
    "    offset = max_pos + 2\n",
    "    var[\"Plot_Positions_UV\"] = var[\"Plot_Positions_NUV\"] + offset - 1\n",
    "    #########\n",
    "    colors = []\n",
    "    for row in var.itertuples():\n",
    "        if row.Signal_UV > row.Prediction_Upper:\n",
    "            colors.append('#8db0fe')\n",
    "        elif row.Signal_UV < row.Prediction_Lower:\n",
    "            colors.append('#f4987a')\n",
    "        else:\n",
    "            colors.append(\"grey\")\n",
    "    var[\"Color\"] = colors\n",
    "    # Convert to a long format df\n",
    "    nuv_data = var[[\"Signal_Non_UV\", \"Is_Significant\", \"Variant\", \"Plot_Positions_NUV\", \"Color\"]]\n",
    "    nuv_data = nuv_data.rename(columns={\"Signal_Non_UV\":\"Signal\",\n",
    "                                        \"Plot_Positions_NUV\":\"Plot_Position\"})\n",
    "    nuv_data[\"Group\"] = \"NUV\"\n",
    "    uv_data = var[[\"Signal_UV\", \"Is_Significant\", \"Variant\", \"Plot_Positions_UV\", \"Color\"]]\n",
    "    uv_data = uv_data.rename(columns={\"Signal_UV\":\"Signal\", \"Plot_Positions_UV\":\"Plot_Position\"})\n",
    "    uv_data[\"Group\"] = \"UV\"\n",
    "    plot_df = pd.concat([nuv_data, uv_data]).reset_index(drop=True)\n",
    "    max_plot_pos = max(plot_df[\"Plot_Position\"])\n",
    "    circle_size=5\n",
    "    plot_y_offset = (max(plot_df[\"Signal\"]) - min(plot_df[\"Signal\"])) * 0.1\n",
    "    y_range = (min(plot_df[\"Signal\"]) - plot_y_offset, max(plot_df[\"Signal\"]) + plot_y_offset)\n",
    "    x_range = (0, max_plot_pos + 1)\n",
    "    \n",
    "    return plot_df\n",
    "\n",
    "\n",
    "plot_df_from_var_df2(\"EGR1\", 6, (np.log(2500), np.log(10000), np.log(40000)), f\"{OUTPUT_MAIN}/Figure_3e.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15b4dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in fig_s5_meta.itertuples():\n",
    "    # Read and convert tick marks to ln space\n",
    "    tick_marks = [row.Tick_Mark_A, row.Tick_Mark_B, row.Tick_Mark_C]\n",
    "    tick_marks_ln = list(map(np.log, tick_marks))\n",
    "    # Generate positional plots\n",
    "    plot_df_from_var_df(row.Protein,\n",
    "                        row.Variant_ID,\n",
    "                        tick_marks_ln,\n",
    "                        output = f\"{OUTPUT_SUPP}/{row.Protein}_{row.Variant_ID}.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6014af63",
   "metadata": {},
   "outputs": [],
   "source": [
    "variation_dict[\"MITF\"].query(\"Variant_ID == 10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8112d0cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_plot(df: pd.DataFrame,\n",
    "                   column_x: str,\n",
    "                   column_y: str,\n",
    "                   hue: str,\n",
    "                   palette: tuple[str, str],\n",
    "                   order: tuple[str, str],\n",
    "                   x_ticks: tuple[int, int],\n",
    "                   output: str,\n",
    "                   ax=ax):\n",
    "    #rcParams['figure.figsize'] = 10, 2.5\n",
    "    sns.set_style('ticks')\n",
    "    #L=5.8\n",
    "    #U=13\n",
    "    p = sns.stripplot(data=df,\n",
    "                      x=column_x,\n",
    "                      y=column_y,\n",
    "                      hue=hue,\n",
    "                      order=order,\n",
    "                      orient='h',\n",
    "                      jitter=False,\n",
    "                      s=15,\n",
    "                      palette=['red', 'blue'],\n",
    "                      alpha=0.5,\n",
    "                      linewidth=0,\n",
    "                      ax=ax)\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label    \n",
    "    #p.set(xticks=[np.log(1000), np.log(10000), np.log(100000)])\n",
    "    #p.set(xticklabels=['1000', '10000', '100000'])  # remove the tick labels\n",
    "    #p.set(yticklabels=[])  # remove the tick labels\n",
    "    #plt.xlim(L, U)\n",
    "    #plt.legend('',frameon=False)\n",
    "    #plt.savefig(output, format=\"svg\")\n",
    "    #plt.close()\n",
    "\n",
    "\n",
    "for protein, highlights in ((\"CREB1\", CREB1_HIGHLIGHT),\n",
    "                            (\"EGR1\", EGR1_HIGHLIGHT)):\n",
    "    plot_df = relabeled_dfs[protein].copy()\n",
    "    plot_df = plot_df[plot_df[\"Name\"].isin(set(highlights))]\n",
    "    plot_df = pd.melt(plot_df,\n",
    "                      id_vars=[\"Name\"],\n",
    "                      value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    highlight_plot(plot_df,\n",
    "                   \"value\",\n",
    "                   \"Name\",\n",
    "                   \"variable\",\n",
    "                   ('blue', 'red'),\n",
    "                   highlights,\n",
    "                   [np.log(1000), np.log(10000), np.log(100000)],\n",
    "                  f\"{OUTPUT_FOLDER_MAIN}/{protein}_Highlight_Plot.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29ab6057",
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_plot(df: pd.DataFrame,\n",
    "                   column_x: str,\n",
    "                   column_y: str,\n",
    "                   hue: str,\n",
    "                   palette: tuple[str, str],\n",
    "                   order: tuple[str, str],\n",
    "                   x_ticks: tuple[int, int],\n",
    "                   output: str,\n",
    "                   ax,\n",
    "                  x_limit):\n",
    "    sns.set_style('ticks')\n",
    "    #L=5.8\n",
    "    #U=13\n",
    "    p = sns.stripplot(data=df,\n",
    "                      x=column_x,\n",
    "                      y=column_y,\n",
    "                      hue=hue,\n",
    "                      order=order,\n",
    "                      orient='h',\n",
    "                      jitter=False,\n",
    "                      s=15,\n",
    "                      palette=['blue', 'red'],\n",
    "                      alpha=0.5,\n",
    "                      linewidth=0,\n",
    "                      ax=ax)\n",
    "    ax.get_legend().remove()\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label  \n",
    "    ax.set_xlim(x_limit[0], x_limit[1])\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "\n",
    "def kde_plot(df,\n",
    "             column_x,\n",
    "             hue,\n",
    "             output,\n",
    "             x_limit,\n",
    "             y_limit,\n",
    "             ax,\n",
    "             show_x_ticks=False):\n",
    "    p = sns.kdeplot(data=df,\n",
    "                    x=column_x,\n",
    "                    hue=hue,\n",
    "                    fill=True,\n",
    "                    common_norm=False,\n",
    "                    palette=[\"blue\", 'red'],\n",
    "                    alpha=0.5,\n",
    "                    linewidth=8,\n",
    "                    ax=ax)\n",
    "    ax.set_xlim(x_limit[0], x_limit[1])\n",
    "    ax.set_ylim(y_limit[0], y_limit[1])\n",
    "    ax.get_legend().remove()\n",
    "    plt.legend('',frameon=False)\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label    \n",
    "    p.yaxis.major_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    if show_x_ticks:\n",
    "        p.set(xticks=[np.log(1000), np.log(10000), np.log(100000)])\n",
    "        p.set(xticklabels=['1000', '10000', '100000'])\n",
    "    else:\n",
    "        ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    return p\n",
    "\n",
    "groups = (\"Concat\",\"SNV\", \"Consensus\")\n",
    "for protein, highlights in ((\"CREB1\", CREB1_HIGHLIGHT),\n",
    "                            (\"EGR1\", EGR1_HIGHLIGHT)):\n",
    "    fig, ax_array = plt.subplots(4)\n",
    "    fig.set_size_inches(10, 10)\n",
    "    kde_x, kde_y = KDE_BOUNDS[protein]\n",
    "    plot_df = relabeled_dfs[protein].copy()\n",
    "    plot_df = plot_df[plot_df[\"Name\"].isin(set(highlights))]\n",
    "    plot_df = pd.melt(plot_df,\n",
    "                      id_vars=[\"Name\"],\n",
    "                      value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    highlight_plot(plot_df,\n",
    "                   \"value\",\n",
    "                   \"Name\",\n",
    "                   \"variable\",\n",
    "                   ('blue', 'red'),\n",
    "                   highlights,\n",
    "                   [np.log(1000), np.log(10000), np.log(100000)],\n",
    "                  f\"{OUTPUT_FOLDER_MAIN}/{protein}_Highlight_Plot.svg\",\n",
    "                  ax_array[0],\n",
    "                  kde_x)\n",
    "    protein_df = relabeled_dfs[protein].copy()\n",
    "    protein_df = pd.melt(protein_df, id_vars=[\"Name\", \"Group\"], value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    for idx, group in enumerate(groups):\n",
    "        plot_df = protein_df[protein_df[\"Group\"] == group]\n",
    "        plot_x_ticks = group == \"Consensus\"\n",
    "        kde_plot(plot_df,\n",
    "                 \"value\",\n",
    "                 \"variable\", \n",
    "                 f\"{OUTPUT_FOLDER_MAIN}/{protein}_{group}_kde_plot.svg\",\n",
    "                 kde_x,\n",
    "                 kde_y,\n",
    "                 ax_array[idx+1],\n",
    "                 plot_x_ticks)\n",
    "    plt.subplots_adjust(hspace=0.0000001)\n",
    "    plt.savefig(fname=f\"{OUTPUT_FOLDER_MAIN}/{protein}_KDE.svg\",format='svg')\n",
    "    plt.close()\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33134632",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_creb1 = list()\n",
    "sequences_creb1 = list(map(lambda x: x[:25], sequences_creb1))\n",
    "\n",
    "sequences_egr1 = list(panel_c_data[(panel_c_data[\"Name\"].str.startswith(\"DN_EGR1_CGCCCACGC\")) &\n",
    "                               (panel_c_data[\"Name\"].str.contains(\"-DN\"))][\"Sequence\"])\n",
    "sequences_egr1 = list(map(lambda x: x[:25], sequences_egr1))\n",
    "\n",
    "sequences_5u01 = panel_a_data[(panel_a_data[\"Protein\"] == \"CREB1\") &\n",
    "                              (panel_a_data[\"Name\"].str.startswith(\"RelA\"))]\n",
    "sequences_5u01 = list(sequences_5u01[\"Sequence\"].apply(lambda x: x[:25]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f82b2da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OLD #\n",
    "# Subset scatterplot meta data by panel a\n",
    "panel_a_meta = meta_data[meta_data[\"Panel\"] == \"A\"].reset_index(drop=True)\n",
    "panel_a_data = []\n",
    "for row in panel_a_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    df = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    # Add group information to the dataframe\n",
    "    df[\"Group\"] = df[\"Name\"].apply(lambda x: \"5U01\" if \"RelA\" in x else \"Other\")\n",
    "    # Transform plot range and ticker settings to ln scale\n",
    "    ln_plot_range = list(map(np.log, (row.Range_Start, row.Range_End)))\n",
    "    ln_plot_ticker = list(map(np.log,\n",
    "                              (row.Ticker_A, row.Ticker_B, row.Ticker_C)))\n",
    "    # Plot\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                  ln_plot_range,\n",
    "                                  10,\n",
    "                                  (PALETTE[2], PALETTE[0]),\n",
    "                                  ln_plot_ticker,\n",
    "                                 (\"Other\", \"5U01\"))\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_SUPP}/Scatterplot_{row.Protein}_A.svg\",\n",
    "               obj=plot)\n",
    "    df[\"Protein\"] = row.Protein\n",
    "    panel_a_data.append(df)\n",
    "panel_a_data = pd.concat(panel_a_data).reset_index(drop=True)\n",
    "panel_a_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47589004",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(data=panel_a_data,\n",
    "                x=\"CREB1_Non_UV\",\n",
    "                y=\"CREB1_UV\",\n",
    "                hue=\"Group\",\n",
    "                hue_order = [\"5U01\", \"Other\"],\n",
    "                palette = [\"#2b8cbe\", \"black\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbdc0e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Is this the real table? Or the first one?\n",
    "is_pdb = []\n",
    "for i in panel_a_data[\"Name\"]:\n",
    "    is_pdb.append(any([j in i for j in PDB_VAR_LIB]))\n",
    "panel_a_data = panel_a_data[is_pdb].reset_index(drop=True)\n",
    "panel_a_data\n",
    "\n",
    "table_panel_a = panel_a_data.copy()\n",
    "table_panel_a = table_panel_a[[\"Group\", \"Sequence\", \"Signal_Non_UV\", \"Signal_UV\", \"Protein\"]]\n",
    "table_panel_a[\"Sequence\"] = table_panel_a[\"Sequence\"].apply(lambda x: x[:25])\n",
    "table_panel_a = table_panel_a.sort_values(by=[\"Protein\", \"Group\"])\n",
    "table_panel_a.to_csv(f\"{OUTPUT_FOLDER_TABLE}/Table_Panel_A.csv\", index=False, float_format='%g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd722e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_a_data = pd.concat(panel_a_data).reset_index(drop=True)\n",
    "panel_a_data = panel_a_data[panel_a_data[\"Group\"].isin(PDB_VAR_LIB)]\n",
    "# Convert to ln space\n",
    "panel_a_data[\"Signal\"] = panel_a_data[\"Signal\"].apply(lambda x: np.log(x))\n",
    "panel_a_data = panel_a_data.pivot(index=[\"Group\",\"Name\", \"Sequence\"],\n",
    "                   values=\"Signal\",\n",
    "                   columns=[\"Protein\", \"Condition\"]).reset_index()\n",
    "panel_a_data.columns = [\"_\".join(i) for i in panel_a_data.columns]\n",
    "panel_a_data.columns = list(map(lambda x: x.rstrip('_'), panel_a_data.columns))\n",
    "panel_a_data.to_csv(f\"{OUTPUT_FOLDER_F6S}/Panel_A_Data.csv\")\n",
    "panel_a_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4b95298",
   "metadata": {},
   "outputs": [],
   "source": [
    "uv_scaling_regression_statistics = pd.concat(regression_reports)\n",
    "uv_scaling_regression_statistics.to_csv(f\"{OUTPUT_FOLDER}/Custom_Sequences_Scaling_Statistics.csv\")\n",
    "uv_scaling_regression_statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41acfdce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import namedtuple\n",
    "\n",
    "Regression_Report = namedtuple(\"Regression_Report\", [\"Condition_X\",\n",
    "                                                     \"Condition_Y\",\n",
    "                                                     \"Prescale_R2\",\n",
    "                                                     \"Prescale_Slope\",\n",
    "                                                     \"Prescale_Slope_CI\",\n",
    "                                                     \"Prescale_Intercept\",\n",
    "                                                     \"Prescale_Intercept_CI\",\n",
    "                                                     \"Postscale_R2\",\n",
    "                                                     \"Postscale_Slope\",\n",
    "                                                     \"Postscale_Slope_CI\",\n",
    "                                                     \"Postscale_Intercept\",\n",
    "                                                     \"Postscale_Intercept_CI\"])\n",
    "\n",
    "\n",
    "def generate_regression_report(prescale_regression,\n",
    "                               postscale_regression,\n",
    "                               len_values,\n",
    "                               column_x,\n",
    "                               column_y,\n",
    "                               ci_percent=0.95):\n",
    "    tinv = lambda p, df: abs(stats.t.ppf(p/2, df))\n",
    "    ts = tinv(1 - ci_percent, len_values-2)\n",
    "    return  Regression_Report(Condition_X=column_x,\n",
    "                             Condition_Y=column_y,\n",
    "                             Prescale_R2=prescale_regression.rvalue ** 2,\n",
    "                             Prescale_Slope=prescale_regression.slope,\n",
    "                             Prescale_Slope_CI=prescale_regression.stderr * ts,\n",
    "                             Prescale_Intercept=prescale_regression.intercept,\n",
    "                             Prescale_Intercept_CI=prescale_regression.intercept_stderr * ts,\n",
    "                             Postscale_R2=postscale_regression.rvalue ** 2,\n",
    "                             Postscale_Slope=postscale_regression.slope,\n",
    "                             Postscale_Slope_CI=postscale_regression.stderr * ts,\n",
    "                             Postscale_Intercept=postscale_regression.intercept,\n",
    "                             Postscale_Intercept_CI=postscale_regression.intercept_stderr * ts)\n",
    "\n",
    "\n",
    "def scale_column(dataframe: pd.DataFrame,\n",
    "                 column_name: str,\n",
    "                 slope: float,\n",
    "                 intercept: float) -> pd.DataFrame:\n",
    "    \"\"\"Scale a column in a dataframe.\n",
    "\n",
    "    Given a dataframe, column_name, slope, and intercept, scales a column's\n",
    "    values such that the slope is equal to 1.\n",
    "    \"\"\"\n",
    "    # Apply the transformation\n",
    "    dataframe[f\"{column_name}_Scaled\"] = dataframe[column_name]\\\n",
    "        .apply(lambda x: (x - intercept) / (slope))\n",
    "    dataframe = dataframe.reset_index(drop=True)\n",
    "    return dataframe\n",
    "\n",
    "def scale_on_no_pydi(dataframe, nouv_column, uv_column):\n",
    "    no_pydi_df = dataframe[~dataframe[\"Has_PyDi\"]]\n",
    "    prescale_regression = scipy.stats.linregress(no_pydi_df[nouv_column],\n",
    "                                                  no_pydi_df[uv_column])\n",
    "    dataframe[f\"{uv_column}_Scaled\"] = dataframe[uv_column]\\\n",
    "        .apply(lambda x: (x - prescale_regression.intercept) / (prescale_regression.slope))\n",
    "    post_no_pydi_df = dataframe[~dataframe[\"Has_PyDi\"]]\n",
    "    postscale_regression = scipy.stats.linregress(post_no_pydi_df[nouv_column],\n",
    "                                                   post_no_pydi_df[f\"{uv_column}_Scaled\"])  \n",
    "    regression_report = generate_regression_report(prescale_regression,\n",
    "                                                   postscale_regression,\n",
    "                                                   len(no_pydi_df),\n",
    "                                                   nouv_column,\n",
    "                                                   uv_column)\n",
    "    return (dataframe, regression_report)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7aec693",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e582c27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
