{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Signal_Increase_After_UV\n",
    "\n",
    "Analysis comparing non-UV and UV conditions to find that the 5U01 sequence seems to increase in binding for CREB1 and EGR1 after UV irradiation. \n",
    "\n",
    "1. Figure S6a; c-d\n",
    "2. Figure 4\n",
    "\n",
    "### Files\n",
    "The following files are used to generate the data:\n",
    "\n",
    "1. Initial observation:\n",
    "    1. CREB1_WC_ID0_alldata.txt\n",
    "    2. CREB1_UV_ID1_alldata.txt\n",
    "    3. EGR1_WC_ID2_alldata.txt\n",
    "    4. EGR1_UV_ID3_alldata.txt\n",
    "    1. CREB1_WC_ID0_7mers_1111111.txt\n",
    "    2. CREB1_UV_ID1_7mers_1111111.txt\n",
    "    3. EGR1_WC_ID2_7mers_1111111.txt\n",
    "    4. EGR1_UV_ID3_7mers_1111111.txt\n",
    "2. Follow-up experiments:\n",
    "    1. CREB1_WC_ID38_alldata.txt\n",
    "    2. CREB1_UV_ID39_alldata.txt\n",
    "    3. EGR1_WC_ID40_alldata.txt\n",
    "    4. EGR1_UV_ID41_alldata.txt\n",
    "3. Meta data file:\n",
    "    1. Signal_Increase_UV_5U01_Scatterplots.csv\n",
    "\n",
    "### Overview\n",
    "\n",
    "A scatterplot of a variation assay design for NDA sequences from protein-DNA structures in the protein databank is made for CREB1 and EGR1, revealing an observation that sequences for 5U01 seem to have higher signal in UV. Scatterplots for a set of follow-up experiments show that the increase is repeatable and if only a partial sequence is tested, a partial effect is given. Then the k-mers for the 5U01 sequences are compared in non_UV and UV conditions, showing tha they have higher E-scores in UV compared to Non-UV via a wilcoxon rank-sum test with a benjamini-hochberg correction for multiple tests. \n",
    "\n",
    "### 3rd Party Packages\n",
    "\n",
    "1. Bokeh - Plotting (Scatterplots)\n",
    "2. Matplotlib - Plotting (Boxplot)\n",
    "3. Numpy - ln function from np.log\n",
    "4. Pandas - Dataframe usage\n",
    "5. Seaborn - Plotting (Boxplots)\n",
    "6. Scipy - Linear regression\n",
    "7. Statsmodels: Multiple test correction\n",
    "\n",
    "### UV Bind Analysis Core Imports\n",
    "\n",
    "- uac.ols: Train OLS regression on sequences that cannot form pyrimidine dimers.\n",
    "- uac.scale_uv_on_non_uv: Scale UV probes on Non-UV by making the slope of a linear regression for sequences that cannot form pyrimidine dimers equal to 1. \n",
    "- uac.read_kmer_file_pair: Given a pair of k-mer files, reads them as dataframes and merges them into one.\n",
    "\n",
    "** Additional details can be found in the uvbind_analysis_core.py script.\n",
    "\n",
    "### Abbreviations:\n",
    "\n",
    "- pydi: Pyrimidine Dinucleotide\n",
    "- df: DataFrame\n",
    "- dn: Dinucleotide\n",
    "- idx: index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import annotations\n",
    "from collections import namedtuple\n",
    "import os\n",
    "\n",
    "from bokeh.io import export_svg\n",
    "from bokeh.plotting import figure, show, ColumnDataSource\n",
    "from bokeh.models import CDSView, GroupFilter\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import uvbind_analysis_core as uac\n",
    "\n",
    "ALLDATA_FOLDER = \"../../Data/AllData_Files\"\n",
    "KMER_FOLDER = \"../../Data/Kmer_Files\"\n",
    "SCATTERPLOT_META_DATA = \"Meta_Data/Signal_Increase_UV_5U01_Scatterplots.csv\"\n",
    "DN_DESIGN_META_DATA = \"Meta_Data/Array_86684_DN_Design.csv\"\n",
    "# Figure 6S Parameters\n",
    "PALETTE = (\"#2b8cbe\", \"#a6bddb\", \"black\", \"black\") \n",
    "BOXPLOT_PALETTE = (\"lightgrey\", \"grey\") # (Non_UV, UV)\n",
    "CREB1_KMERS = (\"CREB1_WC_ID0_7of9mers_1111111.txt\",\n",
    "               \"CREB1_UV_ID1_7of9mers_1111111.txt\")\n",
    "EGR1_KMERS = (\"EGR1_WC_ID2_7of9mers_1111111.txt\",\n",
    "               \"EGR1_UV_ID3_7of9mers_1111111.txt\")\n",
    "OUTPUT_FOLDER_F6S = \"../Figure_S6\"\n",
    "# Figure 4 parameters\n",
    "CREB1_HIGHLIGHT = (\"Clst_CREB1_C1314_TD25_O2\",\n",
    "                   \"Clst_CREB1_C604_TD25\",\n",
    "                   \"Clst_CREB1_C1039_TD25\")\n",
    "EGR1_HIGHLIGHT = (\"Clst_EGR1_C394_TD25\",\n",
    "                  \"Clst_EGR1_C537_TD25\",\n",
    "                  \"Clst_EGR1_C390_TD25\")\n",
    "KDE_BOUNDS = {\"CREB1\": ((5.8, 13),(0, 1.2)), # x_bounds, y_bounds\n",
    "              \"EGR1\": ((5.8, 13),(0, 4))}\n",
    "OUTPUT_FOLDER_F4 = \"../Figure_4\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatterplot of non-Universal probes: Figure S6a\n",
    "\n",
    "Scatterplot of the non-Universal probes in figure S6 for CREB1 and EGR1. The sequences corresponding to a single-nucleotide variation design of the DNA sequence in PDB ID 5U01 are highlighted in a different color. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Pandas' object has no attribute 'Range_Start'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [2]\u001b[0m, in \u001b[0;36m<cell line: 5>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m df[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGroup\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mName\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m5U01\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRelA\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m x \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mOther\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# Transform plot range and ticker settings to ln scale\u001b[39;00m\n\u001b[0;32m---> 12\u001b[0m ln_plot_range \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mmap\u001b[39m(np\u001b[38;5;241m.\u001b[39mlog, (\u001b[43mrow\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mRange_Start\u001b[49m, row\u001b[38;5;241m.\u001b[39mRange_End)))\n\u001b[1;32m     13\u001b[0m ln_plot_ticker \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mmap\u001b[39m(np\u001b[38;5;241m.\u001b[39mlog,\n\u001b[1;32m     14\u001b[0m                           (row\u001b[38;5;241m.\u001b[39mTicker_A, row\u001b[38;5;241m.\u001b[39mTicker_B, row\u001b[38;5;241m.\u001b[39mTicker_C)))\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# Plot\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Pandas' object has no attribute 'Range_Start'"
     ]
    }
   ],
   "source": [
    "# Read meta data and query for data related to panel A\n",
    "meta_data = pd.read_csv(SCATTERPLOT_META_DATA)\n",
    "panel_a_meta = meta_data[meta_data[\"Panel\"] == \"A\"].reset_index(drop=True)\n",
    "panel_a_data = []\n",
    "for row in panel_a_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    df = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    # Add group information to the dataframe\n",
    "    df[\"Group\"] = df[\"Name\"].apply(lambda x: \"5U01\" if \"RelA\" in x else \"Other\")\n",
    "    # Transform plot range and ticker settings to ln scale\n",
    "    ln_plot_range = list(map(np.log, (row.Range_Start, row.Range_End)))\n",
    "    ln_plot_ticker = list(map(np.log,\n",
    "                              (row.Ticker_A, row.Ticker_B, row.Ticker_C)))\n",
    "    # Plot\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                  ln_plot_range,\n",
    "                                  10,\n",
    "                                  (PALETTE[2], PALETTE[0]),\n",
    "                                  ln_plot_ticker,\n",
    "                                 (\"Other\", \"5U01\"))\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_F6S}/Scatterplot_{row.Protein}_A.svg\",\n",
    "               obj=plot)\n",
    "    df[\"Protein\"] = row.Protein\n",
    "    panel_a_data.append(df)\n",
    "panel_a_data = pd.concat(panel_a_data).reset_index(drop=True)\n",
    "panel_a_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Follow-up Scatterplot of non-Universal Probes: Figure S6c\n",
    "\n",
    "Given data from a follow-up experiment, Prediction intervals based on an OLS of non-damageable sequences in the library\n",
    "2. The 5U01 full and half sequences in 2 different colors\n",
    "3. Sequences that either contain a consensus sequence or a 1-nt variation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Signal_Non_UV</th>\n",
       "      <th>Has_PyDi</th>\n",
       "      <th>Signal_UV</th>\n",
       "      <th>Signal_UV_PreScale</th>\n",
       "      <th>Prediction_Upper</th>\n",
       "      <th>Prediction_Lower</th>\n",
       "      <th>Predicted</th>\n",
       "      <th>Group</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ClstD_CREB1_H0_TD25_S0</td>\n",
       "      <td>TTGGAAACCTTTGGAGGGAATTTCCCGCACACATACACATACACAC...</td>\n",
       "      <td>6.864848</td>\n",
       "      <td>True</td>\n",
       "      <td>8.414366</td>\n",
       "      <td>9.079890</td>\n",
       "      <td>7.284846</td>\n",
       "      <td>6.444850</td>\n",
       "      <td>6.864848</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ClstD_CREB1_H0_TD25_S0_O2</td>\n",
       "      <td>GGAAATTCCCTCCAAAGGTTTCCAACGCACACATACACATACACAC...</td>\n",
       "      <td>7.175490</td>\n",
       "      <td>True</td>\n",
       "      <td>8.861874</td>\n",
       "      <td>9.368455</td>\n",
       "      <td>7.593654</td>\n",
       "      <td>6.757326</td>\n",
       "      <td>7.175490</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ClstD_CREB1_H0_TD25_S1</td>\n",
       "      <td>TGGAAACCTTTGGAGGGAATTTCCGCGCACACATACACATACACAC...</td>\n",
       "      <td>6.980076</td>\n",
       "      <td>True</td>\n",
       "      <td>8.715496</td>\n",
       "      <td>9.274066</td>\n",
       "      <td>7.399360</td>\n",
       "      <td>6.560792</td>\n",
       "      <td>6.980076</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ClstD_CREB1_H0_TD25_S1_O2</td>\n",
       "      <td>CGGAAATTCCCTCCAAAGGTTTCCACGCACACATACACATACACAC...</td>\n",
       "      <td>7.193686</td>\n",
       "      <td>True</td>\n",
       "      <td>8.897863</td>\n",
       "      <td>9.391661</td>\n",
       "      <td>7.611751</td>\n",
       "      <td>6.775620</td>\n",
       "      <td>7.193686</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ClstD_CREB1_H0_TD25_S2</td>\n",
       "      <td>TGGAAACCTTTGGAGGGAATTTCCCCGCACACATACACATACACAC...</td>\n",
       "      <td>6.709304</td>\n",
       "      <td>True</td>\n",
       "      <td>8.450171</td>\n",
       "      <td>9.102978</td>\n",
       "      <td>7.130329</td>\n",
       "      <td>6.288280</td>\n",
       "      <td>6.709304</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>CREB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24098</th>\n",
       "      <td>kD_Zif268_F3_GTG_O2</td>\n",
       "      <td>CGGATATATACGCCCACACTATATACGCACACATACACATACACAC...</td>\n",
       "      <td>10.397315</td>\n",
       "      <td>True</td>\n",
       "      <td>10.766001</td>\n",
       "      <td>10.343306</td>\n",
       "      <td>10.902304</td>\n",
       "      <td>9.892325</td>\n",
       "      <td>10.397315</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24099</th>\n",
       "      <td>kD_Zif268_F3_TCG</td>\n",
       "      <td>TATATATCGTGGGCGTATATATCCGCGCACACATACACATACACAC...</td>\n",
       "      <td>10.337637</td>\n",
       "      <td>True</td>\n",
       "      <td>10.212022</td>\n",
       "      <td>9.980657</td>\n",
       "      <td>10.842228</td>\n",
       "      <td>9.833047</td>\n",
       "      <td>10.337637</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24100</th>\n",
       "      <td>kD_Zif268_F3_TCG_O2</td>\n",
       "      <td>CGGATATATACGCCCACGATATATACGCACACATACACATACACAC...</td>\n",
       "      <td>10.365601</td>\n",
       "      <td>True</td>\n",
       "      <td>10.035355</td>\n",
       "      <td>9.865007</td>\n",
       "      <td>10.870377</td>\n",
       "      <td>9.860825</td>\n",
       "      <td>10.365601</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24101</th>\n",
       "      <td>kD_Zif268_consensus</td>\n",
       "      <td>TATATAGCGTGGGCGTATATATCCGCGCACACATACACATACACAC...</td>\n",
       "      <td>10.503752</td>\n",
       "      <td>True</td>\n",
       "      <td>10.914084</td>\n",
       "      <td>10.440244</td>\n",
       "      <td>11.009477</td>\n",
       "      <td>9.998026</td>\n",
       "      <td>10.503752</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24102</th>\n",
       "      <td>kD_Zif268_consensus_O2</td>\n",
       "      <td>CGGATATATACGCCCACGCTATATACGCACACATACACATACACAC...</td>\n",
       "      <td>10.426513</td>\n",
       "      <td>True</td>\n",
       "      <td>11.005808</td>\n",
       "      <td>10.500289</td>\n",
       "      <td>10.931702</td>\n",
       "      <td>9.921325</td>\n",
       "      <td>10.426513</td>\n",
       "      <td>Other_Not_Plotted</td>\n",
       "      <td>EGR1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24103 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0         ClstD_CREB1_H0_TD25_S0   \n",
       "1      ClstD_CREB1_H0_TD25_S0_O2   \n",
       "2         ClstD_CREB1_H0_TD25_S1   \n",
       "3      ClstD_CREB1_H0_TD25_S1_O2   \n",
       "4         ClstD_CREB1_H0_TD25_S2   \n",
       "...                          ...   \n",
       "24098        kD_Zif268_F3_GTG_O2   \n",
       "24099           kD_Zif268_F3_TCG   \n",
       "24100        kD_Zif268_F3_TCG_O2   \n",
       "24101        kD_Zif268_consensus   \n",
       "24102     kD_Zif268_consensus_O2   \n",
       "\n",
       "                                                Sequence  Signal_Non_UV  \\\n",
       "0      TTGGAAACCTTTGGAGGGAATTTCCCGCACACATACACATACACAC...       6.864848   \n",
       "1      GGAAATTCCCTCCAAAGGTTTCCAACGCACACATACACATACACAC...       7.175490   \n",
       "2      TGGAAACCTTTGGAGGGAATTTCCGCGCACACATACACATACACAC...       6.980076   \n",
       "3      CGGAAATTCCCTCCAAAGGTTTCCACGCACACATACACATACACAC...       7.193686   \n",
       "4      TGGAAACCTTTGGAGGGAATTTCCCCGCACACATACACATACACAC...       6.709304   \n",
       "...                                                  ...            ...   \n",
       "24098  CGGATATATACGCCCACACTATATACGCACACATACACATACACAC...      10.397315   \n",
       "24099  TATATATCGTGGGCGTATATATCCGCGCACACATACACATACACAC...      10.337637   \n",
       "24100  CGGATATATACGCCCACGATATATACGCACACATACACATACACAC...      10.365601   \n",
       "24101  TATATAGCGTGGGCGTATATATCCGCGCACACATACACATACACAC...      10.503752   \n",
       "24102  CGGATATATACGCCCACGCTATATACGCACACATACACATACACAC...      10.426513   \n",
       "\n",
       "       Has_PyDi  Signal_UV  Signal_UV_PreScale  Prediction_Upper  \\\n",
       "0          True   8.414366            9.079890          7.284846   \n",
       "1          True   8.861874            9.368455          7.593654   \n",
       "2          True   8.715496            9.274066          7.399360   \n",
       "3          True   8.897863            9.391661          7.611751   \n",
       "4          True   8.450171            9.102978          7.130329   \n",
       "...         ...        ...                 ...               ...   \n",
       "24098      True  10.766001           10.343306         10.902304   \n",
       "24099      True  10.212022            9.980657         10.842228   \n",
       "24100      True  10.035355            9.865007         10.870377   \n",
       "24101      True  10.914084           10.440244         11.009477   \n",
       "24102      True  11.005808           10.500289         10.931702   \n",
       "\n",
       "       Prediction_Lower  Predicted              Group Protein  \n",
       "0              6.444850   6.864848  Other_Not_Plotted   CREB1  \n",
       "1              6.757326   7.175490  Other_Not_Plotted   CREB1  \n",
       "2              6.560792   6.980076  Other_Not_Plotted   CREB1  \n",
       "3              6.775620   7.193686  Other_Not_Plotted   CREB1  \n",
       "4              6.288280   6.709304  Other_Not_Plotted   CREB1  \n",
       "...                 ...        ...                ...     ...  \n",
       "24098          9.892325  10.397315  Other_Not_Plotted    EGR1  \n",
       "24099          9.833047  10.337637  Other_Not_Plotted    EGR1  \n",
       "24100          9.860825  10.365601  Other_Not_Plotted    EGR1  \n",
       "24101          9.998026  10.503752  Other_Not_Plotted    EGR1  \n",
       "24102          9.921325  10.426513  Other_Not_Plotted    EGR1  \n",
       "\n",
       "[24103 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Named tuple for group names based of dinucelotide variant designs\n",
    "Dinucleotide_Group = namedtuple(\"DN_Group\", [\"consensus\", \"snv\", \"dnv\"])\n",
    "\n",
    "def dinucleotide_group_names(query_df: pd.DataFrame,\n",
    "                             design_df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Labels sequences into consensus, snv, or dnv across multiple groups.\n",
    "    \n",
    "    For dinucleotide variation assays, probes can be in one of the following\n",
    "    categories:\n",
    "    \n",
    "    1. Consensus: The sequence is the WT or has a variation outside the binding\n",
    "    site.\n",
    "    2. SNV: The sequence has a 1-nt variation in the binding site\n",
    "    3. DNV: The sequences has a 2-nt variation in the binding site\n",
    "    \n",
    "    A namedtuple is returned containing sets of sequence names belonging to each\n",
    "    group.\n",
    "    \"\"\"\n",
    "    consensus_set = set()\n",
    "    snv_set = set()\n",
    "    dnv_set = set()\n",
    "    for row in design_df.itertuples():\n",
    "        # Query for a group of sequences\n",
    "        df = query_df[query_df[\"Name\"].str.startswith(row.Group)]\n",
    "        df = df.reset_index(drop=True)\n",
    "        for name, sequence in zip(df[\"Name\"], df[\"Sequence\"]):\n",
    "            binding_site = sequence[row.Start:row.End]\n",
    "            if binding_site == row.Binding_Site:\n",
    "                consensus_set.add(name)\n",
    "            elif uac.hamming_distance(binding_site, row.Binding_Site) == 1:\n",
    "                snv_set.add(name)\n",
    "            elif uac.hamming_distance(binding_site, row.Binding_Site) == 2:\n",
    "                dnv_set.add(name)\n",
    "            else:\n",
    "                raise ValueError(\"Unexpected difference found\")\n",
    "    return Dinucleotide_Group(consensus_set, snv_set, dnv_set)\n",
    "\n",
    "# Read meta data and query for data related to panel C\n",
    "panel_c_meta = meta_data[meta_data[\"Panel\"] == \"C\"].reset_index(drop=True)\n",
    "consensus_design = pd.read_csv(DN_DESIGN_META_DATA)\n",
    "panel_c_data = []\n",
    "for row in panel_c_meta.itertuples():\n",
    "    # Read the comparison, scale UV on non-UV\n",
    "    df = uac.scale_uv_on_non_uv(f\"{ALLDATA_FOLDER}/{row.File_Non_UV}\",\n",
    "                                f\"{ALLDATA_FOLDER}/{row.File_UV}\")\n",
    "    # Add OLS analysis\n",
    "    df = uac.ols(df, \"Signal_Non_UV\", \"Signal_UV\", 0.01, \"Has_PyDi\")\n",
    "    # Define sequences with a consensus site or a snv in the consensus\n",
    "    consensus_df = consensus_design[consensus_design[\"Protein\"] == row.Protein]\n",
    "    dn_groups = dinucleotide_group_names(df, consensus_df)\n",
    "    # Label Groups\n",
    "    groups = []\n",
    "    for name in df[\"Name\"]:\n",
    "        if name.startswith(\"DN_RELA_Double\"):\n",
    "            groups.append(\"Full_5U01\")\n",
    "        elif name.startswith(\"DN_RELA_Single\"):\n",
    "            groups.append(\"Half_5U01\")\n",
    "        elif name in dn_groups.consensus:\n",
    "            groups.append(\"Consensus\")\n",
    "        elif name in dn_groups.snv:\n",
    "            groups.append(\"SNV\")\n",
    "        else:\n",
    "            groups.append(\"Other_Not_Plotted\")\n",
    "    df[\"Group\"] = groups\n",
    "    # Transform plot range and ticker settings to ln scale\n",
    "    ln_plot_range = list(map(np.log, (row.Range_Start, row.Range_End)))\n",
    "    ln_plot_ticker = list(map(np.log,\n",
    "                              (row.Ticker_A, row.Ticker_B, row.Ticker_C)))\n",
    "    # Plot\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                 ln_plot_range,\n",
    "                                 10,\n",
    "                                 (\"#2b8cbe\", \"#a6bddb\", \"black\", \"black\"),\n",
    "                                 ln_plot_ticker,\n",
    "                                 (\"Full_5U01\", \"Half_5U01\", \"Consensus\", \"SNV\"),\n",
    "                                 True)\n",
    "    plot.output_backend = \"svg\"\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_F6S}/Scatterplot_{row.Protein}_C.svg\",\n",
    "               obj=plot)\n",
    "    df[\"Protein\"] = row.Protein\n",
    "    panel_c_data.append(df)\n",
    "panel_c_data = pd.concat(panel_c_data).reset_index(drop=True)\n",
    "panel_c_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### E-score comparison: Figure S6d\n",
    "\n",
    "Given the increase is repeatable, could the k-mer data between non-UV and UV offer an explanation as to why the increase occurs? This analysis:\n",
    "\n",
    "1. Retrieves all 5U01 sequences and finds all unique k-mers. \n",
    "2. Subsets the CREB1 and EGR1 k-mer data by thes k-mers\n",
    "3. Determines if there a difference between non-UV and UV conditions for these k-mers\n",
    "\n",
    "A wilcoxon rank-sum test and benjamini-hochberg correction results are saved as a table and a boxplot of the data is saved as a svg file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEGCAYAAAB7DNKzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUkElEQVR4nO3df5BdZ33f8fdXPyyvZRJk2cDaxt5VZEhNSt10ATeUIiLcYifFyTStTVowmWRdPEAgIRR73WnTJpHdSRvCTEg9WkMwlImUBEocRoQxGyu4FFrLxmRia4ztXbBV30pCUoxZLYpW+vaPe4Sv13efXe3uvedq9/2a2dl7znnuOd/dubuf+5zn3OdEZiJJ0mxW1V2AJKm3GRSSpCKDQpJUZFBIkooMCklS0Zq6C+iE888/PwcGBuouQ5LOGA888MB3MvOCdtuWZVAMDAywZ8+eusuQpDNGRHx7tm2eepIkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklRkUEiSigwKSafl8OHD3HLLLRw5cqTuUtQly3L22LqNjo4yMTGxqH00Gg0A+vv7F7yPwcFBhoeHF1WHNNPOnTt55JFH2LFjBzfddFPd5agL7FH0qKmpKaampuouQ3qew4cPMzY2RmYyNjZmr2KFsEfRAUvxLn5kZASAbdu2LXpfdbJ3tbzs3LmTkydPAnDy5El7FSuEPQr1PHtXvWP37t1MT08DMD09ze7du+stSF1hj0IdZe9qedmyZQv33HMP09PTrFmzhi1bttRdkrrAoJBWmMWcDjx+/PgPehQnTpxgfHz8B0F+unrhdOBiT40uxWlR6I3fRYlBIWne1q5dy+rVqzlx4gQbNmxg7dq1dZdUq5VySrTWoIiItwAfAVYDd2bm7bO0ew3wNeC6zPyTLpYoLTuLfef6wQ9+kKeeeooPf/jDbNiwYYmqqsdifxcr5bRobYPZEbEa+ChwNXA58LaIuHyWdv8Z+GJ3K5TUztq1a9m0adMZHxKavzqvenot8Hhmjmfm3wI7gGvbtHsv8BngQDeLkyQ11RkUFwFPtSzvq9b9QERcBPwscMdcO4uIGyNiT0TsOXjw4JIWKkkrWZ1BEW3W5Yzl3wU+lJkn5tpZZm7PzKHMHLrggguWoj5JEvUOZu8DXt6yfDHw9Iw2Q8COiAA4H7gmIqYz83NdqVCSVGtQ3A9cFhGDwP8Frgd+vrVBZg6eehwRnwA+b0hIUnfVFhSZOR0R76F5NdNq4OOZ+XBEvKvaPue4hCSp82r9HEVm7gJ2zVjXNiAy853dqEmS9HxOCihJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSkUEhSSoyKCRJRQaFJKnIoJAkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklRkUEiSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkorW1F2AJC3E6OgoExMTtdYwPj4OwMjISK11DA4OMjw83LH9GxSSzkgTExPs3buXvr6+2mo4fvw4AN/61rdqq2Fqaqrjx6g1KCLiLcBHgNXAnZl5+4zt/wr4ULX4PeCmzPxGd6uU1Kv6+vp45StfWXcZtXr00Uc7fozaxigiYjXwUeBq4HLgbRFx+YxmE8AbM/PVwG8A27tbpSSpzsHs1wKPZ+Z4Zv4tsAO4trVBZv6vzDxSLX4NuLjLNUrSilfnqaeLgKdalvcBryu0/0XgC7NtjIgbgRsBLrnkkqWoTzhg2KrTA4ZSr6ozKKLNumzbMOJNNIPiH822s8zcTnVqamhoqO1+dPocMGzqxoCh1KvqDIp9wMtbli8Gnp7ZKCJeDdwJXJ2Zh7pUm1o4YNidAUOpV9U5RnE/cFlEDEbEWcD1wN2tDSLiEuCzwNsz85s11ChJK15tPYrMnI6I9wBfpHl57Mcz8+GIeFe1/Q7g3wMbgd+PCIDpzByqq2ZJWolq/RxFZu4Cds1Yd0fL418CfqnbdUmSnuNcT5KkIoNCklRkUEiSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpCJvhSrpjNRoNDh69OiKn7Dx6NGjNBqNjh7DHoUkqcgehaQzUn9/P8eOHXMK/Ecfpb+/v6PHsEchSSoyKCRJRQaFJKnIoJAkFTmYLZ1BRkdHmZiYqLWG8fFxAEZGRmqto9OXhOo5BoV0BpmYmODxxx/v+FUuJWvWNP9tTE5O1lZDo9Fg1apVrFrlSZFuMCikM0x/fz/Dw8N1l1Gr0dFR9u/fX3cZK4ZxLEkqMigkSUUGhSSpyKCQJBUZFJKkIoNCklTk5bEqcs7/pm7M+S/1qjl7FBHx0oj4WER8oVq+PCJ+sfOlSZJ6wXx6FJ8A/gC4tVr+JrAT+FiHalIPcc7/pm7M+S/1qvmMUZyfmX8EnATIzGngREerkiT1jPkExWREbAQSICKuBJ7paFWSpJ4xn1NPvwrcDfxIRHwFuAD4uY5WJUnqGXMGRWY+GBFvBF4JBPBoZh7veGWSNIepqalar8g7duwYAOvWrauthqmpqY4fY86giIh3zFj14xFBZn6yQzVJ0pz6+vrYtGlTrTWcujfHwMBArXUMDg52dP/zOfX0mpbHZwNbgQeBRQdFRLwF+AiwGrgzM2+fsT2q7dcAR4F3ZuaDiz2upDNff38/27Ztq7WGUzdvqruOTpvPqaf3ti5HxA8Dn1rsgSNiNfBR4CpgH3B/RNydmY+0NLsauKz6eh3w36rvHeVdxJ7jh8wkLeST2Udp/uNerNcCj2fmOEBE7ACuBVqD4lrgk5mZwNci4sUR0Z+ZHf3v5V3EmryLmCSY3xjFn1FdGkvzctrLgT9agmNfBDzVsryPF/YW2rW5CHhBUETEjcCNAJdccsmii/MuYt5FTFLTfHoU/6Xl8TTw7czctwTHjjbrcgFtmisztwPbAYaGhtq2kSSdvvmMUfxlh469D3h5y/LFwNMLaCNJ6qBZTz5HxLMR8d02X89GxHeX4Nj3A5dFxGBEnAVcT/ODfa3uBt4RTVcCz3R6fEKS9Hyz9igy80WdPHBmTkfEe4Av0rw89uOZ+XBEvKvafgewi+alsY/THET/hU7WJPW6RqPB5OQko6OjdZdSq0ajwfr16+suY8WY91VPEfESmp+jACAzn1zswTNzF80waF13R8vjBN692ONIkhZuPlc9vRX4r8CFwAHgUmAv8KrOliZppv7+fiYnJ70ib3TUHkUXzecC+d8ArgS+mZmDND+Z/ZWOViVJ6hnzCYrjmXkIWBURqzLzXuCKzpYlSeoV8xmj+JuIOBe4D/h0RByg+XkKSdIKMJ8exZeBFwPvA/4ceAL4Zx2sSZLUQ+YTFEHzEtbdwLnAzupUlCRpBZgzKDLzP2bmq2hepnoh8JcR8aWOVyZJ6gmnMy3oAeD/AYeAl3SmHElSr5kzKCLipojYDYwB5wPDmfnqThcmSeoN87nq6VLg/Zn5UIdrkST1oPnMHntzNwqRJPUmb10mSSoyKCRJRQaFJKnIoJAkFRkUkqQig0KSVGRQSJKKDApJUpFBIUkqMigkSUUGhSSpaD6TAkrqIY1Gg9HR0dqOf+hQ875lGzdurK2GRqPB5s2bazv+SmNQSGeQwcHBuktg//79AKxfv762GjZv3twTv4uVwqDQnKampnj00UdrO/6xY8cAWLduXW01TE1N1XbsVsPDw3WXwMjICADbtm2ruRJ1i0Ghor6+PjZt2lRrDePj4wAMDAzUWofvYLVSGRQq6u/vr/2do+9gpXoZFG00Gg0mJydrHTDsBY1Go9bz0JJ6g5fHSpKK7FG00d/fz+TkZE8MHNZpdHTUHoUkexSSpDKDQpJUZFBIkopqCYqIOC8i7omIx6rvG9q0eXlE3BsReyPi4Yh4Xx21StJKV1eP4mZgLDMvA8aq5ZmmgQ9k5t8BrgTeHRGXd7FGSRL1BcW1wF3V47uAn5nZIDMbmflg9fhZYC9wUbcKlCQ11RUUL83MBjQDAXhJqXFEDAB/H/jfhTY3RsSeiNhz8ODBpaxVkla0jn2OIiK+BLyszaZbT3M/5wKfAd6fmd+drV1mbge2AwwNDeXpHEOSNLuOBUVmvnm2bRGxPyL6M7MREf3AgVnaraUZEp/OzM92qFRJUkFdp57uBm6oHt8A/OnMBhERwMeAvZn5O12sTZLUoq6guB24KiIeA66qlomICyNiV9Xm9cDbgZ+MiIeqr2vqKVeSVq5a5nrKzEPA1jbrnwauqR7/TyC6XJokaQY/mS1JKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSUS0fuJOkXjA6OsrExMSCnz8+Pg7AyMjIouoYHBxkeHh4UfvoJINCkhaor6+v7hK6wqCQtGL18rv4XuIYhSSpyB7FLBqNBqOjo7Ud/9ChQwBs3LixthoajQabN2+u7fiSeoNB0cbg4GDdJbB//34A1q9fX1sNmzdv7onfhaR6GRRt9MJ5y1NXUWzbtq3mSiStdAaFtMJ4SahOl4PZkk5LX1/firksdC6HDx/mlltu4ciRI3WX0lH2KKQVxnfxS2fnzp088sgj7Nixg5tuuqnucjrGHoUkLcDhw4cZGxsjMxkbG1vWvQqDQpIWYOfOnZw8eRKAkydPsmPHjpor6hyDQpIWYPfu3UxPTwMwPT3N7t276y2ogwwKSVqALVu2sGZNc5h3zZo1bNmypd6COsigkKQFuO6661i1qvkvdNWqVVx//fU1V9Q5BoUkLcB5553H1q1biQi2bt3Khg0b6i6pY7w8VpIW6LrrruPJJ59c1r0JMCgkacHOO+88brvttrrL6DhPPUmSigwKSVKRQSFJKjIoJElFBoUkqcigkCQVGRSSpKJagiIizouIeyLiser7rB9pjIjVEfH1iPh8N2uUJDXV1aO4GRjLzMuAsWp5Nu8D9nalKknSC9QVFNcCd1WP7wJ+pl2jiLgY+Cngzu6UJUmaqa6geGlmNgCq7y+Zpd3vAv8WODnXDiPixojYExF7Dh48uGSFStJK17G5niLiS8DL2my6dZ7P/2ngQGY+EBFb5mqfmduB7QBDQ0M5/0olSSUdC4rMfPNs2yJif0T0Z2YjIvqBA22avR54a0RcA5wN/FBE/PfM/NcdKlmS1EZdp57uBm6oHt8A/OnMBpl5S2ZenJkDwPXAXxgSktR9dQXF7cBVEfEYcFW1TERcGBG7aqpJktRGLfejyMxDwNY2658Grmmzfjewu+OFSZJewE9mS5KKDApJUpFBIUkqMigkSUW1DGZr5RgdHWViYmJR+xgfHwdgZGRkwfsYHBxkeHh4UXVIK5VBoZ7X19dXdwnSihaZy2+2i6GhodyzZ09tx1/Kd9GbNm1a8D58Fy1pviLigcwcarfNHkWP8l20pF5hUHSA7+IlLSde9SRJKjIoJElFBoUkqcigkCQVGRSSpCKDQpJUZFBIkooMCklSkUEhSSoyKCRJRctyUsCIOAh8u+46lonzge/UXYQ0C1+fS+fSzLyg3YZlGRRaOhGxZ7YZJaW6+frsDk89SZKKDApJUpFBoblsr7sAqcDXZxc4RiFJKrJHIUkqMigkSUUGxQoQES+LiB0R8UREPBIRuyLiFRExFREPVes+GRFrq/ZbIuKZatuprzdX205Uy9+IiAcj4idajvPnEfE3EfH5un5WnblaXlunvm6u1q+JiG0R8VjLtlvbPO+vI+LPIuLFLdt8TS4B75m9zEVEAP8DuCszr6/WXQG8FHgiM6+IiNXAPcC/BD5dPfW+zPzpNrucyswrqv38U+A24I3Vtt8GzgH+TWd+Gi1zP3htzfCbwMuAv5uZ34+IFwEfaPe8iLgLeDfwW9U2X5NLwB7F8vcm4Hhm3nFqRWY+BDzVsnwC+D/ARae57x8CjrTsZwx4djHFSq0i4hxgGHhvZn4fIDOfzcxfn+UpX6XldexrcmnYo1j+fgx4oNQgIs4GXge8r2X1GyLioZblf56ZTwB91fqzgX7gJ5e0Wq1kfTNec7cBe4EnM3POf/ZVz3gr8LHOlLdyGRQr249Uf5iXAX+SmX/Vsm0+p57+IfDJiPix9DprLd4LTj1FxKtnLP8CzTc0G4GfyMyneC5gBmi+KbqnG8WuJJ56Wv4eBv7BLNueqP4wNwNXRsRbT2fHmflVmpOytZ1ITFoCjwOXVOMSZOYfVK/ZZ4DVVZtTAXMpcBbNMQotIYNi+fsLYF1EDJ9aERGvoflHBUBmNoCbgVtOZ8cR8aM0/1gPLU2p0vNl5lGap5J+rzpFeuoU01lt2j4D/DLwa6eu4NPSMCiWueqU0M8CV1WXxz4M/Drw9IymnwPOiYg3VMtvmHGp4s9V6/tOrQN2AjdUg+FExH3AHwNbI2JfdVWUNF99M15zt1frbwUawF9HxNeB+4C7eOFrmMz8OvAN4NQVfr4ml4BTeEiSiuxRSJKKDApJUpFBIUkqMigkSUUGhSSpyKCQ5tAyO+nD1ay5vxoRxb+diBiIiJ/vQC3vr+Y/krrGoJDmNpWZV2Tmq4CrgGuA/zDHcwaAJQ8K4P00Z0OVusagkE5DZh4AbgTeE00DEXFfdW+O1vtz3M5zH1r8ldnaRUR/RHy55X4Kb6jW/5OI+GrV9o8j4tyI+GXgQuDeiLi3jp9fK5MfuJPmEBHfy8xzZ6w7AvwozSmsT1b3SbgM+MPMHIqILcCvnZpYsTpd1K7dB4CzM/O3qqkpzgHWAZ8Frs7MyYj4ELAuM/9TRHwLGMrM73Tlh5dw9lhpoaL6vpbmPERXACeAV8zSfrZ29wMfr+Ym+lxmPhQRbwQuB77SvO8UZ9G8z4JUC4NCOk0RsYnmP/sDNMcq9gN/j+ap3O/P8rRfadcuM78cEf8Y+CngUxHx2zRvBnVPZr6tkz+HNF+OUUinISIuAO4Afq+acPGHgUZmngTeznNTXz8LvKjlqW3bRcSlwIHMHKU5S+qPA18DXh8Rm6s250TEK2bZr9Rx9iikuZ26Mc5aYBr4FPA71bbfBz4TEf8CuBeYrNb/FTAdEd8APlFotwX4YEQcB74HvCMzD0bEO4E/jIh1Vbt/B3wT2A58ISIamfmmzvy40vM5mC1JKvLUkySpyKCQJBUZFJKkIoNCklRkUEiSigwKSVKRQSFJKvr/cgoKAK12b7gAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Get 5u01 sequences from panel A as a list\n",
    "sequences_5u01 = panel_a_data[(panel_a_data[\"Protein\"] == \"CREB1\") &\n",
    "                              (panel_a_data[\"Name\"].str.startswith(\"RelA\"))]\n",
    "sequences_5u01 = list(sequences_5u01[\"Sequence\"].apply(lambda x: x[:25]))\n",
    "# Determine all unique 7-mers from the list of 5u01 sequences\n",
    "unique_5u01_kmers = set()\n",
    "for sequence in sequences_5u01:\n",
    "    unique_5u01_kmers.update(uac.unique_kmers(sequence, 7))\n",
    "# For each transcription factor\n",
    "kmer_dfs = []\n",
    "for kmer_files, tf in ((CREB1_KMERS, \"CREB1\"),\n",
    "                         (EGR1_KMERS, \"EGR1\")):\n",
    "    # Read the k-mer data as a pd.DataFrame\n",
    "    non_uv_file, uv_file = kmer_files\n",
    "    kmers = uac.read_kmer_file_pair(f\"{KMER_FOLDER}/{non_uv_file}\",\n",
    "                                    f\"{KMER_FOLDER}/{uv_file}\",\n",
    "                                    (\"_Non_UV\", \"_UV\"))\n",
    "    # Subset it by those k-mers present in 5U01 sequences\n",
    "    kmers_5u01 = kmers[(kmers[\"7-mer\"].isin(unique_5u01_kmers)) |\n",
    "                      (kmers[\"7-mer.1\"].isin(unique_5u01_kmers))]\n",
    "    kmers_5u01 = kmers_5u01.reset_index(drop=True)\n",
    "    # Add information on the TF, append it to the dataframes processed\n",
    "    kmers_5u01[\"Dataset\"] = tf\n",
    "    kmer_dfs.append(kmers_5u01)\n",
    "# Concatintate dataframes into one\n",
    "kmer_measurements_5u01 = pd.concat(kmer_dfs).reset_index(drop=True)\n",
    "# Compare groups with wilcoxon rank sum test\n",
    "wilcoxon_results = []\n",
    "for dataset, df in kmer_measurements_5u01.groupby(by=\"Dataset\"):\n",
    "    ranksum_result = stats.ranksums(df[\"E-score_Non_UV\"], df[\"E-score_UV\"])\n",
    "    wilcoxon_results.append((dataset, ranksum_result.pvalue))\n",
    "# Store the results as a pd.DataFrame\n",
    "escore_comparison = pd.DataFrame(wilcoxon_results)\n",
    "escore_comparison = escore_comparison.rename(columns={0:\"Protein\", 1:\"P_Value\"})\n",
    "# Benjamini correction for multiple tests to produce false discovery rates\n",
    "fdr_values = multipletests(escore_comparison[\"P_Value\"], method=\"fdr_bh\")[1]\n",
    "escore_comparison[\"FDR\"] = fdr_values\n",
    "# Save results\n",
    "escore_comparison.to_csv(f\"{OUTPUT_FOLDER_F6S}/Panel_E_Statistics.csv\")\n",
    "# Create a boxplot of the data and save\n",
    "# Transform the data to long form\n",
    "plot_df = pd.melt(kmer_measurements_5u01,\n",
    "                  id_vars=['7-mer', \"Dataset\"],\n",
    "                  value_vars=['E-score_Non_UV', \"E-score_UV\"])\n",
    "# Create a boxplot\n",
    "sns.boxplot(x=\"Dataset\", y=\"value\", hue=\"variable\",\n",
    "            data=plot_df,\n",
    "            palette = BOXPLOT_PALETTE,\n",
    "           hue_order=[\"E-score_Non_UV\", \"E-score_UV\"])\n",
    "plt.ylim(-0.52, 0.52)\n",
    "plt.legend([],[], frameon=False)\n",
    "plt.savefig(f\"{OUTPUT_FOLDER_F6S}/Panel_E_Boxplot.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatinated k-mer site scatterplots: FIgure 4a\n",
    "\n",
    "Scatterplot comparing sequences made of concatinated k-mers that had higher E-scores in UV conditions than non-UV conditions (UV-preffered). The plots are colored in the following priority:\n",
    "\n",
    "1. Green: Sequences with a consensus sequence of the given transcription factor\n",
    "2. Light Green: Sequences with a 1-nt variation from the consensus sequences\n",
    "3. Black: Sequences made from concatinated UV-preffered k-mers\n",
    "4. Grey: Sequences used to train the OLS model\n",
    "\n",
    "The data uses the same information as panel C for Figure S6, with the only changes being the sequences being plotted. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "relabeled_dfs = {}\n",
    "for protein in (\"CREB1\", \"EGR1\"):\n",
    "    df = panel_c_data[panel_c_data[\"Protein\"] == protein]\n",
    "    df = df.reset_index(drop=True)\n",
    "    # Rename groups to plot\n",
    "    groups = []\n",
    "    for row in df.itertuples():\n",
    "        if row.Name.startswith(f\"Clst_{protein}\"):\n",
    "            groups.append(\"Concat\")\n",
    "        elif (row.Has_PyDi is False) and (row.Group not in (\"Consensus\", \"SNV\")):\n",
    "            groups.append(\"Training\")\n",
    "        else:\n",
    "            groups.append(row.Group)\n",
    "    df[\"Group\"] = groups\n",
    "    plot = uac.scatterplot_of_groups(df,\n",
    "                                     ln_plot_range,\n",
    "                                     10,\n",
    "                                     (\"lightgrey\", \"black\", \"green\", \"lightgreen\"),\n",
    "                                     ln_plot_ticker,\n",
    "                                     (\"Training\", \"Concat\", \"Consensus\", \"SNV\"),\n",
    "                                     True)\n",
    "    plot.output_backend = 'svg'\n",
    "    export_svg(filename=f\"{OUTPUT_FOLDER_F4}/Scatterplot_{protein}.svg\",\n",
    "               obj=plot)\n",
    "    relabeled_dfs[protein] = df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4b - Highlights and KDE comparisons\n",
    "\n",
    "This panel is comprised of a figure with the following sub-plots:\n",
    "\n",
    "1. A plot at the top showing the change from non-UV to UV for 3 selected sequences as highlights\n",
    "2. A set of plots showing KDE comparisons for the following groups:\n",
    "    1. Consensus sequences\n",
    "    2. 1-nt variation sequences\n",
    "    3. Concatinated UV-preffered k-mer sequences\n",
    "\n",
    "These components are then combined into a single plot which shows the data at the same scale to compare across componenets. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Highlight subpanels\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_plot(df: pd.DataFrame,\n",
    "                   column_x: str,\n",
    "                   column_y: str,\n",
    "                   hue: str,\n",
    "                   palette: tuple[str, str],\n",
    "                   order: tuple[str, str],\n",
    "                   x_ticks: tuple[int, int],\n",
    "                   output: str):\n",
    "    rcParams['figure.figsize'] = 10, 2.5\n",
    "    sns.set_style('ticks')\n",
    "    L=5.8\n",
    "    U=13\n",
    "    p = sns.stripplot(data=df,\n",
    "                      x=column_x,\n",
    "                      y=column_y,\n",
    "                      hue=hue,\n",
    "                      order=order,\n",
    "                      orient='h',\n",
    "                      jitter=False,\n",
    "                      s=15,\n",
    "                      palette=['red', 'blue'],\n",
    "                      alpha=0.5,\n",
    "                      linewidth=0)\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label    \n",
    "    p.set(xticks=[np.log(1000), np.log(10000), np.log(100000)])\n",
    "    p.set(xticklabels=['1000', '10000', '100000'])  # remove the tick labels\n",
    "    p.set(yticklabels=[])  # remove the tick labels\n",
    "    plt.xlim(L, U)\n",
    "    plt.legend('',frameon=False)\n",
    "    plt.savefig(output, format=\"svg\")\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "for protein, highlights in ((\"CREB1\", CREB1_HIGHLIGHT),\n",
    "                            (\"EGR1\", EGR1_HIGHLIGHT)):\n",
    "    plot_df = relabeled_dfs[protein].copy()\n",
    "    plot_df = plot_df[plot_df[\"Name\"].isin(set(highlights))]\n",
    "    plot_df = pd.melt(plot_df,\n",
    "                      id_vars=[\"Name\"],\n",
    "                      value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    highlight_plot(plot_df,\n",
    "                   \"value\",\n",
    "                   \"Name\",\n",
    "                   \"variable\",\n",
    "                   ('blue', 'red'),\n",
    "                   highlights,\n",
    "                   [np.log(1000), np.log(10000), np.log(100000)],\n",
    "                  f\"{OUTPUT_FOLDER_F4}/{protein}_Highlight_Plot.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### KDE comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kde_plot(df,\n",
    "             column_x,\n",
    "             hue,\n",
    "             output,\n",
    "             x_limit,\n",
    "             y_limit,\n",
    "             show_x_ticks=False):\n",
    "    p = sns.kdeplot(data=df,\n",
    "                    x=column_x,\n",
    "                    hue=hue,\n",
    "                    fill=True,\n",
    "                    common_norm=False,\n",
    "                    palette=[\"blue\", 'red'],\n",
    "                    alpha=.5,\n",
    "                    linewidth=0)\n",
    "    plt.xlim(x_limit[0], x_limit[1])\n",
    "    plt.ylim(y_limit[0], y_limit[1])\n",
    "    plt.legend('',frameon=False)\n",
    "    p.set(xlabel=None)  # remove the axis label   \n",
    "    p.set(ylabel=None)  # remove the axis label    \n",
    "    p.yaxis.major_tick_line_color = None\n",
    "    p.yaxis.minor_tick_line_color = None\n",
    "    if show_x_ticks:\n",
    "        p.set(xticks=[np.log(1000), np.log(10000), np.log(100000)])\n",
    "        p.set(xticklabels=['1000', '10000', '100000'])\n",
    "    else:\n",
    "        p.xaxis.major_tick_line_color = None\n",
    "        p.xaxis.minor_tick_line_color = None\n",
    "        p.set(xticklabels=[])\n",
    "    p.set(yticklabels=[])  # remove the tick labels\n",
    "    plt.savefig(output, format=\"svg\")\n",
    "    plt.close()\n",
    "\n",
    "groups = (\"Consensus\", \"SNV\", \"Concat\")\n",
    "for protein in (\"CREB1\", \"EGR1\"):\n",
    "    kde_x, kde_y = KDE_BOUNDS[protein]\n",
    "    protein_df = relabeled_dfs[protein].copy()\n",
    "    protein_df = pd.melt(protein_df, id_vars=[\"Name\", \"Group\"], value_vars=[\"Signal_Non_UV\", \"Signal_UV\"])\n",
    "    for group in groups:\n",
    "        rcParams['figure.figsize'] = 10, 2.5\n",
    "        plot_df = protein_df[protein_df[\"Group\"] == group]\n",
    "        plot_x_ticks = group == \"Consensus\"\n",
    "        kde_plot(plot_df,\n",
    "                 \"value\",\n",
    "                 \"variable\", \n",
    "                 f\"{OUTPUT_FOLDER_F4}/{protein}_{group}_kde_plot.svg\",\n",
    "                 kde_x,\n",
    "                 kde_y,\n",
    "                 plot_x_ticks)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
